{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "nyXy4lr7ghUz"
   },
   "source": [
    "## Misc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "id": "nhYM6MfBGLE7"
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import subprocess as sp\n",
    "\n",
    "from datapaths import *\n",
    "\n",
    "from Bio import SeqIO\n",
    "from collections import Counter\n",
    "from dataclasses import dataclass\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "!mkdir -p {D_DATA} {D_UTILS} {D_GENOME} {D_LIFTOVER} {D_TABLES} {D_FLIPONS_BED} {D_FLIPONS_FA} {D_CCRE} {D_RMSK} {D_IMG} {D_GEBR}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "ByjfR8Et84PS"
   },
   "source": [
    "## Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Get Kouzine ssDNA data.\"\"\"\n",
    "\n",
    "!wget -c -q --show-progress -P {D_DATA} https://www.ncbi.nlm.nih.gov/CBBresearch/Przytycka/software/nonbdna/nonB_DNA_ssDNA_enriched.tar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "id": "oZQ23x3cRzcC"
   },
   "outputs": [],
   "source": [
    "\"\"\"Get mm10 genome data.\"\"\"\n",
    "\n",
    "!wget -c -q --show-progress -O {F_CHROM_SIZES} http://hgdownload.cse.ucsc.edu/goldenpath/mm10/bigZips/mm10.chrom.sizes\n",
    "!wget -c -q --show-progress -P {D_GENOME} https://hgdownload.soe.ucsc.edu/goldenPath/mm10/bigZips/mm10.fa.gz\n",
    "!cd {D_GENOME} && gzip -cdn mm10.fa.gz > mm10.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Get mm10 gene annotation.\"\"\"\n",
    "\n",
    "!wget -c -q --show-progress -O {F_GENCODE} http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.annotation.gtf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Get liftOver tool files.\"\"\"\n",
    "\n",
    "!wget -c -q --show-progress -O {F_LIFTOVER} http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/liftOver\n",
    "!wget -c -q --show-progress -O {F_LIFTOVER_CHAIN} https://hgdownload.cse.ucsc.edu/goldenpath/mm9/liftOver/mm9ToMm10.over.chain.gz\n",
    "!chmod +x {F_LIFTOVER}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Get bigBedToBed.\"\"\"\n",
    "\n",
    "!wget -c -q --show-progress -O {F_BIGBEDTOBED} https://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/bigBedToBed\n",
    "!chmod +x {F_BIGBEDTOBED}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Get cCRE data.\"\"\"\n",
    "\n",
    "!chmod +x {F_BIGBEDTOBED}\n",
    "!{F_BIGBEDTOBED} http://hgdownload.soe.ucsc.edu/gbdb/mm10/encode3/ccre/encodeCcreCombined.bb stdout > {F_CCRE}\n",
    "\n",
    "(\n",
    "    pd.read_table(F_CCRE, header=None)\n",
    "    .assign(\n",
    "        name_preformatted=lambda x: x[9] + \"@\" + x[3],\n",
    "        name=lambda x: x[\"name_preformatted\"].str.replace(\n",
    "            \",CTCF-bound\", \" (CTCF-bound)\"\n",
    "        ),\n",
    "    )\n",
    "    .loc[:, [0, 1, 2, \"name\"]]\n",
    ").to_csv(F_CCRE, sep=\"\\t\", header=False, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Get rmsk data.\"\"\"\n",
    "\n",
    "!wget -c -q --show-progress -O {F_RMSK} https://hgdownload.soe.ucsc.edu/goldenPath/mm10/database/rmsk.txt.gz\n",
    "\n",
    "columns = [\n",
    "    \"bin\", \"swScore\", \"milliDiv\", \"milliDel\", \"milliIns\", \"genoName\", \"genoStart\", \"genoEnd\", \"genoLeft\", \"strand\", \"repName\", \"repClass\", \"repFamily\", \"repStart\", \"repEnd\", \"repLeft\", \"id\"\n",
    "]\n",
    "\n",
    "(\n",
    "    pd.read_table(F_RMSK, header=None, names=columns, compression=\"gzip\")\n",
    "    .query('repClass.str.contains(\"LTR\") or repClass.str.contains(\"LINE\")')\n",
    "    .assign(\n",
    "        name=lambda x: x[\"repClass\"] + \"@\" + x[\"repName\"] + \"(\" + x[\"strand\"] + \")\"\n",
    "    )\n",
    "    .loc[:, [\"genoName\", \"genoStart\", \"genoEnd\", \"name\", \"swScore\", \"strand\"]]\n",
    "    .to_csv(F_RMSK_LINE_LTR, sep=\"\\t\", header=None, index=None)\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "pvhqUOYgXsQm"
   },
   "source": [
    "## Preprocessing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "MRiEpXx8a_F9"
   },
   "source": [
    "### Uplift Kouzine peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "xBrspN0PPoWA",
    "outputId": "43861da6-99a3-43b5-ceb4-6ff147d60a16"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nonB_DNA_ssDNA_enriched/\n",
      "nonB_DNA_ssDNA_enriched/human_hg19/\n",
      "nonB_DNA_ssDNA_enriched/human_hg19/Raji_ssDNA_enriched_H-DNA.bed.gz\n",
      "nonB_DNA_ssDNA_enriched/human_hg19/Raji_ssDNA_enriched_Quadruplex.bed.gz\n",
      "nonB_DNA_ssDNA_enriched/human_hg19/Raji_ssDNA_enriched_SIDD.bed.gz\n",
      "nonB_DNA_ssDNA_enriched/human_hg19/Raji_ssDNA_enriched_Z-DNA.bed.gz\n",
      "nonB_DNA_ssDNA_enriched/mouse_mm9/\n",
      "nonB_DNA_ssDNA_enriched/mouse_mm9/actB_ssDNA_enriched_H-DNA.bed.gz\n",
      "nonB_DNA_ssDNA_enriched/mouse_mm9/actB_ssDNA_enriched_Quadruplex.bed.gz\n",
      "nonB_DNA_ssDNA_enriched/mouse_mm9/actB_ssDNA_enriched_SIDD.bed.gz\n",
      "nonB_DNA_ssDNA_enriched/mouse_mm9/actB_ssDNA_enriched_Z-DNA.bed.gz\n",
      "rm: cannot remove '/home/fpavlov/projects/article_conserved_miRNA/data/nonB_DNA_ssDNA_enriched/mouse_mm10': No such file or directory\n",
      "actB_ssDNA_enriched_Z-DNA.bed -> mm10.actb_ssdna_enriched_z-dna.bed\n",
      "Reading liftover chains\n",
      "Mapping coordinates\n",
      "2\n",
      "actB_ssDNA_enriched_H-DNA.bed -> mm10.actb_ssdna_enriched_h-dna.bed\n",
      "Reading liftover chains\n",
      "Mapping coordinates\n",
      "4\n",
      "actB_ssDNA_enriched_SIDD.bed -> mm10.actb_ssdna_enriched_sidd.bed\n",
      "Reading liftover chains\n",
      "Mapping coordinates\n",
      "2\n",
      "actB_ssDNA_enriched_Quadruplex.bed -> mm10.actb_ssdna_enriched_g4.bed\n",
      "Reading liftover chains\n",
      "Mapping coordinates\n",
      "5\n"
     ]
    }
   ],
   "source": [
    "\"\"\"Uplift from mm9 to mm10.\"\"\"\n",
    "\n",
    "!rm -r {D_DATA}/nonB_DNA_ssDNA_enriched\n",
    "!tar -xvf {D_DATA}/nonB_DNA_ssDNA_enriched.tar -C {D_DATA}\n",
    "!cd {D_DATA}/nonB_DNA_ssDNA_enriched/mouse_mm9 && ls | xargs gzip -df\n",
    "\n",
    "!rm -r {D_DATA}/nonB_DNA_ssDNA_enriched/mouse_mm10 ; mkdir -p {D_DATA}/nonB_DNA_ssDNA_enriched/mouse_mm10\n",
    "!rm -r {D_FLIPONS_BED} ; mkdir -p {D_FLIPONS_BED}\n",
    "\n",
    "for file in (D_DATA / 'nonB_DNA_ssDNA_enriched/mouse_mm9').iterdir():\n",
    "    new_file_path = Path(str(file).replace('mm9', 'mm10')).parent / (\"mm10.\" + file.name.replace('Quadruplex', 'G4').lower())\n",
    "    print(f\"{file.name} -> {new_file_path.name}\")\n",
    "\n",
    "    !tail -n +2 {file} > temp && mv temp {file}\n",
    "    !{F_LIFTOVER} {file} {F_LIFTOVER_CHAIN} {new_file_path} unmapped_regions.bed\n",
    "    !cat unmapped_regions.bed | grep \"#Deleted in new\" | wc -l ; rm unmapped_regions.bed\n",
    "\n",
    "    !sort --unique -k1,1 -k2,2n -o {new_file_path} {new_file_path}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "iEm3x-FHbBPR"
   },
   "source": [
    "### Slops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mm10.actb_ssdna_enriched_g4.bed\n",
      "mm10.actb_ssdna_enriched_z-dna.bed\n",
      "mm10.actb_ssdna_enriched_sidd.bed\n",
      "mm10.actb_ssdna_enriched_h-dna.bed\n"
     ]
    }
   ],
   "source": [
    "\"\"\"Calculate slops for each bedfile (required to find overlaps with miRNA).\"\"\"\n",
    "\n",
    "slop_values = [100, 200, 500]\n",
    "for file in (D_DATA / \"nonB_DNA_ssDNA_enriched/mouse_mm10\").iterdir():\n",
    "    print(file.name)\n",
    "    flipon_name = file.name.split(\"ed_\")[1][:-4]\n",
    "    path_to_flipon = D_FLIPONS_BED / file.name\n",
    "\n",
    "    flipon_data = (\n",
    "        pd.read_table(file, header=None).astype({1: int, 2: int}).sort_values([0, 1])\n",
    "    )\n",
    "    flipon_data[3] = flipon_data[0] + \":\" + flipon_data[1].astype(str) + \"-\" + flipon_data[2].astype(str)\n",
    "    if len(flipon_data.columns) == 4:\n",
    "        flipon_data[4] = 0\n",
    "        flipon_data[5] = \".\"\n",
    "    flipon_data.to_csv(path_to_flipon, sep=\"\\t\", header=False, index=False)\n",
    "\n",
    "    for slop_val in slop_values:\n",
    "        path_to_slop = str(path_to_flipon).replace(\".bed\", f\".slop{slop_val}.bed\")\n",
    "        !bedtools slop -i {path_to_flipon} -g {F_CHROM_SIZES} -b {slop_val} > {path_to_slop}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "9LVdK7sudIC5"
   },
   "source": [
    "### Intersections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "Jlk2uFRHYEeS",
    "outputId": "701b830c-0b36-42ce-9292-cff6635b7225"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mm10.actb_ssdna_enriched_z-dna_and_sidd.bed: 256\n",
      "mm10.actb_ssdna_enriched_z-dna_and_sidd.slop100.bed: 758\n",
      "mm10.actb_ssdna_enriched_z-dna_and_sidd.slop200.bed: 1,117\n",
      "mm10.actb_ssdna_enriched_z-dna_and_sidd.slop500.bed: 1,802\n",
      "mm10.actb_ssdna_enriched_g4_and_sidd.bed: 1,374\n",
      "mm10.actb_ssdna_enriched_g4_and_sidd.slop100.bed: 2,250\n",
      "mm10.actb_ssdna_enriched_g4_and_sidd.slop200.bed: 2,534\n",
      "mm10.actb_ssdna_enriched_g4_and_sidd.slop500.bed: 3,089\n",
      "mm10.actb_ssdna_enriched_h-dna_and_sidd.bed: 1,627\n",
      "mm10.actb_ssdna_enriched_h-dna_and_sidd.slop100.bed: 2,016\n",
      "mm10.actb_ssdna_enriched_h-dna_and_sidd.slop200.bed: 2,249\n",
      "mm10.actb_ssdna_enriched_h-dna_and_sidd.slop500.bed: 2,560\n"
     ]
    }
   ],
   "source": [
    "\"\"\"Intersect other flipons with SIDD.\"\"\"\n",
    "\n",
    "kouzine_data = [\n",
    "    D_FLIPONS_BED / \"mm10.actb_ssdna_enriched_z-dna.bed\",\n",
    "    D_FLIPONS_BED / \"mm10.actb_ssdna_enriched_g4.bed\",\n",
    "    D_FLIPONS_BED / \"mm10.actb_ssdna_enriched_h-dna.bed\",\n",
    "]\n",
    "\n",
    "sidd_slops = [\n",
    "    D_FLIPONS_BED / \"mm10.actb_ssdna_enriched_sidd.bed\",\n",
    "    *[D_FLIPONS_BED / f\"mm10.actb_ssdna_enriched_sidd.slop{x}.bed\" for x in slop_values],\n",
    "]\n",
    "\n",
    "for a in kouzine_data:\n",
    "    for b in sidd_slops:\n",
    "        path_to_intersection = str(a).replace(\n",
    "            \".bed\", f'_and{str(b).split(\"enriched\")[1]}'\n",
    "        )\n",
    "        with open(path_to_intersection, \"w\") as f_out:\n",
    "            sp.run(\n",
    "                [\"bedtools\", \"intersect\", \"-a\", a, \"-b\", b, \"-u\"],\n",
    "                check=True,\n",
    "                stdout=f_out,\n",
    "            )\n",
    "\n",
    "        shape = sp.run(\n",
    "            f\"wc -l {path_to_intersection}\",\n",
    "            shell=True,\n",
    "            capture_output=True,\n",
    "            encoding=\"utf-8\",\n",
    "        ).stdout\n",
    "        print(f\"{Path(shape.split()[1]).name}: {int(shape.split()[0]):,d}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "index file /home/fpavlov/projects/article_conserved_miRNA/data/genome/mm10.fa.fai not found, generating...\n"
     ]
    }
   ],
   "source": [
    "\"\"\"Calculate fasta files.\"\"\"\n",
    "\n",
    "!rm -r {F_GENOME}.fai\n",
    "!rm -r {D_FLIPONS_FA} ; mkdir -p {D_FLIPONS_FA}\n",
    "\n",
    "for file in D_FLIPONS_BED.iterdir():\n",
    "    path_fa = D_FLIPONS_FA / file.name.replace('bed', 'fa')\n",
    "    with open(path_fa, 'w') as f_out:\n",
    "        sp.run([\"bedtools\", \"getfasta\", \"-fi\", F_GENOME, \"-bed\", file], check=True, stdout=f_out)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Map flipons to miRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "@dataclass\n",
    "class Data:\n",
    "    name: str\n",
    "    path_fa: str\n",
    "    path_bed: str\n",
    "    path_bed_200: str\n",
    "    shape: int = None\n",
    "    path_ccre_intersection: str = None\n",
    "    path_rmsk_intersection: str = None\n",
    "\n",
    "    def __post_init__(self):\n",
    "        self.shape = self._get_shape()\n",
    "        self.path_ccre_intersection = D_FLIPONS_BED / self.path_bed_200.name.replace('.bed', '_and_ccre.bed')\n",
    "        self.path_rmsk_intersection = D_FLIPONS_BED / self.path_bed_200.name.replace('.bed', '_and_rmsk.bed')\n",
    "\n",
    "    def _get_shape(self):\n",
    "        with open(self.path_bed, \"r\") as f_in:\n",
    "            return sum(1 for _ in f_in)\n",
    "\n",
    "\n",
    "flipon_data = [\n",
    "    Data(\n",
    "        \"g4\",\n",
    "        D_FLIPONS_FA / \"mm10.actb_ssdna_enriched_g4.fa\",\n",
    "        D_FLIPONS_BED / \"mm10.actb_ssdna_enriched_g4.bed\",\n",
    "        D_FLIPONS_BED / \"mm10.actb_ssdna_enriched_g4.slop200.bed\",\n",
    "    ),\n",
    "    Data(\n",
    "        \"sidd\",\n",
    "        D_FLIPONS_FA / \"mm10.actb_ssdna_enriched_sidd.fa\",\n",
    "        D_FLIPONS_BED / \"mm10.actb_ssdna_enriched_sidd.bed\",\n",
    "        D_FLIPONS_BED / \"mm10.actb_ssdna_enriched_sidd.slop200.bed\",\n",
    "    ),\n",
    "    Data(\n",
    "        \"z-dna\",\n",
    "        D_FLIPONS_FA / \"mm10.actb_ssdna_enriched_z-dna.fa\",\n",
    "        D_FLIPONS_BED / \"mm10.actb_ssdna_enriched_z-dna.bed\",\n",
    "        D_FLIPONS_BED / \"mm10.actb_ssdna_enriched_z-dna.slop200.bed\",\n",
    "    ),\n",
    "    Data(\n",
    "        \"h-dna\",\n",
    "        D_FLIPONS_FA / \"mm10.actb_ssdna_enriched_h-dna.fa\",\n",
    "        D_FLIPONS_BED / \"mm10.actb_ssdna_enriched_h-dna.bed\",\n",
    "        D_FLIPONS_BED / \"mm10.actb_ssdna_enriched_h-dna.slop200.bed\",\n",
    "    ),\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_mirna_counts(mirna_counts: list):\n",
    "    \"\"\"Count each miRNA occurrence for each region.\"\"\"\n",
    "    res = []\n",
    "    for region in mirna_counts:\n",
    "        temp_res = []\n",
    "        for mirna_family, counts in region.items():\n",
    "            temp_res.append(f\"{mirna_family} ({counts:,d})\")\n",
    "        res.append(', '.join(sorted(temp_res)))\n",
    "    return res\n",
    "\n",
    "def reverse_complement(x: str):\n",
    "    t_d = {\"A\": \"T\", \"C\": \"G\", \"T\": \"A\", \"G\": \"C\", \"N\": \"N\"}\n",
    "    return \"\".join([t_d[y] for y in x[::-1]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(177, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Conservation</th>\n",
       "      <th>miRNA gene family</th>\n",
       "      <th># of genes</th>\n",
       "      <th>Seed region</th>\n",
       "      <th>Seed region (RC)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Bilateria</td>\n",
       "      <td>miR-34/449</td>\n",
       "      <td>6</td>\n",
       "      <td>[GGCAGTG]</td>\n",
       "      <td>[CACTGCC]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Bilateria</td>\n",
       "      <td>miR-96/1271</td>\n",
       "      <td>2</td>\n",
       "      <td>[TTGGCAC]</td>\n",
       "      <td>[GTGCCAA]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Bilateria</td>\n",
       "      <td>miR-99/100</td>\n",
       "      <td>3</td>\n",
       "      <td>[ACCCGTA]</td>\n",
       "      <td>[TACGGGT]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Bilateria</td>\n",
       "      <td>miR-124</td>\n",
       "      <td>3</td>\n",
       "      <td>[AAGGCAC, TAAGGCA]</td>\n",
       "      <td>[GTGCCTT, TGCCTTA]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Bilateria</td>\n",
       "      <td>miR-125</td>\n",
       "      <td>3</td>\n",
       "      <td>[CCCTGAG]</td>\n",
       "      <td>[CTCAGGG]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Bilateria</td>\n",
       "      <td>miR-133</td>\n",
       "      <td>3</td>\n",
       "      <td>[TGGTCCC, TTGGTCC]</td>\n",
       "      <td>[GGGACCA, GGACCAA]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Bilateria</td>\n",
       "      <td>miR-153</td>\n",
       "      <td>2</td>\n",
       "      <td>[TGCATAG]</td>\n",
       "      <td>[CTATGCA]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Bilateria</td>\n",
       "      <td>miR-183</td>\n",
       "      <td>1</td>\n",
       "      <td>[ATGGCAC, TGGCACT]</td>\n",
       "      <td>[GTGCCAT, AGTGCCA]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Bilateria</td>\n",
       "      <td>miR-184</td>\n",
       "      <td>1</td>\n",
       "      <td>[GGACGGA]</td>\n",
       "      <td>[TCCGTCC]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Bilateria</td>\n",
       "      <td>miR-190</td>\n",
       "      <td>2</td>\n",
       "      <td>[GATATGT]</td>\n",
       "      <td>[ACATATC]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Conservation miRNA gene family  # of genes         Seed region  \\\n",
       "10    Bilateria        miR-34/449           6           [GGCAGTG]   \n",
       "11    Bilateria       miR-96/1271           2           [TTGGCAC]   \n",
       "12    Bilateria        miR-99/100           3           [ACCCGTA]   \n",
       "13    Bilateria           miR-124           3  [AAGGCAC, TAAGGCA]   \n",
       "14    Bilateria           miR-125           3           [CCCTGAG]   \n",
       "15    Bilateria           miR-133           3  [TGGTCCC, TTGGTCC]   \n",
       "16    Bilateria           miR-153           2           [TGCATAG]   \n",
       "17    Bilateria           miR-183           1  [ATGGCAC, TGGCACT]   \n",
       "18    Bilateria           miR-184           1           [GGACGGA]   \n",
       "19    Bilateria           miR-190           2           [GATATGT]   \n",
       "\n",
       "      Seed region (RC)  \n",
       "10           [CACTGCC]  \n",
       "11           [GTGCCAA]  \n",
       "12           [TACGGGT]  \n",
       "13  [GTGCCTT, TGCCTTA]  \n",
       "14           [CTCAGGG]  \n",
       "15  [GGGACCA, GGACCAA]  \n",
       "16           [CTATGCA]  \n",
       "17  [GTGCCAT, AGTGCCA]  \n",
       "18           [TCCGTCC]  \n",
       "19           [ACATATC]  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kwargs = {\n",
    "    \"Seed region\": lambda x: x[\"Seed region\"].str.replace(\"U\", \"T\").str.split(\", \"),\n",
    "    \"Seed region (RC)\": lambda x: x[\"Seed region\"].apply(\n",
    "        lambda x: [reverse_complement(y) for y in x]\n",
    "    ),\n",
    "}\n",
    "mirna_df = pd.read_table(F_MIRNA_LIST).assign(**kwargs)\n",
    "\n",
    "print(mirna_df.shape)\n",
    "mirna_df.iloc[10:20, :]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ee3f56faeb0c40039751c531b98b6e56",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "g4: 20,253 regions\n",
      "sidd: 15,296 regions\n",
      "z-dna: 25,059 regions\n",
      "h-dna: 17,100 regions\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Flipon</th>\n",
       "      <th>Coordinates</th>\n",
       "      <th>miRNA (+)</th>\n",
       "      <th>miRNA (-)</th>\n",
       "      <th>miRNA (intersection)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr10:100015775-100015790</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr10:100016214-100016246</td>\n",
       "      <td></td>\n",
       "      <td>miR-149 (1)</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr10:10012216-10012266</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr10:100147225-100147247</td>\n",
       "      <td>miR-365 (1)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr10:100401253-100401285</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77703</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrY:4200098-4200122</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77704</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrY:4208594-4208610</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77705</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrY:4223501-4223522</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77706</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrY:4231086-4231113</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77707</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrY:4235026-4235042</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77708 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Flipon                Coordinates    miRNA (+)    miRNA (-)  \\\n",
       "0         g4  chr10:100015775-100015790                             \n",
       "1         g4  chr10:100016214-100016246               miR-149 (1)   \n",
       "2         g4    chr10:10012216-10012266                             \n",
       "3         g4  chr10:100147225-100147247  miR-365 (1)                \n",
       "4         g4  chr10:100401253-100401285                             \n",
       "...      ...                        ...          ...          ...   \n",
       "77703  h-dna       chrY:4200098-4200122                             \n",
       "77704  h-dna       chrY:4208594-4208610                             \n",
       "77705  h-dna       chrY:4223501-4223522                             \n",
       "77706  h-dna       chrY:4231086-4231113                             \n",
       "77707  h-dna       chrY:4235026-4235042                             \n",
       "\n",
       "      miRNA (intersection)  \n",
       "0                           \n",
       "1                           \n",
       "2                           \n",
       "3                           \n",
       "4                           \n",
       "...                    ...  \n",
       "77703                       \n",
       "77704                       \n",
       "77705                       \n",
       "77706                       \n",
       "77707                       \n",
       "\n",
       "[77708 rows x 5 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flipon_to_mirna_df = pd.DataFrame()\n",
    "\n",
    "for flipon in tqdm(flipon_data):\n",
    "    print(f\"{flipon.name}: {flipon.shape:,d} regions\")\n",
    "\n",
    "    mirna_list_fwd = [{} for _ in range(flipon.shape)]\n",
    "    mirna_list_rcm = [{} for _ in range(flipon.shape)]\n",
    "\n",
    "    fasta = list(SeqIO.parse(flipon.path_fa, \"fasta\"))\n",
    "    fasta_fwd_seq = [x.seq.upper() for x in fasta]\n",
    "\n",
    "    for j, rec_fwd in enumerate(fasta_fwd_seq):\n",
    "        counter_fwd = Counter()\n",
    "        counter_rcm = Counter()\n",
    "        for mirna_family, motiffs_f, motiffs_rc in mirna_df[\n",
    "            [\"miRNA gene family\", \"Seed region\", \"Seed region (RC)\"]\n",
    "        ].values:\n",
    "            for motiff_fwd, motiff_rcm in zip(motiffs_f, motiffs_rc):\n",
    "                counter_fwd[mirna_family] += rec_fwd.count(motiff_fwd)\n",
    "                counter_rcm[mirna_family] += rec_fwd.count(motiff_rcm)\n",
    "        mirna_list_fwd[j] = +counter_fwd\n",
    "        mirna_list_rcm[j] = +counter_rcm\n",
    "\n",
    "    kwargs = {\n",
    "        \"miRNA (intersection)\": lambda x: x[[\"miRNA (+)\", \"miRNA (-)\"]].apply(\n",
    "            lambda y: \", \".join(sorted(set(y[0]) & set(y[1]))), axis=1\n",
    "        ),\n",
    "        \"miRNA (+)\": lambda x: format_mirna_counts(x[\"miRNA (+)\"].values),\n",
    "        \"miRNA (-)\": lambda x: format_mirna_counts(x[\"miRNA (-)\"].values),\n",
    "    }\n",
    "    df = pd.DataFrame(\n",
    "        {\n",
    "            \"Flipon\": flipon.name,\n",
    "            \"Coordinates\": [x.id for x in fasta],\n",
    "            \"miRNA (+)\": mirna_list_fwd,\n",
    "            \"miRNA (-)\": mirna_list_rcm,\n",
    "        }\n",
    "    ).assign(**kwargs)\n",
    "\n",
    "    flipon_to_mirna_df = pd.concat(\n",
    "        [flipon_to_mirna_df, df.sort_values(\"Coordinates\")], ignore_index=True\n",
    "    )\n",
    "\n",
    "flipon_to_mirna_df.to_csv(F_FLIPON_TO_MIRNA, sep=\"\\t\", quoting=2, index=False)\n",
    "flipon_to_mirna_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Export mirna-enriched flipons.\"\"\"\n",
    "\n",
    "for flipon_type, group_df in flipon_to_mirna_df.groupby(\"Flipon\"):\n",
    "    (\n",
    "        group_df.replace(\"\", None)\n",
    "        .dropna(subset=[\"miRNA (+)\", \"miRNA (-)\"], how=\"all\")[\"Coordinates\"]\n",
    "        .str.split(r\":|-\", regex=True, expand=True)\n",
    "        .astype({1: int, 2: int})\n",
    "        .sort_values([0, 1])\n",
    "        .to_csv(\n",
    "            D_FLIPONS_BED / f\"mm10.actb_ssdna_enriched_{flipon_type}_and_mirna.bed\",\n",
    "            sep=\"\\t\",\n",
    "            index=False,\n",
    "            header=False,\n",
    "        )\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Map flipons to cCRE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mm10.actb_ssdna_enriched_g4.slop200_and_ccre.bed: 14,637\n",
      "mm10.actb_ssdna_enriched_sidd.slop200_and_ccre.bed: 1,683\n",
      "mm10.actb_ssdna_enriched_z-dna.slop200_and_ccre.bed: 20,638\n",
      "mm10.actb_ssdna_enriched_h-dna.slop200_and_ccre.bed: 3,756\n"
     ]
    }
   ],
   "source": [
    "for flipon in flipon_data:\n",
    "    with open(flipon.path_ccre_intersection, \"w\") as f_out:\n",
    "        sp.run(\n",
    "            [\"bedtools\", \"intersect\", \"-a\", flipon.path_bed_200, \"-b\", F_CCRE, \"-wo\"],\n",
    "            check=True,\n",
    "            stdout=f_out,\n",
    "        )\n",
    "        shape = sp.run(\n",
    "            f\"wc -l {flipon.path_ccre_intersection}\",\n",
    "            shell=True,\n",
    "            capture_output=True,\n",
    "            encoding=\"utf-8\",\n",
    "        ).stdout\n",
    "        print(f\"{Path(shape.split()[1]).name}: {int(shape.split()[0]):,d}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Flipon</th>\n",
       "      <th>Coordinates</th>\n",
       "      <th>cCRE (+-200bp)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3671869-3671902</td>\n",
       "      <td>pELS (CTCF-bound)@EM10E0431220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:4493714-4493748</td>\n",
       "      <td>dELS@EM10E0431244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:4571896-4571924</td>\n",
       "      <td>dELS (CTCF-bound)@EM10E0431262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:5018367-5018390</td>\n",
       "      <td>DNase-H3K4me3 (CTCF-bound)@EM10E0431331, PLS@E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:5019245-5019272</td>\n",
       "      <td>pELS@EM10E0431333, PLS (CTCF-bound)@EM10E0431334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24415</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrX:152769666-152769687</td>\n",
       "      <td>pELS (CTCF-bound)@EM10E0930778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24416</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrX:159987892-159987912</td>\n",
       "      <td>DNase-H3K4me3 (CTCF-bound)@EM10E0931180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24417</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrX:161717977-161717993</td>\n",
       "      <td>PLS@EM10E0931282, pELS@EM10E0931283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24418</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrX:162643118-162643148</td>\n",
       "      <td>pELS@EM10E0931392, pELS@EM10E0931393, pELS (CT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24419</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrX:166585870-166585891</td>\n",
       "      <td>PLS@EM10E0931752, DNase-H3K4me3@EM10E0931753</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24420 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Flipon               Coordinates  \\\n",
       "0         g4      chr1:3671869-3671902   \n",
       "1         g4      chr1:4493714-4493748   \n",
       "2         g4      chr1:4571896-4571924   \n",
       "3         g4      chr1:5018367-5018390   \n",
       "4         g4      chr1:5019245-5019272   \n",
       "...      ...                       ...   \n",
       "24415  h-dna  chrX:152769666-152769687   \n",
       "24416  h-dna  chrX:159987892-159987912   \n",
       "24417  h-dna  chrX:161717977-161717993   \n",
       "24418  h-dna  chrX:162643118-162643148   \n",
       "24419  h-dna  chrX:166585870-166585891   \n",
       "\n",
       "                                          cCRE (+-200bp)  \n",
       "0                         pELS (CTCF-bound)@EM10E0431220  \n",
       "1                                      dELS@EM10E0431244  \n",
       "2                         dELS (CTCF-bound)@EM10E0431262  \n",
       "3      DNase-H3K4me3 (CTCF-bound)@EM10E0431331, PLS@E...  \n",
       "4       pELS@EM10E0431333, PLS (CTCF-bound)@EM10E0431334  \n",
       "...                                                  ...  \n",
       "24415                     pELS (CTCF-bound)@EM10E0930778  \n",
       "24416            DNase-H3K4me3 (CTCF-bound)@EM10E0931180  \n",
       "24417                PLS@EM10E0931282, pELS@EM10E0931283  \n",
       "24418  pELS@EM10E0931392, pELS@EM10E0931393, pELS (CT...  \n",
       "24419       PLS@EM10E0931752, DNase-H3K4me3@EM10E0931753  \n",
       "\n",
       "[24420 rows x 3 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flipon_to_ccre_df = pd.DataFrame()\n",
    "\n",
    "for flipon in flipon_data:\n",
    "    kwargs = {\n",
    "        \"Flipon\": flipon.name,\n",
    "        \"cCRE (+-200bp)\": lambda df: df.groupby([3])[9].transform(\n",
    "            lambda x: \", \".join(x)\n",
    "        ),\n",
    "    }\n",
    "\n",
    "    df = (\n",
    "        pd.read_table(flipon.path_ccre_intersection, header=None)\n",
    "        .assign(**kwargs)\n",
    "        .drop_duplicates(subset=[3])[[\"Flipon\", 3, \"cCRE (+-200bp)\"]]\n",
    "        .rename(columns={3: \"Coordinates\"})\n",
    "    )\n",
    "\n",
    "    flipon_to_ccre_df = pd.concat([flipon_to_ccre_df, df], ignore_index=True)\n",
    "\n",
    "flipon_to_ccre_df.to_csv(F_FLIPON_TO_CCRE, index=False, sep=\"\\t\")\n",
    "flipon_to_ccre_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Export ccre-enriched flipons.\"\"\"\n",
    "\n",
    "for flipon_type, group_df in flipon_to_ccre_df.groupby(\"Flipon\"):\n",
    "    (\n",
    "        group_df.replace(\"\", None)\n",
    "        .dropna(subset=[\"cCRE (+-200bp)\"])[\"Coordinates\"]\n",
    "        .str.split(r\":|-\", regex=True, expand=True)\n",
    "        .astype({1: int, 2: int})\n",
    "        .sort_values([0, 1])\n",
    "        .to_csv(\n",
    "            D_FLIPONS_BED / f\"mm10.actb_ssdna_enriched_{flipon_type}_and_ccre.bed\",\n",
    "            sep=\"\\t\",\n",
    "            index=False,\n",
    "            header=False,\n",
    "        )\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Map flipons to LINE/LTRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mm10.actb_ssdna_enriched_g4.slop200_and_rmsk.bed: 13,336\n",
      "mm10.actb_ssdna_enriched_sidd.slop200_and_rmsk.bed: 23,477\n",
      "mm10.actb_ssdna_enriched_z-dna.slop200_and_rmsk.bed: 11,835\n",
      "mm10.actb_ssdna_enriched_h-dna.slop200_and_rmsk.bed: 15,871\n"
     ]
    }
   ],
   "source": [
    "for flipon in flipon_data:\n",
    "    with open(flipon.path_rmsk_intersection, \"w\") as f_out:\n",
    "        sp.run(\n",
    "            [\"bedtools\", \"intersect\", \"-a\", flipon.path_bed_200, \"-b\", F_RMSK_LINE_LTR, \"-wo\"],\n",
    "            check=True,\n",
    "            stdout=f_out,\n",
    "        )\n",
    "        shape = sp.run(\n",
    "            f\"wc -l {flipon.path_rmsk_intersection}\",\n",
    "            shell=True,\n",
    "            capture_output=True,\n",
    "            encoding=\"utf-8\",\n",
    "        ).stdout\n",
    "        print(f\"{Path(shape.split()[1]).name}: {int(shape.split()[0]):,d}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Flipon</th>\n",
       "      <th>Coordinates</th>\n",
       "      <th>LINE/LTR (+-200bp)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3014794-3014871</td>\n",
       "      <td>LINE@L1Md_F2(-), LINE@L1VL1(+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3099888-3099963</td>\n",
       "      <td>LINE@L1Md_F2(+), LTR@MTC(+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3287445-3287468</td>\n",
       "      <td>LINE@L1MD(-)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3472953-3472969</td>\n",
       "      <td>LINE@L1_Mus1(+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3535948-3535996</td>\n",
       "      <td>LINE@L1_Mus2(-), LINE@L1VL1(+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39775</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrY:4200098-4200122</td>\n",
       "      <td>LINE@L1MD(+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39776</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrY:4208594-4208610</td>\n",
       "      <td>LTR@ORR1E(+), LINE@L1Md_T(-)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39777</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrY:4223501-4223522</td>\n",
       "      <td>LINE@L1_Mus2(-)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39778</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrY:4231086-4231113</td>\n",
       "      <td>LTR@MuRRS4-int(-)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39779</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrY:4235026-4235042</td>\n",
       "      <td>LTR@MuRRS4-int(-), LTR@MuRRS4-int(-)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>39780 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Flipon           Coordinates                    LINE/LTR (+-200bp)\n",
       "0         g4  chr1:3014794-3014871        LINE@L1Md_F2(-), LINE@L1VL1(+)\n",
       "1         g4  chr1:3099888-3099963           LINE@L1Md_F2(+), LTR@MTC(+)\n",
       "2         g4  chr1:3287445-3287468                          LINE@L1MD(-)\n",
       "3         g4  chr1:3472953-3472969                       LINE@L1_Mus1(+)\n",
       "4         g4  chr1:3535948-3535996        LINE@L1_Mus2(-), LINE@L1VL1(+)\n",
       "...      ...                   ...                                   ...\n",
       "39775  h-dna  chrY:4200098-4200122                          LINE@L1MD(+)\n",
       "39776  h-dna  chrY:4208594-4208610          LTR@ORR1E(+), LINE@L1Md_T(-)\n",
       "39777  h-dna  chrY:4223501-4223522                       LINE@L1_Mus2(-)\n",
       "39778  h-dna  chrY:4231086-4231113                     LTR@MuRRS4-int(-)\n",
       "39779  h-dna  chrY:4235026-4235042  LTR@MuRRS4-int(-), LTR@MuRRS4-int(-)\n",
       "\n",
       "[39780 rows x 3 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flipon_to_repeat_df = pd.DataFrame()\n",
    "\n",
    "for flipon in flipon_data:\n",
    "    kwargs = {\n",
    "        \"Flipon\": flipon.name,\n",
    "        \"LINE/LTR (+-200bp)\": lambda df: df.groupby([3])[9].transform(\n",
    "            lambda x: \", \".join(x)\n",
    "        ),\n",
    "    }\n",
    "\n",
    "    df = (\n",
    "        pd.read_table(flipon.path_rmsk_intersection, header=None)\n",
    "        .assign(**kwargs)\n",
    "        .drop_duplicates(subset=[3])[[\"Flipon\", 3, \"LINE/LTR (+-200bp)\"]]\n",
    "        .rename(columns={3: \"Coordinates\"})\n",
    "    )\n",
    "\n",
    "    flipon_to_repeat_df = pd.concat([flipon_to_repeat_df, df], ignore_index=True)\n",
    "\n",
    "flipon_to_repeat_df.to_csv(F_FLIPON_TO_RMSK, index=False, sep=\"\\t\")\n",
    "flipon_to_repeat_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Export line/ltr-enriched flipons.\"\"\"\n",
    "\n",
    "for flipon_type, group_df in flipon_to_repeat_df.groupby(\"Flipon\"):\n",
    "    (\n",
    "        group_df.replace(\"\", None)\n",
    "        .dropna(subset=[\"LINE/LTR (+-200bp)\"])[\"Coordinates\"]\n",
    "        .str.split(r\":|-\", regex=True, expand=True)\n",
    "        .astype({1: int, 2: int})\n",
    "        .sort_values([0, 1])\n",
    "        .to_csv(\n",
    "            D_FLIPONS_BED / f\"mm10.actb_ssdna_enriched_{flipon_type}_and_line_ltr.bed\",\n",
    "            sep=\"\\t\",\n",
    "            index=False,\n",
    "            header=False,\n",
    "        )\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Track names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess as sp\n",
    "\n",
    "def get_color_by_name(name: str):\n",
    "    if 'G4' in name:\n",
    "        # orange\n",
    "        return \"245,139,0\"\n",
    "    if 'Z-DNA' in name:\n",
    "        # blue\n",
    "        return \"0,112,245\"\n",
    "    if 'SIDD' in name:\n",
    "        # red\n",
    "        return \"255,49,38\"\n",
    "    if 'H-DNA' in name:\n",
    "        # purple\n",
    "        return \"159,75,201\"\n",
    "\n",
    "for file in D_FLIPONS_BED.iterdir():\n",
    "    size = int(sp.run(f\"wc -l {file}\", shell=True, capture_output=True, encoding='utf-8').stdout.split(\" \")[0])\n",
    "    name = (\n",
    "        file.name.split(\"enriched_\")[1]\n",
    "        .replace(\".bed\", \"\")\n",
    "        .replace(\"_\", \" \")\n",
    "        .replace(\"and\", \"&\")\n",
    "        .replace(\".slop\", \"±\")\n",
    "        .upper()\n",
    "        .replace(\"MIRNA\", \"miRNA\")\n",
    "        .replace(\"LINE \", \"LINE/\")\n",
    "        .replace(\"CCRE\", \"cCRE\")\n",
    "    )\n",
    "    description = \"Kouzine experimental \" + name.split(\" \")[0]\n",
    "    if len(name.split(\" & \")) > 1:\n",
    "        description += \" and \" + name.split(\" & \")[1].replace(\n",
    "            \"miRNA\", \"conserved miRNA\"\n",
    "        ).replace(\"/LTR\", \"/LTR repeats\")\n",
    "    description += f\" ({size:,d})\"\n",
    "    color = get_color_by_name(name.split(' & ')[0])\n",
    "\n",
    "    track_info = f'track name=\"{name}\" description=\"{description}\" color={color}'\n",
    "\n",
    "    !printf '%s\\n%s\\n' '{track_info}' \"$(cat {file})\" > {file}\n"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "name": "2022-05-12_01_Kouzine_data_preprocessing.ipynb",
   "provenance": [],
   "toc_visible": true
  },
  "kernelspec": {
   "display_name": "Python 3.11.0 ('mirna')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  },
  "vscode": {
   "interpreter": {
    "hash": "aa4ca5e43b0b74394dc05f972f57eb4922712b4f4450a8ee4f1bf77784cf2ccc"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
