{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from Bio import SeqIO\n",
    "from tqdm.notebook import tqdm\n",
    "from dataclasses import dataclass\n",
    "from collections import Counter\n",
    "\n",
    "from datapaths import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "@dataclass\n",
    "class Data:\n",
    "    name: str\n",
    "    path_fa: str\n",
    "    path_bed: str\n",
    "    path_bed_200: str\n",
    "    shape: int = None\n",
    "    path_ccre_intersection: str = None\n",
    "    path_repeat_intersection: str = None\n",
    "\n",
    "    def __post_init__(self):\n",
    "        self.shape = self._get_shape()\n",
    "\n",
    "    def _get_shape(self):\n",
    "        with open(self.path_fa, \"r\") as f_in:\n",
    "            return sum(1 for _ in f_in) // 2\n",
    "\n",
    "\n",
    "flipon_data = [\n",
    "    Data(\n",
    "        \"g4\",\n",
    "        D_FLIPONS_FA / \"mm10.actb_ssdna_enriched_g4.fa\",\n",
    "        D_FLIPONS_BED / \"mm10.actb_ssdna_enriched_g4.bed\",\n",
    "        D_FLIPONS_BED / \"mm10.actb_ssdna_enriched_g4.slop200.bed\",\n",
    "    ),\n",
    "    Data(\n",
    "        \"sidd\",\n",
    "        D_FLIPONS_FA / \"mm10.actb_ssdna_enriched_sidd.fa\",\n",
    "        D_FLIPONS_BED / \"mm10.actb_ssdna_enriched_sidd.bed\",\n",
    "        D_FLIPONS_BED / \"mm10.actb_ssdna_enriched_sidd.slop200.bed\",\n",
    "    ),\n",
    "    Data(\n",
    "        \"z-dna\",\n",
    "        D_FLIPONS_FA / \"mm10.actb_ssdna_enriched_z-dna.fa\",\n",
    "        D_FLIPONS_BED / \"mm10.actb_ssdna_enriched_z-dna.bed\",\n",
    "        D_FLIPONS_BED / \"mm10.actb_ssdna_enriched_z-dna.slop200.bed\",\n",
    "    ),\n",
    "    Data(\n",
    "        \"h-dna\",\n",
    "        D_FLIPONS_FA / \"mm10.actb_ssdna_enriched_h-dna.fa\",\n",
    "        D_FLIPONS_BED / \"mm10.actb_ssdna_enriched_h-dna.bed\",\n",
    "        D_FLIPONS_BED / \"mm10.actb_ssdna_enriched_h-dna.slop200.bed\",\n",
    "    ),\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(177, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Conservation</th>\n",
       "      <th>miRNA gene family</th>\n",
       "      <th># of genes</th>\n",
       "      <th>Seed region</th>\n",
       "      <th>Seed region (RC)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Bilateria</td>\n",
       "      <td>miR-34/449</td>\n",
       "      <td>6</td>\n",
       "      <td>[GGCAGTG]</td>\n",
       "      <td>[CACTGCC]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Bilateria</td>\n",
       "      <td>miR-96/1271</td>\n",
       "      <td>2</td>\n",
       "      <td>[TTGGCAC]</td>\n",
       "      <td>[GTGCCAA]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Bilateria</td>\n",
       "      <td>miR-99/100</td>\n",
       "      <td>3</td>\n",
       "      <td>[ACCCGTA]</td>\n",
       "      <td>[TACGGGT]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Bilateria</td>\n",
       "      <td>miR-124</td>\n",
       "      <td>3</td>\n",
       "      <td>[AAGGCAC, TAAGGCA]</td>\n",
       "      <td>[GTGCCTT, TGCCTTA]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Bilateria</td>\n",
       "      <td>miR-125</td>\n",
       "      <td>3</td>\n",
       "      <td>[CCCTGAG]</td>\n",
       "      <td>[CTCAGGG]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Bilateria</td>\n",
       "      <td>miR-133</td>\n",
       "      <td>3</td>\n",
       "      <td>[TGGTCCC, TTGGTCC]</td>\n",
       "      <td>[GGGACCA, GGACCAA]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Bilateria</td>\n",
       "      <td>miR-153</td>\n",
       "      <td>2</td>\n",
       "      <td>[TGCATAG]</td>\n",
       "      <td>[CTATGCA]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Bilateria</td>\n",
       "      <td>miR-183</td>\n",
       "      <td>1</td>\n",
       "      <td>[ATGGCAC, TGGCACT]</td>\n",
       "      <td>[GTGCCAT, AGTGCCA]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Bilateria</td>\n",
       "      <td>miR-184</td>\n",
       "      <td>1</td>\n",
       "      <td>[GGACGGA]</td>\n",
       "      <td>[TCCGTCC]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Bilateria</td>\n",
       "      <td>miR-190</td>\n",
       "      <td>2</td>\n",
       "      <td>[GATATGT]</td>\n",
       "      <td>[ACATATC]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Conservation miRNA gene family  # of genes         Seed region  \\\n",
       "10    Bilateria        miR-34/449           6           [GGCAGTG]   \n",
       "11    Bilateria       miR-96/1271           2           [TTGGCAC]   \n",
       "12    Bilateria        miR-99/100           3           [ACCCGTA]   \n",
       "13    Bilateria           miR-124           3  [AAGGCAC, TAAGGCA]   \n",
       "14    Bilateria           miR-125           3           [CCCTGAG]   \n",
       "15    Bilateria           miR-133           3  [TGGTCCC, TTGGTCC]   \n",
       "16    Bilateria           miR-153           2           [TGCATAG]   \n",
       "17    Bilateria           miR-183           1  [ATGGCAC, TGGCACT]   \n",
       "18    Bilateria           miR-184           1           [GGACGGA]   \n",
       "19    Bilateria           miR-190           2           [GATATGT]   \n",
       "\n",
       "      Seed region (RC)  \n",
       "10           [CACTGCC]  \n",
       "11           [GTGCCAA]  \n",
       "12           [TACGGGT]  \n",
       "13  [GTGCCTT, TGCCTTA]  \n",
       "14           [CTCAGGG]  \n",
       "15  [GGGACCA, GGACCAA]  \n",
       "16           [CTATGCA]  \n",
       "17  [GTGCCAT, AGTGCCA]  \n",
       "18           [TCCGTCC]  \n",
       "19           [ACATATC]  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def reverse_complement(x: str):\n",
    "    t_d = {\"A\": \"T\", \"C\": \"G\", \"T\": \"A\", \"G\": \"C\", \"N\": \"N\"}\n",
    "    return \"\".join([t_d[y] for y in x[::-1]])\n",
    "\n",
    "\n",
    "kwargs = {\n",
    "    \"Seed region\": lambda x: x[\"Seed region\"].str.replace(\"U\", \"T\").str.split(\", \"),\n",
    "    \"Seed region (RC)\": lambda x: x[\"Seed region\"].apply(\n",
    "        lambda x: [reverse_complement(y) for y in x]\n",
    "    ),\n",
    "}\n",
    "mirna_df = pd.read_table(F_MIRNA_LIST).assign(**kwargs)\n",
    "\n",
    "print(mirna_df.shape)\n",
    "mirna_df.iloc[10:20, :]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Flipons to miRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_mirna_counts(mirna_counts: list):\n",
    "    \"\"\"Count each miRNA occurrence for each region.\"\"\"\n",
    "    res = []\n",
    "    for region in mirna_counts:\n",
    "        temp_res = []\n",
    "        for mirna_family, counts in region.items():\n",
    "            temp_res.append(f\"{mirna_family} ({counts:,d})\")\n",
    "        res.append(', '.join(sorted(temp_res)))\n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3bc8dad99db74d3f9d43379835a1b69b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "g4: 20,253 regions\n",
      "20253 20253 20253\n",
      "sidd: 15,296 regions\n",
      "15296 15296 15296\n",
      "z-dna: 25,059 regions\n",
      "25059 25059 25059\n",
      "h-dna: 17,100 regions\n",
      "17100 17100 17100\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Flipon</th>\n",
       "      <th>Coordinates</th>\n",
       "      <th>miRNA (+)</th>\n",
       "      <th>miRNA (-)</th>\n",
       "      <th>miRNA (intersection)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr10:100015775-100015790</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr10:100016214-100016246</td>\n",
       "      <td></td>\n",
       "      <td>miR-149 (1)</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr10:10012216-10012266</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr10:100147225-100147247</td>\n",
       "      <td>miR-365 (1)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr10:100401253-100401285</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77703</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrY:4200098-4200122</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77704</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrY:4208594-4208610</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77705</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrY:4223501-4223522</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77706</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrY:4231086-4231113</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77707</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrY:4235026-4235042</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77708 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Flipon                Coordinates    miRNA (+)    miRNA (-)  \\\n",
       "0         g4  chr10:100015775-100015790                             \n",
       "1         g4  chr10:100016214-100016246               miR-149 (1)   \n",
       "2         g4    chr10:10012216-10012266                             \n",
       "3         g4  chr10:100147225-100147247  miR-365 (1)                \n",
       "4         g4  chr10:100401253-100401285                             \n",
       "...      ...                        ...          ...          ...   \n",
       "77703  h-dna       chrY:4200098-4200122                             \n",
       "77704  h-dna       chrY:4208594-4208610                             \n",
       "77705  h-dna       chrY:4223501-4223522                             \n",
       "77706  h-dna       chrY:4231086-4231113                             \n",
       "77707  h-dna       chrY:4235026-4235042                             \n",
       "\n",
       "      miRNA (intersection)  \n",
       "0                           \n",
       "1                           \n",
       "2                           \n",
       "3                           \n",
       "4                           \n",
       "...                    ...  \n",
       "77703                       \n",
       "77704                       \n",
       "77705                       \n",
       "77706                       \n",
       "77707                       \n",
       "\n",
       "[77708 rows x 5 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flipon_to_mirna_df = pd.DataFrame()\n",
    "\n",
    "for flipon in tqdm(flipon_data):\n",
    "    print(f\"{flipon.name}: {flipon.shape:,d} regions\")\n",
    "\n",
    "    mirna_list_fwd = [{} for _ in range(flipon.shape)]\n",
    "    mirna_list_rcm = [{} for _ in range(flipon.shape)]\n",
    "\n",
    "    fasta = list(SeqIO.parse(flipon.path_fa, \"fasta\"))\n",
    "    fasta_fwd_seq = [x.seq.upper() for x in fasta]\n",
    "\n",
    "    for j, rec_fwd in enumerate(fasta_fwd_seq):\n",
    "        counter_fwd = Counter()\n",
    "        counter_rcm = Counter()\n",
    "        for mirna_family, motiffs_f, motiffs_rc in mirna_df[\n",
    "            [\"miRNA gene family\", \"Seed region\", \"Seed region (RC)\"]\n",
    "        ].values:\n",
    "            for motiff_fwd, motiff_rcm in zip(motiffs_f, motiffs_rc):\n",
    "                counter_fwd[mirna_family] += rec_fwd.count(motiff_fwd)\n",
    "                counter_rcm[mirna_family] += rec_fwd.count(motiff_rcm)\n",
    "        mirna_list_fwd[j] = +counter_fwd\n",
    "        mirna_list_rcm[j] = +counter_rcm\n",
    "\n",
    "    kwargs = {\n",
    "        \"miRNA (intersection)\": lambda x: x[[\"miRNA (+)\", \"miRNA (-)\"]].apply(\n",
    "            lambda y: \", \".join(sorted(set(y[0]) & set(y[1]))), axis=1\n",
    "        ),\n",
    "        \"miRNA (+)\": lambda x: format_mirna_counts(x[\"miRNA (+)\"].values),\n",
    "        \"miRNA (-)\": lambda x: format_mirna_counts(x[\"miRNA (-)\"].values),\n",
    "    }\n",
    "    df = pd.DataFrame(\n",
    "        {\n",
    "            \"Flipon\": flipon.name,\n",
    "            \"Coordinates\": [x.id for x in fasta],\n",
    "            \"miRNA (+)\": mirna_list_fwd,\n",
    "            \"miRNA (-)\": mirna_list_rcm,\n",
    "        }\n",
    "    ).assign(**kwargs)\n",
    "\n",
    "    flipon_to_mirna_df = pd.concat(\n",
    "        [flipon_to_mirna_df, df.sort_values(\"Coordinates\")], ignore_index=True\n",
    "    )\n",
    "\n",
    "flipon_to_mirna_df.to_csv(F_FLIPON_TO_MIRNA, sep=\"\\t\", quoting=2)\n",
    "flipon_to_mirna_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Flipons to Gene Features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Flipon</th>\n",
       "      <th>Coordinates</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Strand</th>\n",
       "      <th>Gene Feature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3014794-3014871</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>+</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3099888-3099963</td>\n",
       "      <td>Gm26206</td>\n",
       "      <td>+</td>\n",
       "      <td>Promoter (2-3kb)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3287445-3287468</td>\n",
       "      <td>Gm18956</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3472953-3472969</td>\n",
       "      <td>Gm37686</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3535948-3535996</td>\n",
       "      <td>Gm7341</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77703</th>\n",
       "      <td>z-dna</td>\n",
       "      <td>chrY:3866251-3866287</td>\n",
       "      <td>Gm8521</td>\n",
       "      <td>+</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77704</th>\n",
       "      <td>z-dna</td>\n",
       "      <td>chrY:4195990-4196008</td>\n",
       "      <td>Gm29038</td>\n",
       "      <td>-</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77705</th>\n",
       "      <td>z-dna</td>\n",
       "      <td>chrY:4202845-4202862</td>\n",
       "      <td>Gm28191</td>\n",
       "      <td>+</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77706</th>\n",
       "      <td>z-dna</td>\n",
       "      <td>chrY:4203100-4203146</td>\n",
       "      <td>Gm28191</td>\n",
       "      <td>+</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77707</th>\n",
       "      <td>z-dna</td>\n",
       "      <td>chrY:4207093-4207113</td>\n",
       "      <td>Gm28191</td>\n",
       "      <td>+</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77708 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Flipon           Coordinates      Gene Name Gene Strand  \\\n",
       "0         g4  chr1:3014794-3014871  4933401J01Rik           +   \n",
       "1         g4  chr1:3099888-3099963        Gm26206           +   \n",
       "2         g4  chr1:3287445-3287468        Gm18956           +   \n",
       "3         g4  chr1:3472953-3472969        Gm37686           -   \n",
       "4         g4  chr1:3535948-3535996         Gm7341           +   \n",
       "...      ...                   ...            ...         ...   \n",
       "77703  z-dna  chrY:3866251-3866287         Gm8521           +   \n",
       "77704  z-dna  chrY:4195990-4196008        Gm29038           -   \n",
       "77705  z-dna  chrY:4202845-4202862        Gm28191           +   \n",
       "77706  z-dna  chrY:4203100-4203146        Gm28191           +   \n",
       "77707  z-dna  chrY:4207093-4207113        Gm28191           +   \n",
       "\n",
       "            Gene Feature  \n",
       "0      Distal Intergenic  \n",
       "1       Promoter (2-3kb)  \n",
       "2                 Intron  \n",
       "3                 Intron  \n",
       "4                 Intron  \n",
       "...                  ...  \n",
       "77703  Distal Intergenic  \n",
       "77704  Distal Intergenic  \n",
       "77705  Distal Intergenic  \n",
       "77706  Distal Intergenic  \n",
       "77707  Distal Intergenic  \n",
       "\n",
       "[77708 rows x 5 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flipon_to_gene_df = pd.read_table(F_FLIPON_TO_GENE)\n",
    "flipon_to_gene_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Flipons to cCRE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14637 /home/fpavlov/projects/article_conserved_miRNA/data/cCRE/mm10.actb_ssdna_enriched_g4.slop200_and_ccre.bed\n",
      "1683 /home/fpavlov/projects/article_conserved_miRNA/data/cCRE/mm10.actb_ssdna_enriched_sidd.slop200_and_ccre.bed\n",
      "20638 /home/fpavlov/projects/article_conserved_miRNA/data/cCRE/mm10.actb_ssdna_enriched_z-dna.slop200_and_ccre.bed\n",
      "3756 /home/fpavlov/projects/article_conserved_miRNA/data/cCRE/mm10.actb_ssdna_enriched_h-dna.slop200_and_ccre.bed\n"
     ]
    }
   ],
   "source": [
    "for flipon in flipon_data:\n",
    "    flipon.path_ccre_intersection = D_CCRE / flipon.path_bed_200.name.replace('.bed', '_and_ccre.bed')\n",
    "    !bedtools intersect -a {flipon.path_bed_200} -b {F_CCRE} -wo > {flipon.path_ccre_intersection}\n",
    "    !wc -l {flipon.path_ccre_intersection}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Flipon</th>\n",
       "      <th>Coordinates</th>\n",
       "      <th>cCRE (+-200bp)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3671869-3671902</td>\n",
       "      <td>pELS (CTCF-bound)@EM10E0431220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:4493714-4493748</td>\n",
       "      <td>dELS@EM10E0431244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:4571896-4571924</td>\n",
       "      <td>dELS (CTCF-bound)@EM10E0431262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:5018367-5018390</td>\n",
       "      <td>DNase-H3K4me3 (CTCF-bound)@EM10E0431331, PLS@E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:5019245-5019272</td>\n",
       "      <td>pELS@EM10E0431333, PLS (CTCF-bound)@EM10E0431334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24415</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrX:152769666-152769687</td>\n",
       "      <td>pELS (CTCF-bound)@EM10E0930778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24416</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrX:159987892-159987912</td>\n",
       "      <td>DNase-H3K4me3 (CTCF-bound)@EM10E0931180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24417</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrX:161717977-161717993</td>\n",
       "      <td>PLS@EM10E0931282, pELS@EM10E0931283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24418</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrX:162643118-162643148</td>\n",
       "      <td>pELS@EM10E0931392, pELS@EM10E0931393, pELS (CT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24419</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrX:166585870-166585891</td>\n",
       "      <td>PLS@EM10E0931752, DNase-H3K4me3@EM10E0931753</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24420 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Flipon               Coordinates  \\\n",
       "0         g4      chr1:3671869-3671902   \n",
       "1         g4      chr1:4493714-4493748   \n",
       "2         g4      chr1:4571896-4571924   \n",
       "3         g4      chr1:5018367-5018390   \n",
       "4         g4      chr1:5019245-5019272   \n",
       "...      ...                       ...   \n",
       "24415  h-dna  chrX:152769666-152769687   \n",
       "24416  h-dna  chrX:159987892-159987912   \n",
       "24417  h-dna  chrX:161717977-161717993   \n",
       "24418  h-dna  chrX:162643118-162643148   \n",
       "24419  h-dna  chrX:166585870-166585891   \n",
       "\n",
       "                                          cCRE (+-200bp)  \n",
       "0                         pELS (CTCF-bound)@EM10E0431220  \n",
       "1                                      dELS@EM10E0431244  \n",
       "2                         dELS (CTCF-bound)@EM10E0431262  \n",
       "3      DNase-H3K4me3 (CTCF-bound)@EM10E0431331, PLS@E...  \n",
       "4       pELS@EM10E0431333, PLS (CTCF-bound)@EM10E0431334  \n",
       "...                                                  ...  \n",
       "24415                     pELS (CTCF-bound)@EM10E0930778  \n",
       "24416            DNase-H3K4me3 (CTCF-bound)@EM10E0931180  \n",
       "24417                PLS@EM10E0931282, pELS@EM10E0931283  \n",
       "24418  pELS@EM10E0931392, pELS@EM10E0931393, pELS (CT...  \n",
       "24419       PLS@EM10E0931752, DNase-H3K4me3@EM10E0931753  \n",
       "\n",
       "[24420 rows x 3 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flipon_to_ccre_df = pd.DataFrame()\n",
    "columns = [\n",
    "    \"flipon_chr\",\n",
    "    \"flipon_start\",\n",
    "    \"flipon_end\",\n",
    "    \"Coordinates\",\n",
    "    \"flipon_score\",\n",
    "    \"flipon_strand\",\n",
    "    \"ccre_chr\",\n",
    "    \"ccre_start\",\n",
    "    \"ccre_end\",\n",
    "    \"ccre\",\n",
    "    \"intersection_size\",\n",
    "]\n",
    "\n",
    "for flipon in flipon_data:\n",
    "    kwargs = {\n",
    "        \"Flipon\": flipon.name,\n",
    "        \"cCRE (+-200bp)\": lambda df: df.groupby(\"Coordinates\")[\"ccre\"].transform(\n",
    "            lambda x: \", \".join(x)\n",
    "        ),\n",
    "    }\n",
    "\n",
    "    df = (\n",
    "        pd.read_table(flipon.path_ccre_intersection, header=None, names=columns)\n",
    "        .assign(**kwargs)\n",
    "        .drop_duplicates(subset=[\"Coordinates\"])[\n",
    "            [\"Flipon\", \"Coordinates\", \"cCRE (+-200bp)\"]\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    flipon_to_ccre_df = pd.concat([flipon_to_ccre_df, df], ignore_index=True)\n",
    "\n",
    "flipon_to_ccre_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Flipon to repeats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13336 /home/fpavlov/projects/article_conserved_miRNA/data/rmsk/mm10.actb_ssdna_enriched_g4.slop200_and_repeats.bed\n",
      "23477 /home/fpavlov/projects/article_conserved_miRNA/data/rmsk/mm10.actb_ssdna_enriched_sidd.slop200_and_repeats.bed\n",
      "11835 /home/fpavlov/projects/article_conserved_miRNA/data/rmsk/mm10.actb_ssdna_enriched_z-dna.slop200_and_repeats.bed\n",
      "15871 /home/fpavlov/projects/article_conserved_miRNA/data/rmsk/mm10.actb_ssdna_enriched_h-dna.slop200_and_repeats.bed\n"
     ]
    }
   ],
   "source": [
    "for flipon in flipon_data:\n",
    "    flipon.path_repeat_intersection = D_RMSK / flipon.path_bed_200.name.replace('.bed', '_and_repeats.bed')\n",
    "    !bedtools intersect -a {flipon.path_bed_200} -b {F_RMSK_LINE_LTR} -wo > {flipon.path_repeat_intersection}\n",
    "    !wc -l {flipon.path_repeat_intersection}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Flipon</th>\n",
       "      <th>Coordinates</th>\n",
       "      <th>LINE/LTR (+-200bp)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3014794-3014871</td>\n",
       "      <td>LINE@L1Md_F2(-), LINE@L1VL1(+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3099888-3099963</td>\n",
       "      <td>LINE@L1Md_F2(+), LTR@MTC(+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3287445-3287468</td>\n",
       "      <td>LINE@L1MD(-)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3472953-3472969</td>\n",
       "      <td>LINE@L1_Mus1(+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3535948-3535996</td>\n",
       "      <td>LINE@L1_Mus2(-), LINE@L1VL1(+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39775</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrY:4200098-4200122</td>\n",
       "      <td>LINE@L1MD(+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39776</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrY:4208594-4208610</td>\n",
       "      <td>LTR@ORR1E(+), LINE@L1Md_T(-)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39777</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrY:4223501-4223522</td>\n",
       "      <td>LINE@L1_Mus2(-)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39778</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrY:4231086-4231113</td>\n",
       "      <td>LTR@MuRRS4-int(-)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39779</th>\n",
       "      <td>h-dna</td>\n",
       "      <td>chrY:4235026-4235042</td>\n",
       "      <td>LTR@MuRRS4-int(-), LTR@MuRRS4-int(-)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>39780 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Flipon           Coordinates                    LINE/LTR (+-200bp)\n",
       "0         g4  chr1:3014794-3014871        LINE@L1Md_F2(-), LINE@L1VL1(+)\n",
       "1         g4  chr1:3099888-3099963           LINE@L1Md_F2(+), LTR@MTC(+)\n",
       "2         g4  chr1:3287445-3287468                          LINE@L1MD(-)\n",
       "3         g4  chr1:3472953-3472969                       LINE@L1_Mus1(+)\n",
       "4         g4  chr1:3535948-3535996        LINE@L1_Mus2(-), LINE@L1VL1(+)\n",
       "...      ...                   ...                                   ...\n",
       "39775  h-dna  chrY:4200098-4200122                          LINE@L1MD(+)\n",
       "39776  h-dna  chrY:4208594-4208610          LTR@ORR1E(+), LINE@L1Md_T(-)\n",
       "39777  h-dna  chrY:4223501-4223522                       LINE@L1_Mus2(-)\n",
       "39778  h-dna  chrY:4231086-4231113                     LTR@MuRRS4-int(-)\n",
       "39779  h-dna  chrY:4235026-4235042  LTR@MuRRS4-int(-), LTR@MuRRS4-int(-)\n",
       "\n",
       "[39780 rows x 3 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flipon_to_repeat_df = pd.DataFrame()\n",
    "columns = [\n",
    "    \"flipon_chr\",\n",
    "    \"flipon_start\",\n",
    "    \"flipon_end\",\n",
    "    \"Coordinates\",\n",
    "    \"flipon_score\",\n",
    "    \"flipon_strand\",\n",
    "    \"rep_chr\",\n",
    "    \"rep_start\",\n",
    "    \"rep_end\",\n",
    "    \"rep_name\",\n",
    "    \"rep_score\",\n",
    "    \"rep_strand\",\n",
    "    \"int_len\",\n",
    "]\n",
    "\n",
    "for flipon in flipon_data:\n",
    "    kwargs = {\n",
    "        \"Flipon\": flipon.name,\n",
    "        \"LINE/LTR (+-200bp)\": lambda df: df.groupby(\"Coordinates\")[\n",
    "            \"rep_name\"\n",
    "        ].transform(lambda x: \", \".join(x)),\n",
    "    }\n",
    "\n",
    "    df = (\n",
    "        pd.read_table(flipon.path_repeat_intersection, header=None, names=columns)\n",
    "        .assign(**kwargs)\n",
    "        .drop_duplicates(subset=[\"Coordinates\"])[\n",
    "            [\"Flipon\", \"Coordinates\", \"LINE/LTR (+-200bp)\"]\n",
    "        ]\n",
    "    )\n",
    "\n",
    "    flipon_to_repeat_df = pd.concat([flipon_to_repeat_df, df], ignore_index=True)\n",
    "\n",
    "flipon_to_repeat_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Flipon to everything"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_feature_group(x: list):\n",
    "    components = []\n",
    "    if (\n",
    "        x[0]\n",
    "        .replace(\" (CTCF-bound)\", \"\")\n",
    "        .replace(\"CTCF-only\", \"\")\n",
    "        .replace(\"CTCF\", \"\")\n",
    "        .strip(\", \")\n",
    "        != \"\"\n",
    "    ):\n",
    "        components.append(\"cCRE\")\n",
    "    if x[1] != \"\":\n",
    "        components.append(\"LINE/LTR\")\n",
    "    if \"CTCF\" in x[0]:\n",
    "        components.append(\"CTCF\")\n",
    "\n",
    "    return \" & \".join(components)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Flipon</th>\n",
       "      <th>Coordinates</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Strand</th>\n",
       "      <th>Gene Feature</th>\n",
       "      <th>miRNA (+)</th>\n",
       "      <th>miRNA (-)</th>\n",
       "      <th>miRNA (intersection)</th>\n",
       "      <th>cCRE (+-200bp)</th>\n",
       "      <th>LINE/LTR (+-200bp)</th>\n",
       "      <th>Feature Group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3014794-3014871</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>+</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td>miR-328 (1)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>LINE@L1Md_F2(-), LINE@L1VL1(+)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3099888-3099963</td>\n",
       "      <td>Gm26206</td>\n",
       "      <td>+</td>\n",
       "      <td>Promoter (2-3kb)</td>\n",
       "      <td></td>\n",
       "      <td>miR-328 (1)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>LINE@L1Md_F2(+), LTR@MTC(+)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3287445-3287468</td>\n",
       "      <td>Gm18956</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>LINE@L1MD(-)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3472953-3472969</td>\n",
       "      <td>Gm37686</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>LINE@L1_Mus1(+)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3535948-3535996</td>\n",
       "      <td>Gm7341</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>miR-129 (1)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>LINE@L1_Mus2(-), LINE@L1VL1(+)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77703</th>\n",
       "      <td>z-dna</td>\n",
       "      <td>chrY:3866251-3866287</td>\n",
       "      <td>Gm8521</td>\n",
       "      <td>+</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>LTR@ORR1E(-)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77704</th>\n",
       "      <td>z-dna</td>\n",
       "      <td>chrY:4195990-4196008</td>\n",
       "      <td>Gm29038</td>\n",
       "      <td>-</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>LINE@L1_Mus3(+), LTR@IAPLTR3(+)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77705</th>\n",
       "      <td>z-dna</td>\n",
       "      <td>chrY:4202845-4202862</td>\n",
       "      <td>Gm28191</td>\n",
       "      <td>+</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>LTR@IAPEY3-int(-), LTR@IAPEY_LTR(-)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77706</th>\n",
       "      <td>z-dna</td>\n",
       "      <td>chrY:4203100-4203146</td>\n",
       "      <td>Gm28191</td>\n",
       "      <td>+</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>LINE@L1_Mur1(-)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77707</th>\n",
       "      <td>z-dna</td>\n",
       "      <td>chrY:4207093-4207113</td>\n",
       "      <td>Gm28191</td>\n",
       "      <td>+</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>LINE@L1_Mus3(+), LINE@L1Md_A(+), LTR@ORR1B1(+)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77708 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Flipon           Coordinates      Gene Name Gene Strand  \\\n",
       "0         g4  chr1:3014794-3014871  4933401J01Rik           +   \n",
       "1         g4  chr1:3099888-3099963        Gm26206           +   \n",
       "2         g4  chr1:3287445-3287468        Gm18956           +   \n",
       "3         g4  chr1:3472953-3472969        Gm37686           -   \n",
       "4         g4  chr1:3535948-3535996         Gm7341           +   \n",
       "...      ...                   ...            ...         ...   \n",
       "77703  z-dna  chrY:3866251-3866287         Gm8521           +   \n",
       "77704  z-dna  chrY:4195990-4196008        Gm29038           -   \n",
       "77705  z-dna  chrY:4202845-4202862        Gm28191           +   \n",
       "77706  z-dna  chrY:4203100-4203146        Gm28191           +   \n",
       "77707  z-dna  chrY:4207093-4207113        Gm28191           +   \n",
       "\n",
       "            Gene Feature    miRNA (+)    miRNA (-) miRNA (intersection)  \\\n",
       "0      Distal Intergenic  miR-328 (1)                                     \n",
       "1       Promoter (2-3kb)               miR-328 (1)                        \n",
       "2                 Intron                                                  \n",
       "3                 Intron                                                  \n",
       "4                 Intron  miR-129 (1)                                     \n",
       "...                  ...          ...          ...                  ...   \n",
       "77703  Distal Intergenic                                                  \n",
       "77704  Distal Intergenic                                                  \n",
       "77705  Distal Intergenic                                                  \n",
       "77706  Distal Intergenic                                                  \n",
       "77707  Distal Intergenic                                                  \n",
       "\n",
       "      cCRE (+-200bp)                              LINE/LTR (+-200bp)  \\\n",
       "0                                     LINE@L1Md_F2(-), LINE@L1VL1(+)   \n",
       "1                                        LINE@L1Md_F2(+), LTR@MTC(+)   \n",
       "2                                                       LINE@L1MD(-)   \n",
       "3                                                    LINE@L1_Mus1(+)   \n",
       "4                                     LINE@L1_Mus2(-), LINE@L1VL1(+)   \n",
       "...              ...                                             ...   \n",
       "77703                                                   LTR@ORR1E(-)   \n",
       "77704                                LINE@L1_Mus3(+), LTR@IAPLTR3(+)   \n",
       "77705                            LTR@IAPEY3-int(-), LTR@IAPEY_LTR(-)   \n",
       "77706                                                LINE@L1_Mur1(-)   \n",
       "77707                 LINE@L1_Mus3(+), LINE@L1Md_A(+), LTR@ORR1B1(+)   \n",
       "\n",
       "      Feature Group  \n",
       "0          LINE/LTR  \n",
       "1          LINE/LTR  \n",
       "2          LINE/LTR  \n",
       "3          LINE/LTR  \n",
       "4          LINE/LTR  \n",
       "...             ...  \n",
       "77703      LINE/LTR  \n",
       "77704      LINE/LTR  \n",
       "77705      LINE/LTR  \n",
       "77706      LINE/LTR  \n",
       "77707      LINE/LTR  \n",
       "\n",
       "[77708 rows x 11 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flipon_to_data = (\n",
    "    flipon_to_gene_df.merge(\n",
    "        flipon_to_mirna_df, on=[\"Flipon\", \"Coordinates\"], how=\"left\"\n",
    "    )\n",
    "    .merge(flipon_to_ccre_df, on=[\"Flipon\", \"Coordinates\"], how=\"left\")\n",
    "    .merge(flipon_to_repeat_df, on=[\"Flipon\", \"Coordinates\"], how=\"left\")\n",
    "    .fillna(\"\")\n",
    ")\n",
    "flipon_to_data[\"Feature Group\"] = flipon_to_data.replace(\n",
    "    regex=r\"(@(.+?), )|(@(.+?)$)\", value=\",\"\n",
    ")[[\"cCRE (+-200bp)\", \"LINE/LTR (+-200bp)\"]].apply(get_feature_group, axis=1)\n",
    "\n",
    "flipon_to_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Flipon</th>\n",
       "      <th>Coordinates</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Strand</th>\n",
       "      <th>Gene Feature</th>\n",
       "      <th>miRNA (+)</th>\n",
       "      <th>miRNA (-)</th>\n",
       "      <th>miRNA (intersection)</th>\n",
       "      <th>cCRE (+-200bp)</th>\n",
       "      <th>LINE/LTR (+-200bp)</th>\n",
       "      <th>Feature Group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3014794-3014871</td>\n",
       "      <td>-</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>+</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td>miR-328 (1)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>LINE@L1Md_F2(-), LINE@L1VL1(+)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3099888-3099963</td>\n",
       "      <td>+</td>\n",
       "      <td>Gm26206</td>\n",
       "      <td>+</td>\n",
       "      <td>Promoter (2-3kb)</td>\n",
       "      <td></td>\n",
       "      <td>miR-328 (1)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>LINE@L1Md_F2(+), LTR@MTC(+)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3287445-3287468</td>\n",
       "      <td>-</td>\n",
       "      <td>Gm18956</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>LINE@L1MD(-)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3472953-3472969</td>\n",
       "      <td>-</td>\n",
       "      <td>Gm37686</td>\n",
       "      <td>-</td>\n",
       "      <td>Intron</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>LINE@L1_Mus1(+)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3535948-3535996</td>\n",
       "      <td>-</td>\n",
       "      <td>Gm7341</td>\n",
       "      <td>+</td>\n",
       "      <td>Intron</td>\n",
       "      <td>miR-129 (1)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>LINE@L1_Mus2(-), LINE@L1VL1(+)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77703</th>\n",
       "      <td>z-dna</td>\n",
       "      <td>chrY:3866251-3866287</td>\n",
       "      <td>.</td>\n",
       "      <td>Gm8521</td>\n",
       "      <td>+</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>LTR@ORR1E(-)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77704</th>\n",
       "      <td>z-dna</td>\n",
       "      <td>chrY:4195990-4196008</td>\n",
       "      <td>.</td>\n",
       "      <td>Gm29038</td>\n",
       "      <td>-</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>LINE@L1_Mus3(+), LTR@IAPLTR3(+)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77705</th>\n",
       "      <td>z-dna</td>\n",
       "      <td>chrY:4202845-4202862</td>\n",
       "      <td>.</td>\n",
       "      <td>Gm28191</td>\n",
       "      <td>+</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>LTR@IAPEY3-int(-), LTR@IAPEY_LTR(-)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77706</th>\n",
       "      <td>z-dna</td>\n",
       "      <td>chrY:4203100-4203146</td>\n",
       "      <td>.</td>\n",
       "      <td>Gm28191</td>\n",
       "      <td>+</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>LINE@L1_Mur1(-)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77707</th>\n",
       "      <td>z-dna</td>\n",
       "      <td>chrY:4207093-4207113</td>\n",
       "      <td>.</td>\n",
       "      <td>Gm28191</td>\n",
       "      <td>+</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>LINE@L1_Mus3(+), LINE@L1Md_A(+), LTR@ORR1B1(+)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77708 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Flipon           Coordinates Strand      Gene Name Gene Strand  \\\n",
       "0         g4  chr1:3014794-3014871      -  4933401J01Rik           +   \n",
       "1         g4  chr1:3099888-3099963      +        Gm26206           +   \n",
       "2         g4  chr1:3287445-3287468      -        Gm18956           +   \n",
       "3         g4  chr1:3472953-3472969      -        Gm37686           -   \n",
       "4         g4  chr1:3535948-3535996      -         Gm7341           +   \n",
       "...      ...                   ...    ...            ...         ...   \n",
       "77703  z-dna  chrY:3866251-3866287      .         Gm8521           +   \n",
       "77704  z-dna  chrY:4195990-4196008      .        Gm29038           -   \n",
       "77705  z-dna  chrY:4202845-4202862      .        Gm28191           +   \n",
       "77706  z-dna  chrY:4203100-4203146      .        Gm28191           +   \n",
       "77707  z-dna  chrY:4207093-4207113      .        Gm28191           +   \n",
       "\n",
       "            Gene Feature    miRNA (+)    miRNA (-) miRNA (intersection)  \\\n",
       "0      Distal Intergenic  miR-328 (1)                                     \n",
       "1       Promoter (2-3kb)               miR-328 (1)                        \n",
       "2                 Intron                                                  \n",
       "3                 Intron                                                  \n",
       "4                 Intron  miR-129 (1)                                     \n",
       "...                  ...          ...          ...                  ...   \n",
       "77703  Distal Intergenic                                                  \n",
       "77704  Distal Intergenic                                                  \n",
       "77705  Distal Intergenic                                                  \n",
       "77706  Distal Intergenic                                                  \n",
       "77707  Distal Intergenic                                                  \n",
       "\n",
       "      cCRE (+-200bp)                              LINE/LTR (+-200bp)  \\\n",
       "0                                     LINE@L1Md_F2(-), LINE@L1VL1(+)   \n",
       "1                                        LINE@L1Md_F2(+), LTR@MTC(+)   \n",
       "2                                                       LINE@L1MD(-)   \n",
       "3                                                    LINE@L1_Mus1(+)   \n",
       "4                                     LINE@L1_Mus2(-), LINE@L1VL1(+)   \n",
       "...              ...                                             ...   \n",
       "77703                                                   LTR@ORR1E(-)   \n",
       "77704                                LINE@L1_Mus3(+), LTR@IAPLTR3(+)   \n",
       "77705                            LTR@IAPEY3-int(-), LTR@IAPEY_LTR(-)   \n",
       "77706                                                LINE@L1_Mur1(-)   \n",
       "77707                 LINE@L1_Mus3(+), LINE@L1Md_A(+), LTR@ORR1B1(+)   \n",
       "\n",
       "      Feature Group  \n",
       "0          LINE/LTR  \n",
       "1          LINE/LTR  \n",
       "2          LINE/LTR  \n",
       "3          LINE/LTR  \n",
       "4          LINE/LTR  \n",
       "...             ...  \n",
       "77703      LINE/LTR  \n",
       "77704      LINE/LTR  \n",
       "77705      LINE/LTR  \n",
       "77706      LINE/LTR  \n",
       "77707      LINE/LTR  \n",
       "\n",
       "[77708 rows x 12 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flipon_beds = pd.DataFrame()\n",
    "for flipon in flipon_data:\n",
    "    flipon_bed = pd.read_table(flipon.path_bed, sep='\\t', skiprows=1, header=None, names=['chr', 'start', 'end', 'Coordinates', 'score', 'Strand']).assign(Flipon=flipon.name)\n",
    "    flipon_beds = pd.concat([flipon_beds, flipon_bed], ignore_index=True)\n",
    "\n",
    "flipon_to_data = flipon_to_data.merge(flipon_beds[['Flipon', 'Coordinates', 'Strand']], on=['Flipon', 'Coordinates'], how='left')\n",
    "\n",
    "flipon_to_data = flipon_to_data[['Flipon', 'Coordinates', 'Strand', 'Gene Name', 'Gene Strand', 'Gene Feature', 'miRNA (+)', 'miRNA (-)', 'miRNA (intersection)', 'cCRE (+-200bp)', 'LINE/LTR (+-200bp)', 'Feature Group']]    \n",
    "flipon_to_data    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# mirna only\n",
    "(\n",
    "    flipon_to_data\n",
    "    .replace(\"\", None)\n",
    "    .dropna(subset=[\"miRNA (+)\", \"miRNA (-)\"], how=\"all\")\n",
    "    .fillna(\"\")\n",
    "    .to_csv(F_FLIPON_TO_DATA_MIRNA, sep=\"\\t\", index=False)\n",
    ")\n",
    "\n",
    "flipon_to_data.to_csv(F_FLIPON_TO_DATA, sep='\\t', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.11.0 ('mirna')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "aa4ca5e43b0b74394dc05f972f57eb4922712b4f4450a8ee4f1bf77784cf2ccc"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
