{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import tempfile\n",
    "\n",
    "from gtfparse import read_gtf\n",
    "from collections import Counter\n",
    "from dataclasses import dataclass\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "from datapaths import *\n",
    "\n",
    "pd.set_option('display.max_rows', 4)\n",
    "pd.set_option('display.max_columns', None)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Flipon</th>\n",
       "      <th>Coordinates</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene Feature</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Strand</th>\n",
       "      <th>Gene Type</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Transcript ID</th>\n",
       "      <th>c. M miR (+)</th>\n",
       "      <th>c. M miR (-)</th>\n",
       "      <th>c. MJ miR (+)</th>\n",
       "      <th>c. MJ miR (-)</th>\n",
       "      <th>c. J miR (+)</th>\n",
       "      <th>c. J miR (-)</th>\n",
       "      <th>all M miR (+)</th>\n",
       "      <th>all M miR (-)</th>\n",
       "      <th>all MJ miR (+)</th>\n",
       "      <th>all MJ miR (-)</th>\n",
       "      <th>all J miR (+)</th>\n",
       "      <th>all J miR (-)</th>\n",
       "      <th>cCRE (+-200bp)</th>\n",
       "      <th>LINE/LTR (+-200bp)</th>\n",
       "      <th>Feature Group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3014794-3014871</td>\n",
       "      <td>-</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>+</td>\n",
       "      <td>TEC</td>\n",
       "      <td>ENSMUSG00000102693.1</td>\n",
       "      <td>ENSMUST00000193812.1</td>\n",
       "      <td>miR-328 (1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-486 (1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-6994 (1), miR-129b (1)</td>\n",
       "      <td>miR-92a-2 (1), miR-698/miR-7078 (1), miR-3572 ...</td>\n",
       "      <td>miR-486a/miR-486b (1)</td>\n",
       "      <td>miR-3076 (1), miR-6418 (1), miR-6987 (1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-7649 (1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LINE@L1Md_F2(-), LINE@L1VL1(+)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3099888-3099963</td>\n",
       "      <td>+</td>\n",
       "      <td>Promoter (2-3kb)</td>\n",
       "      <td>Gm26206</td>\n",
       "      <td>+</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>ENSMUSG00000064842.1</td>\n",
       "      <td>ENSMUST00000082908.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-328 (1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-365-1/miR-365-2 (1), miR-698/miR-7078 (1),...</td>\n",
       "      <td>miR-221 (1), miR-129b (1)</td>\n",
       "      <td>miR-6418 (1), miR-7067 (1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LINE@L1Md_F2(+), LTR@MTC(+)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77706</th>\n",
       "      <td>z-dna</td>\n",
       "      <td>chrY:4203100-4203146</td>\n",
       "      <td>.</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td>Gm28191</td>\n",
       "      <td>+</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>ENSMUSG00000100300.1</td>\n",
       "      <td>ENSMUST00000189112.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-493 (1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-466d/miR-466i (4), miR-466l (5)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-493 (1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LINE@L1_Mur1(-)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77707</th>\n",
       "      <td>z-dna</td>\n",
       "      <td>chrY:4207093-4207113</td>\n",
       "      <td>.</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td>Gm28191</td>\n",
       "      <td>+</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>ENSMUSG00000100300.1</td>\n",
       "      <td>ENSMUST00000189112.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-466a/miR-466e/miR-466p (1), miR-466d/miR-4...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LINE@L1_Mus3(+), LINE@L1Md_A(+), LTR@ORR1B1(+)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77708 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Flipon           Coordinates Strand       Gene Feature      Gene Name  \\\n",
       "0         g4  chr1:3014794-3014871      -  Distal Intergenic  4933401J01Rik   \n",
       "1         g4  chr1:3099888-3099963      +   Promoter (2-3kb)        Gm26206   \n",
       "...      ...                   ...    ...                ...            ...   \n",
       "77706  z-dna  chrY:4203100-4203146      .  Distal Intergenic        Gm28191   \n",
       "77707  z-dna  chrY:4207093-4207113      .  Distal Intergenic        Gm28191   \n",
       "\n",
       "      Gene Strand               Gene Type               Gene ID  \\\n",
       "0               +                     TEC  ENSMUSG00000102693.1   \n",
       "1               +                   snRNA  ENSMUSG00000064842.1   \n",
       "...           ...                     ...                   ...   \n",
       "77706           +  unprocessed_pseudogene  ENSMUSG00000100300.1   \n",
       "77707           +  unprocessed_pseudogene  ENSMUSG00000100300.1   \n",
       "\n",
       "              Transcript ID c. M miR (+) c. M miR (-) c. MJ miR (+)  \\\n",
       "0      ENSMUST00000193812.1  miR-328 (1)          NaN   miR-486 (1)   \n",
       "1      ENSMUST00000082908.1          NaN  miR-328 (1)           NaN   \n",
       "...                     ...          ...          ...           ...   \n",
       "77706  ENSMUST00000189112.1          NaN          NaN   miR-493 (1)   \n",
       "77707  ENSMUST00000189112.1          NaN          NaN           NaN   \n",
       "\n",
       "      c. MJ miR (-) c. J miR (+) c. J miR (-)  \\\n",
       "0               NaN          NaN          NaN   \n",
       "1               NaN          NaN          NaN   \n",
       "...             ...          ...          ...   \n",
       "77706           NaN          NaN          NaN   \n",
       "77707           NaN          NaN          NaN   \n",
       "\n",
       "                                           all M miR (+)  \\\n",
       "0                             miR-6994 (1), miR-129b (1)   \n",
       "1      miR-365-1/miR-365-2 (1), miR-698/miR-7078 (1),...   \n",
       "...                                                  ...   \n",
       "77706                miR-466d/miR-466i (4), miR-466l (5)   \n",
       "77707  miR-466a/miR-466e/miR-466p (1), miR-466d/miR-4...   \n",
       "\n",
       "                                           all M miR (-)  \\\n",
       "0      miR-92a-2 (1), miR-698/miR-7078 (1), miR-3572 ...   \n",
       "1                              miR-221 (1), miR-129b (1)   \n",
       "...                                                  ...   \n",
       "77706                                                NaN   \n",
       "77707                                                NaN   \n",
       "\n",
       "                   all MJ miR (+)                            all MJ miR (-)  \\\n",
       "0           miR-486a/miR-486b (1)  miR-3076 (1), miR-6418 (1), miR-6987 (1)   \n",
       "1      miR-6418 (1), miR-7067 (1)                                       NaN   \n",
       "...                           ...                                       ...   \n",
       "77706                 miR-493 (1)                                       NaN   \n",
       "77707                         NaN                                       NaN   \n",
       "\n",
       "      all J miR (+) all J miR (-) cCRE (+-200bp)  \\\n",
       "0               NaN  miR-7649 (1)            NaN   \n",
       "1               NaN           NaN            NaN   \n",
       "...             ...           ...            ...   \n",
       "77706           NaN           NaN            NaN   \n",
       "77707           NaN           NaN            NaN   \n",
       "\n",
       "                                   LINE/LTR (+-200bp) Feature Group  \n",
       "0                      LINE@L1Md_F2(-), LINE@L1VL1(+)      LINE/LTR  \n",
       "1                         LINE@L1Md_F2(+), LTR@MTC(+)      LINE/LTR  \n",
       "...                                               ...           ...  \n",
       "77706                                 LINE@L1_Mur1(-)      LINE/LTR  \n",
       "77707  LINE@L1_Mus3(+), LINE@L1Md_A(+), LTR@ORR1B1(+)      LINE/LTR  \n",
       "\n",
       "[77708 rows x 24 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flipon_to_data = pd.read_table(F_FLIPON_TO_DATA)\n",
    "flipon_to_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/fpavlov/miniconda3/envs/mirna/lib/python3.11/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "/home/fpavlov/miniconda3/envs/mirna/lib/python3.11/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The warn_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'level', 'mgi_id', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name', 'transcript_support_level', 'tag', 'havana_transcript', 'exon_number', 'exon_id', 'protein_id', 'ccdsid', 'ont']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>mgi_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>ont</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000102693.1</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1918292</td>\n",
       "      <td>OTTMUSG00000049935.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000102693.1</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1918292</td>\n",
       "      <td>OTTMUSG00000049935.1</td>\n",
       "      <td>ENSMUST00000193812.1</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTMUST00000127109.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1302161</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000064372.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>mt-Tp</td>\n",
       "      <td>3</td>\n",
       "      <td>MGI:102478</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000082423.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>mt-Tp-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1302162</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000064372.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>mt-Tp</td>\n",
       "      <td>3</td>\n",
       "      <td>MGI:102478</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000082423.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>mt-Tp-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000521550.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1302163 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname   source     feature    start      end  score strand frame  \\\n",
       "0          chr1   HAVANA        gene  3073253  3074322    NaN      +     0   \n",
       "1          chr1   HAVANA  transcript  3073253  3074322    NaN      +     0   \n",
       "...         ...      ...         ...      ...      ...    ...    ...   ...   \n",
       "1302161    chrM  ENSEMBL  transcript    15356    15422    NaN      -     0   \n",
       "1302162    chrM  ENSEMBL        exon    15356    15422    NaN      -     0   \n",
       "\n",
       "                      gene_id gene_type      gene_name level       mgi_id  \\\n",
       "0        ENSMUSG00000102693.1       TEC  4933401J01Rik     2  MGI:1918292   \n",
       "1        ENSMUSG00000102693.1       TEC  4933401J01Rik     2  MGI:1918292   \n",
       "...                       ...       ...            ...   ...          ...   \n",
       "1302161  ENSMUSG00000064372.1   Mt_tRNA          mt-Tp     3   MGI:102478   \n",
       "1302162  ENSMUSG00000064372.1   Mt_tRNA          mt-Tp     3   MGI:102478   \n",
       "\n",
       "                  havana_gene         transcript_id transcript_type  \\\n",
       "0        OTTMUSG00000049935.1                                         \n",
       "1        OTTMUSG00000049935.1  ENSMUST00000193812.1             TEC   \n",
       "...                       ...                   ...             ...   \n",
       "1302161                        ENSMUST00000082423.1         Mt_tRNA   \n",
       "1302162                        ENSMUST00000082423.1         Mt_tRNA   \n",
       "\n",
       "           transcript_name transcript_support_level    tag  \\\n",
       "0                                                            \n",
       "1        4933401J01Rik-201                       NA  basic   \n",
       "...                    ...                      ...    ...   \n",
       "1302161          mt-Tp-201                       NA  basic   \n",
       "1302162          mt-Tp-201                       NA  basic   \n",
       "\n",
       "            havana_transcript exon_number               exon_id protein_id  \\\n",
       "0                                                                            \n",
       "1        OTTMUST00000127109.1                                                \n",
       "...                       ...         ...                   ...        ...   \n",
       "1302161                                                                      \n",
       "1302162                                 1  ENSMUSE00000521550.1              \n",
       "\n",
       "        ccdsid ont  \n",
       "0                   \n",
       "1                   \n",
       "...        ...  ..  \n",
       "1302161             \n",
       "1302162             \n",
       "\n",
       "[1302163 rows x 25 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf = read_gtf(F_GENCODE)\n",
    "gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "@dataclass\n",
    "class Data:\n",
    "    name: str\n",
    "    path: str\n",
    "\n",
    "\n",
    "flipon_data = (\n",
    "    Data(\"g4\", D_FLIPONS_BED / \"mm10.actb_ssdna_enriched_g4.bed\"),\n",
    "    Data(\"z-dna\", D_FLIPONS_BED / \"mm10.actb_ssdna_enriched_z-dna.bed\"),\n",
    "    Data(\"sidd\", D_FLIPONS_BED / \"mm10.actb_ssdna_enriched_sidd.bed\"),\n",
    "    Data(\"h-dna\", D_FLIPONS_BED / \"mm10.actb_ssdna_enriched_h-dna.bed\"),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_overlaps_df(flipon_df: pd.DataFrame):\n",
    "    overlaps_df: pd.DataFrame\n",
    "    with tempfile.TemporaryDirectory() as tmpdir:\n",
    "        # 1. Get transcript coordinates        \n",
    "        transcript_list = flipon_df['Transcript ID'].unique()\n",
    "        annotated_transcripts = gtf[(gtf['feature']=='transcript') & gtf['transcript_id'].isin(transcript_list)].drop_duplicates('transcript_id')[['seqname', 'start', 'end', 'transcript_id', 'score', 'strand']]\n",
    "\n",
    "        # 2. Get TSS coordinates\n",
    "        annotated_transcripts['end'] = annotated_transcripts.apply(lambda row: row['start']+1 if row['strand']=='+' else row['end'], axis=1)\n",
    "        annotated_transcripts['start'] = annotated_transcripts.apply(lambda row: row['end']-1 if row['strand']=='-' else row['start'], axis=1)\n",
    "\n",
    "        path_tss = tmpdir + '/tss.bed'\n",
    "        annotated_transcripts.to_csv(path_tss, sep='\\t', header=False, index=False)\n",
    "\n",
    "        # 3. Get promoter coordinates\n",
    "        path_proximal_promoters = tmpdir + '/proximal_promoters.bed'\n",
    "        !bedtools slop -i {path_tss} -g {F_CHROM_SIZES} -r 999 -l 1000 -s > {path_proximal_promoters}\n",
    "\n",
    "        # 4.1 Get 100 bp-windows\n",
    "        path_windows = tmpdir + '/windows.bed'\n",
    "        !bedtools makewindows -b {path_proximal_promoters} -w 100 -i srcwinnum > {path_windows}\n",
    "\n",
    "        # 4.2. Reverse window numbers for + strand\n",
    "        windows_df = pd.read_table(path_windows, header=None, names=['chr', 'start', 'end', 'window_name'])\n",
    "        windows_df[['transcript_id', 'window_num']] = windows_df['window_name'].str.split('_', expand=True)\n",
    "        windows_df = windows_df.merge(annotated_transcripts[['transcript_id', 'strand']].drop_duplicates(), how='left', on='transcript_id')\n",
    "\n",
    "        new_windows = []\n",
    "        for idx, group_df in windows_df.groupby('transcript_id', sort=False):\n",
    "            if group_df.iloc[0,-1] == \"+\":\n",
    "                group_df['window_num'] = [*range(-10, 0), *range(1, 11)]\n",
    "            else:\n",
    "                group_df['window_num'] = [*range(10, 0, -1), *range(-1, -11, -1)]\n",
    "            new_windows.extend(group_df['window_num'].values)\n",
    "        windows_df['window_num'] = new_windows\n",
    "        windows_df['new_name'] = windows_df['transcript_id'] + '_' + windows_df['window_num'].astype(str)\n",
    "        windows_df[['chr', 'start', 'end', 'new_name']].to_csv(path_windows, sep='\\t', header=False, index=False)\n",
    "        \n",
    "        overlaps_df = pd.DataFrame()\n",
    "        for flipon in flipon_data:\n",
    "            path_to_intersection_count = tmpdir + f'/{flipon.name}_intersections.tsv'\n",
    "            !bedtools intersect -a {path_windows} -b {flipon.path} -c > {path_to_intersection_count}\n",
    "\n",
    "            print(flipon.name)\n",
    "            !bedtools intersect -a {path_proximal_promoters} -b {flipon.path} -wo | wc -l\n",
    "            \n",
    "            df = pd.read_table(path_to_intersection_count, header=None, names=['Gene Chr', 'Gene Start', 'Gene End', 'Window Name', 'Count'])\n",
    "            df['Flipon'] = flipon.name\n",
    "\n",
    "            overlaps_df = pd.concat([overlaps_df, df], ignore_index=True)\n",
    "        \n",
    "        overlaps_df[['Gene Name', 'Window Num']] = overlaps_df['Window Name'].str.split('_', expand=True)\n",
    "\n",
    "    return overlaps_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "g4\n",
      "9809\n",
      "z-dna\n",
      "14009\n",
      "sidd\n",
      "1057\n",
      "h-dna\n",
      "2446\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Chr</th>\n",
       "      <th>Gene Start</th>\n",
       "      <th>Gene End</th>\n",
       "      <th>Window Name</th>\n",
       "      <th>Count</th>\n",
       "      <th>Flipon</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Window Num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3072253</td>\n",
       "      <td>3072353</td>\n",
       "      <td>ENSMUST00000193812.1_-10</td>\n",
       "      <td>0</td>\n",
       "      <td>g4</td>\n",
       "      <td>ENSMUST00000193812.1</td>\n",
       "      <td>-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3072353</td>\n",
       "      <td>3072453</td>\n",
       "      <td>ENSMUST00000193812.1_-9</td>\n",
       "      <td>0</td>\n",
       "      <td>g4</td>\n",
       "      <td>ENSMUST00000193812.1</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1909518</th>\n",
       "      <td>chrY</td>\n",
       "      <td>4228380</td>\n",
       "      <td>4228480</td>\n",
       "      <td>ENSMUST00000189112.1_9</td>\n",
       "      <td>0</td>\n",
       "      <td>h-dna</td>\n",
       "      <td>ENSMUST00000189112.1</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1909519</th>\n",
       "      <td>chrY</td>\n",
       "      <td>4228480</td>\n",
       "      <td>4228580</td>\n",
       "      <td>ENSMUST00000189112.1_10</td>\n",
       "      <td>0</td>\n",
       "      <td>h-dna</td>\n",
       "      <td>ENSMUST00000189112.1</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1909520 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene Chr  Gene Start  Gene End               Window Name  Count  \\\n",
       "0           chr1     3072253   3072353  ENSMUST00000193812.1_-10      0   \n",
       "1           chr1     3072353   3072453   ENSMUST00000193812.1_-9      0   \n",
       "...          ...         ...       ...                       ...    ...   \n",
       "1909518     chrY     4228380   4228480    ENSMUST00000189112.1_9      0   \n",
       "1909519     chrY     4228480   4228580   ENSMUST00000189112.1_10      0   \n",
       "\n",
       "        Flipon             Gene Name Window Num  \n",
       "0           g4  ENSMUST00000193812.1        -10  \n",
       "1           g4  ENSMUST00000193812.1         -9  \n",
       "...        ...                   ...        ...  \n",
       "1909518  h-dna  ENSMUST00000189112.1          9  \n",
       "1909519  h-dna  ENSMUST00000189112.1         10  \n",
       "\n",
       "[1909520 rows x 8 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlaps_df_all = get_overlaps_df(flipon_to_data)\n",
    "overlaps_df_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "g4\n",
      "8469\n",
      "z-dna\n",
      "11766\n",
      "sidd\n",
      "992\n",
      "h-dna\n",
      "2021\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Chr</th>\n",
       "      <th>Gene Start</th>\n",
       "      <th>Gene End</th>\n",
       "      <th>Window Name</th>\n",
       "      <th>Count</th>\n",
       "      <th>Flipon</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Window Num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3072253</td>\n",
       "      <td>3072353</td>\n",
       "      <td>ENSMUST00000193812.1_-10</td>\n",
       "      <td>0</td>\n",
       "      <td>g4</td>\n",
       "      <td>ENSMUST00000193812.1</td>\n",
       "      <td>-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3072353</td>\n",
       "      <td>3072453</td>\n",
       "      <td>ENSMUST00000193812.1_-9</td>\n",
       "      <td>0</td>\n",
       "      <td>g4</td>\n",
       "      <td>ENSMUST00000193812.1</td>\n",
       "      <td>-9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1578478</th>\n",
       "      <td>chrY</td>\n",
       "      <td>4228380</td>\n",
       "      <td>4228480</td>\n",
       "      <td>ENSMUST00000189112.1_9</td>\n",
       "      <td>0</td>\n",
       "      <td>h-dna</td>\n",
       "      <td>ENSMUST00000189112.1</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1578479</th>\n",
       "      <td>chrY</td>\n",
       "      <td>4228480</td>\n",
       "      <td>4228580</td>\n",
       "      <td>ENSMUST00000189112.1_10</td>\n",
       "      <td>0</td>\n",
       "      <td>h-dna</td>\n",
       "      <td>ENSMUST00000189112.1</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1578480 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene Chr  Gene Start  Gene End               Window Name  Count  \\\n",
       "0           chr1     3072253   3072353  ENSMUST00000193812.1_-10      0   \n",
       "1           chr1     3072353   3072453   ENSMUST00000193812.1_-9      0   \n",
       "...          ...         ...       ...                       ...    ...   \n",
       "1578478     chrY     4228380   4228480    ENSMUST00000189112.1_9      0   \n",
       "1578479     chrY     4228480   4228580   ENSMUST00000189112.1_10      0   \n",
       "\n",
       "        Flipon             Gene Name Window Num  \n",
       "0           g4  ENSMUST00000193812.1        -10  \n",
       "1           g4  ENSMUST00000193812.1         -9  \n",
       "...        ...                   ...        ...  \n",
       "1578478  h-dna  ENSMUST00000189112.1          9  \n",
       "1578479  h-dna  ENSMUST00000189112.1         10  \n",
       "\n",
       "[1578480 rows x 8 columns]"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlaps_df_mirna = get_overlaps_df(flipon_to_data[flipon_to_data.filter(like='c. ').notna().sum(axis=1)>0])\n",
    "overlaps_df_mirna"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Flipon distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "40f041beef0243648474e8c17c302532",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/23869 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>-10</th>\n",
       "      <th>-9</th>\n",
       "      <th>-8</th>\n",
       "      <th>-7</th>\n",
       "      <th>-6</th>\n",
       "      <th>-5</th>\n",
       "      <th>-4</th>\n",
       "      <th>-3</th>\n",
       "      <th>-2</th>\n",
       "      <th>-1</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>g4</th>\n",
       "      <td>163</td>\n",
       "      <td>161</td>\n",
       "      <td>227</td>\n",
       "      <td>255</td>\n",
       "      <td>314</td>\n",
       "      <td>386</td>\n",
       "      <td>549</td>\n",
       "      <td>703</td>\n",
       "      <td>1176</td>\n",
       "      <td>1782</td>\n",
       "      <td>942</td>\n",
       "      <td>1012</td>\n",
       "      <td>1002</td>\n",
       "      <td>875</td>\n",
       "      <td>769</td>\n",
       "      <td>621</td>\n",
       "      <td>483</td>\n",
       "      <td>410</td>\n",
       "      <td>340</td>\n",
       "      <td>267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>h-dna</th>\n",
       "      <td>84</td>\n",
       "      <td>82</td>\n",
       "      <td>92</td>\n",
       "      <td>118</td>\n",
       "      <td>127</td>\n",
       "      <td>134</td>\n",
       "      <td>165</td>\n",
       "      <td>223</td>\n",
       "      <td>252</td>\n",
       "      <td>376</td>\n",
       "      <td>191</td>\n",
       "      <td>168</td>\n",
       "      <td>152</td>\n",
       "      <td>172</td>\n",
       "      <td>145</td>\n",
       "      <td>170</td>\n",
       "      <td>125</td>\n",
       "      <td>111</td>\n",
       "      <td>109</td>\n",
       "      <td>94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sidd</th>\n",
       "      <td>83</td>\n",
       "      <td>98</td>\n",
       "      <td>119</td>\n",
       "      <td>128</td>\n",
       "      <td>126</td>\n",
       "      <td>138</td>\n",
       "      <td>165</td>\n",
       "      <td>198</td>\n",
       "      <td>189</td>\n",
       "      <td>126</td>\n",
       "      <td>89</td>\n",
       "      <td>76</td>\n",
       "      <td>85</td>\n",
       "      <td>114</td>\n",
       "      <td>116</td>\n",
       "      <td>107</td>\n",
       "      <td>101</td>\n",
       "      <td>113</td>\n",
       "      <td>105</td>\n",
       "      <td>94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>z-dna</th>\n",
       "      <td>183</td>\n",
       "      <td>244</td>\n",
       "      <td>244</td>\n",
       "      <td>293</td>\n",
       "      <td>445</td>\n",
       "      <td>578</td>\n",
       "      <td>677</td>\n",
       "      <td>896</td>\n",
       "      <td>1274</td>\n",
       "      <td>1935</td>\n",
       "      <td>1739</td>\n",
       "      <td>1458</td>\n",
       "      <td>1419</td>\n",
       "      <td>1200</td>\n",
       "      <td>1062</td>\n",
       "      <td>844</td>\n",
       "      <td>679</td>\n",
       "      <td>551</td>\n",
       "      <td>452</td>\n",
       "      <td>373</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       -10  -9   -8   -7   -6   -5   -4   -3    -2    -1     1     2     3   \\\n",
       "g4     163  161  227  255  314  386  549  703  1176  1782   942  1012  1002   \n",
       "h-dna   84   82   92  118  127  134  165  223   252   376   191   168   152   \n",
       "sidd    83   98  119  128  126  138  165  198   189   126    89    76    85   \n",
       "z-dna  183  244  244  293  445  578  677  896  1274  1935  1739  1458  1419   \n",
       "\n",
       "         4     5    6    7    8    9    10  \n",
       "g4      875   769  621  483  410  340  267  \n",
       "h-dna   172   145  170  125  111  109   94  \n",
       "sidd    114   116  107  101  113  105   94  \n",
       "z-dna  1200  1062  844  679  551  452  373  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "total_sum = np.zeros((4, 20))\n",
    "\n",
    "for idx, group_df in tqdm(overlaps_df_all.groupby(\"Gene Name\")):\n",
    "    pivot_df = group_df.astype({\"Window Num\": int}).pivot_table(\n",
    "        values=[\"Count\"], index=[\"Flipon\"], columns=[\"Window Num\"], sort=True\n",
    "    )\n",
    "    total_sum += pivot_df.values\n",
    "\n",
    "total_sum_df = pd.DataFrame(\n",
    "    total_sum, index=[\"g4\", \"h-dna\", \"sidd\", \"z-dna\"], columns=[*range(-10, 0), *range(1, 11)], dtype=int\n",
    ")\n",
    "\n",
    "total_sum_df.to_csv(D_TABLES / \"total_window_flipons_all.tsv\", sep='\\t', index=True)\n",
    "display(total_sum_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a43cd2d0a6ba4fdf986c72efd1c9919b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/19731 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>-10</th>\n",
       "      <th>-9</th>\n",
       "      <th>-8</th>\n",
       "      <th>-7</th>\n",
       "      <th>-6</th>\n",
       "      <th>-5</th>\n",
       "      <th>-4</th>\n",
       "      <th>-3</th>\n",
       "      <th>-2</th>\n",
       "      <th>-1</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>g4</th>\n",
       "      <td>136</td>\n",
       "      <td>141</td>\n",
       "      <td>188</td>\n",
       "      <td>218</td>\n",
       "      <td>270</td>\n",
       "      <td>329</td>\n",
       "      <td>477</td>\n",
       "      <td>619</td>\n",
       "      <td>1023</td>\n",
       "      <td>1544</td>\n",
       "      <td>818</td>\n",
       "      <td>883</td>\n",
       "      <td>865</td>\n",
       "      <td>760</td>\n",
       "      <td>635</td>\n",
       "      <td>522</td>\n",
       "      <td>419</td>\n",
       "      <td>356</td>\n",
       "      <td>295</td>\n",
       "      <td>244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>h-dna</th>\n",
       "      <td>72</td>\n",
       "      <td>72</td>\n",
       "      <td>74</td>\n",
       "      <td>88</td>\n",
       "      <td>93</td>\n",
       "      <td>110</td>\n",
       "      <td>142</td>\n",
       "      <td>173</td>\n",
       "      <td>212</td>\n",
       "      <td>321</td>\n",
       "      <td>168</td>\n",
       "      <td>136</td>\n",
       "      <td>120</td>\n",
       "      <td>133</td>\n",
       "      <td>112</td>\n",
       "      <td>135</td>\n",
       "      <td>105</td>\n",
       "      <td>98</td>\n",
       "      <td>95</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sidd</th>\n",
       "      <td>80</td>\n",
       "      <td>92</td>\n",
       "      <td>110</td>\n",
       "      <td>118</td>\n",
       "      <td>118</td>\n",
       "      <td>133</td>\n",
       "      <td>159</td>\n",
       "      <td>186</td>\n",
       "      <td>177</td>\n",
       "      <td>119</td>\n",
       "      <td>85</td>\n",
       "      <td>70</td>\n",
       "      <td>77</td>\n",
       "      <td>103</td>\n",
       "      <td>105</td>\n",
       "      <td>100</td>\n",
       "      <td>97</td>\n",
       "      <td>108</td>\n",
       "      <td>98</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>z-dna</th>\n",
       "      <td>155</td>\n",
       "      <td>206</td>\n",
       "      <td>210</td>\n",
       "      <td>241</td>\n",
       "      <td>390</td>\n",
       "      <td>496</td>\n",
       "      <td>570</td>\n",
       "      <td>762</td>\n",
       "      <td>1074</td>\n",
       "      <td>1600</td>\n",
       "      <td>1432</td>\n",
       "      <td>1188</td>\n",
       "      <td>1208</td>\n",
       "      <td>1005</td>\n",
       "      <td>883</td>\n",
       "      <td>725</td>\n",
       "      <td>567</td>\n",
       "      <td>475</td>\n",
       "      <td>395</td>\n",
       "      <td>318</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       -10  -9   -8   -7   -6   -5   -4   -3    -2    -1     1     2     3   \\\n",
       "g4     136  141  188  218  270  329  477  619  1023  1544   818   883   865   \n",
       "h-dna   72   72   74   88   93  110  142  173   212   321   168   136   120   \n",
       "sidd    80   92  110  118  118  133  159  186   177   119    85    70    77   \n",
       "z-dna  155  206  210  241  390  496  570  762  1074  1600  1432  1188  1208   \n",
       "\n",
       "         4    5    6    7    8    9    10  \n",
       "g4      760  635  522  419  356  295  244  \n",
       "h-dna   133  112  135  105   98   95   80  \n",
       "sidd    103  105  100   97  108   98   88  \n",
       "z-dna  1005  883  725  567  475  395  318  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "total_sum = np.zeros((4, 20))\n",
    "\n",
    "for idx, group_df in tqdm(overlaps_df_mirna.groupby(\"Gene Name\")):\n",
    "    pivot_df = group_df.astype({\"Window Num\": int}).pivot_table(\n",
    "        values=[\"Count\"], index=[\"Flipon\"], columns=[\"Window Num\"], sort=True\n",
    "    )\n",
    "    total_sum += pivot_df.values\n",
    "\n",
    "total_sum_df = pd.DataFrame(\n",
    "    total_sum, index=[\"g4\", \"h-dna\", \"sidd\", \"z-dna\"], columns=[*range(-10, 0), *range(1, 11)], dtype=int\n",
    ")\n",
    "\n",
    "total_sum_df.to_csv(D_TABLES / \"total_window_flipons_mirna.tsv\", sep='\\t', index=True)\n",
    "display(total_sum_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pairwise sums"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8101848822aa43b19c8425822b49ab92",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/23869 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>-10</th>\n",
       "      <th>-9</th>\n",
       "      <th>-8</th>\n",
       "      <th>-7</th>\n",
       "      <th>-6</th>\n",
       "      <th>-5</th>\n",
       "      <th>-4</th>\n",
       "      <th>-3</th>\n",
       "      <th>-2</th>\n",
       "      <th>-1</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>106.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>141.0</td>\n",
       "      <td>178.0</td>\n",
       "      <td>189.0</td>\n",
       "      <td>258.0</td>\n",
       "      <td>343.0</td>\n",
       "      <td>443.0</td>\n",
       "      <td>734.0</td>\n",
       "      <td>993.0</td>\n",
       "      <td>566.0</td>\n",
       "      <td>646.0</td>\n",
       "      <td>664.0</td>\n",
       "      <td>592.0</td>\n",
       "      <td>539.0</td>\n",
       "      <td>425.0</td>\n",
       "      <td>332.0</td>\n",
       "      <td>291.0</td>\n",
       "      <td>253.0</td>\n",
       "      <td>190.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G &amp; H</th>\n",
       "      <td>14.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G &amp; H &amp; S</th>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2.0</td>\n",
       "      <td></td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G &amp; H &amp; S &amp; Z</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G &amp; H &amp; Z</th>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G &amp; S</th>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G &amp; S &amp; Z</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td>2.0</td>\n",
       "      <td></td>\n",
       "      <td>2.0</td>\n",
       "      <td></td>\n",
       "      <td>2.0</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G &amp; Z</th>\n",
       "      <td>30.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>119.0</td>\n",
       "      <td>137.0</td>\n",
       "      <td>248.0</td>\n",
       "      <td>434.0</td>\n",
       "      <td>253.0</td>\n",
       "      <td>257.0</td>\n",
       "      <td>223.0</td>\n",
       "      <td>188.0</td>\n",
       "      <td>141.0</td>\n",
       "      <td>125.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>38.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>41.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>111.0</td>\n",
       "      <td>108.0</td>\n",
       "      <td>126.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>87.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>59.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H &amp; S</th>\n",
       "      <td>6.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H &amp; S &amp; Z</th>\n",
       "      <td></td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td></td>\n",
       "      <td>2.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H &amp; Z</th>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>72.0</td>\n",
       "      <td>82.0</td>\n",
       "      <td>98.0</td>\n",
       "      <td>109.0</td>\n",
       "      <td>110.0</td>\n",
       "      <td>118.0</td>\n",
       "      <td>144.0</td>\n",
       "      <td>159.0</td>\n",
       "      <td>163.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>87.0</td>\n",
       "      <td>82.0</td>\n",
       "      <td>78.0</td>\n",
       "      <td>76.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>83.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S &amp; Z</th>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Z</th>\n",
       "      <td>123.0</td>\n",
       "      <td>172.0</td>\n",
       "      <td>165.0</td>\n",
       "      <td>208.0</td>\n",
       "      <td>295.0</td>\n",
       "      <td>388.0</td>\n",
       "      <td>432.0</td>\n",
       "      <td>592.0</td>\n",
       "      <td>776.0</td>\n",
       "      <td>1057.0</td>\n",
       "      <td>1147.0</td>\n",
       "      <td>996.0</td>\n",
       "      <td>977.0</td>\n",
       "      <td>836.0</td>\n",
       "      <td>763.0</td>\n",
       "      <td>604.0</td>\n",
       "      <td>503.0</td>\n",
       "      <td>388.0</td>\n",
       "      <td>337.0</td>\n",
       "      <td>297.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 -10    -9     -8     -7     -6     -5     -4     -3     -2   \\\n",
       "G              106.0   99.0  141.0  178.0  189.0  258.0  343.0  443.0  734.0   \n",
       "G & H           14.0   14.0   13.0   19.0   15.0   18.0   34.0   51.0   83.0   \n",
       "G & H & S        1.0                                                     4.0   \n",
       "G & H & S & Z                                                            1.0   \n",
       "G & H & Z        3.0    2.0    2.0    4.0    4.0    4.0    4.0    9.0   12.0   \n",
       "G & S            4.0    4.0    4.0    5.0    6.0    6.0    3.0    5.0    3.0   \n",
       "G & S & Z                                                         2.0          \n",
       "G & Z           30.0   31.0   45.0   39.0   75.0   69.0  119.0  137.0  248.0   \n",
       "H               41.0   45.0   45.0   68.0   77.0   85.0   91.0  111.0  108.0   \n",
       "H & S            6.0    8.0   13.0    8.0    8.0    7.0   14.0   24.0   10.0   \n",
       "H & S & Z               2.0    1.0    1.0           1.0    1.0           1.0   \n",
       "H & Z            4.0    3.0    7.0    4.0    9.0    9.0    9.0   13.0   16.0   \n",
       "S               72.0   82.0   98.0  109.0  110.0  118.0  144.0  159.0  163.0   \n",
       "S & Z                   1.0    2.0    5.0    1.0    6.0    3.0    8.0    7.0   \n",
       "Z              123.0  172.0  165.0  208.0  295.0  388.0  432.0  592.0  776.0   \n",
       "\n",
       "                  -1       1      2      3      4      5      6      7   \\\n",
       "G               993.0   566.0  646.0  664.0  592.0  539.0  425.0  332.0   \n",
       "G & H           129.0    37.0   32.0   30.0   31.0   27.0   31.0   19.0   \n",
       "G & H & S                        1.0                  2.0           4.0   \n",
       "G & H & S & Z     5.0            1.0                                      \n",
       "G & H & Z        33.0    16.0    5.0    5.0    4.0    2.0    4.0    2.0   \n",
       "G & S             8.0     3.0    5.0    3.0    4.0    8.0    6.0   10.0   \n",
       "G & S & Z                 1.0           2.0           2.0           2.0   \n",
       "G & Z           434.0   253.0  257.0  223.0  188.0  141.0  125.0   85.0   \n",
       "H               126.0    94.0   91.0   87.0   90.0   83.0   90.0   74.0   \n",
       "H & S            20.0    15.0   18.0   13.0   14.0   15.0   16.0    7.0   \n",
       "H & S & Z         1.0                                 1.0    2.0          \n",
       "H & Z            25.0    17.0    9.0    9.0   13.0    5.0    6.0    6.0   \n",
       "S                88.0    61.0   48.0   63.0   87.0   82.0   78.0   76.0   \n",
       "S & Z             4.0     9.0    3.0    4.0    6.0    6.0    5.0    1.0   \n",
       "Z              1057.0  1147.0  996.0  977.0  836.0  763.0  604.0  503.0   \n",
       "\n",
       "                  8      9      10  \n",
       "G              291.0  253.0  190.0  \n",
       "G & H           15.0   11.0   19.0  \n",
       "G & H & S        1.0           1.0  \n",
       "G & H & S & Z                       \n",
       "G & H & Z        3.0    2.0         \n",
       "G & S            1.0    1.0    3.0  \n",
       "G & S & Z               1.0    1.0  \n",
       "G & Z           73.0   47.0   38.0  \n",
       "H               68.0   65.0   59.0  \n",
       "H & S           11.0    5.0    4.0  \n",
       "H & S & Z        2.0                \n",
       "H & Z            3.0    9.0    1.0  \n",
       "S               96.0   96.0   83.0  \n",
       "S & Z            2.0    2.0    2.0  \n",
       "Z              388.0  337.0  297.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "flipon_type_num = {\"g4\": 1, \"h-dna\": 2, \"sidd\": 3, \"z-dna\": 4}\n",
    "flipon_num_type = {\"1\": \"G\", \"2\": \"H\", \"3\": \"S\", \"4\": \"Z\"}\n",
    "# flipon_num_type = {str(val):key for key,val in flipon_type_num.items()}\n",
    "counter = Counter()\n",
    "\n",
    "for idx, group_df in tqdm(overlaps_df_all.groupby(\"Gene Name\")):\n",
    "    group_df[\"Count\"] = group_df[\"Count\"].apply(lambda x: int(x != 0))\n",
    "    group_df[\"Count_type\"] = group_df[[\"Flipon\", \"Count\"]].apply(\n",
    "        lambda x: flipon_type_num[x[\"Flipon\"]] if x[\"Count\"] != 0 else 0, axis=1\n",
    "    )\n",
    "\n",
    "    pivot_df = (\n",
    "        group_df.astype({\"Window Num\": int})\n",
    "        .pivot_table(\n",
    "            values=[\"Count_type\"], index=[\"Flipon\"], columns=[\"Window Num\"], sort=True\n",
    "        )\n",
    "        .astype(str)\n",
    "    )\n",
    "\n",
    "    pivot_sum = pivot_df.sum().str.replace(\"0\", \"\")\n",
    "    for col, val in zip(pivot_sum.index, pivot_sum.values):\n",
    "        if val == \"\":\n",
    "            continue\n",
    "        counter[(col[1], val)] += 1\n",
    "\n",
    "result = pd.DataFrame()\n",
    "for idx, val in counter.items():\n",
    "    result.loc[idx[1], idx[0]] = val\n",
    "\n",
    "result = result.fillna('')\n",
    "result = result.loc[:,sorted(result.columns)]\n",
    "# result.columns = [f\"[{c*100}, {(c-1)*100}]\" for c in result.columns]\n",
    "result = result.sort_index()\n",
    "result.index = [' & '.join([flipon_num_type[y] for y in x]) for x in result.index]\n",
    "\n",
    "with pd.option_context('display.max_rows', None):\n",
    "    display(result)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5e5ba3146a904162971bf60166e40d96",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/19731 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>-10</th>\n",
       "      <th>-9</th>\n",
       "      <th>-8</th>\n",
       "      <th>-7</th>\n",
       "      <th>-6</th>\n",
       "      <th>-5</th>\n",
       "      <th>-4</th>\n",
       "      <th>-3</th>\n",
       "      <th>-2</th>\n",
       "      <th>-1</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>87.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>118.0</td>\n",
       "      <td>157.0</td>\n",
       "      <td>162.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>302.0</td>\n",
       "      <td>387.0</td>\n",
       "      <td>633.0</td>\n",
       "      <td>847.0</td>\n",
       "      <td>476.0</td>\n",
       "      <td>558.0</td>\n",
       "      <td>562.0</td>\n",
       "      <td>505.0</td>\n",
       "      <td>446.0</td>\n",
       "      <td>353.0</td>\n",
       "      <td>292.0</td>\n",
       "      <td>253.0</td>\n",
       "      <td>218.0</td>\n",
       "      <td>173.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G &amp; H</th>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>112.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G &amp; H &amp; S</th>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2.0</td>\n",
       "      <td></td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G &amp; H &amp; S &amp; Z</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G &amp; H &amp; Z</th>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G &amp; S</th>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G &amp; S &amp; Z</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td>2.0</td>\n",
       "      <td></td>\n",
       "      <td>2.0</td>\n",
       "      <td></td>\n",
       "      <td>2.0</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G &amp; Z</th>\n",
       "      <td>25.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>117.0</td>\n",
       "      <td>222.0</td>\n",
       "      <td>375.0</td>\n",
       "      <td>224.0</td>\n",
       "      <td>226.0</td>\n",
       "      <td>202.0</td>\n",
       "      <td>165.0</td>\n",
       "      <td>115.0</td>\n",
       "      <td>110.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>34.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>33.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>87.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>64.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>48.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H &amp; S</th>\n",
       "      <td>6.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H &amp; S &amp; Z</th>\n",
       "      <td></td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H &amp; Z</th>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>69.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>92.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>104.0</td>\n",
       "      <td>113.0</td>\n",
       "      <td>140.0</td>\n",
       "      <td>151.0</td>\n",
       "      <td>153.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>58.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>92.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>78.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S &amp; Z</th>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td></td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Z</th>\n",
       "      <td>103.0</td>\n",
       "      <td>143.0</td>\n",
       "      <td>146.0</td>\n",
       "      <td>172.0</td>\n",
       "      <td>258.0</td>\n",
       "      <td>329.0</td>\n",
       "      <td>371.0</td>\n",
       "      <td>501.0</td>\n",
       "      <td>645.0</td>\n",
       "      <td>837.0</td>\n",
       "      <td>918.0</td>\n",
       "      <td>790.0</td>\n",
       "      <td>818.0</td>\n",
       "      <td>691.0</td>\n",
       "      <td>630.0</td>\n",
       "      <td>512.0</td>\n",
       "      <td>422.0</td>\n",
       "      <td>337.0</td>\n",
       "      <td>290.0</td>\n",
       "      <td>251.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 -10    -9     -8     -7     -6     -5     -4     -3     -2   \\\n",
       "G               87.0   88.0  118.0  157.0  162.0  221.0  302.0  387.0  633.0   \n",
       "G & H           12.0   12.0   10.0   14.0   10.0   16.0   32.0   45.0   70.0   \n",
       "G & H & S        1.0                                                     4.0   \n",
       "G & H & S & Z                                                            1.0   \n",
       "G & H & Z        2.0    1.0    2.0    2.0    3.0    2.0    4.0    9.0   10.0   \n",
       "G & S            4.0    4.0    4.0    5.0    6.0    6.0    2.0    5.0    3.0   \n",
       "G & S & Z                                                         1.0          \n",
       "G & Z           25.0   26.0   35.0   30.0   66.0   58.0   96.0  117.0  222.0   \n",
       "H               33.0   40.0   35.0   48.0   56.0   68.0   73.0   77.0   87.0   \n",
       "H & S            6.0    7.0   10.0    7.0    7.0    7.0   14.0   21.0    9.0   \n",
       "H & S & Z               2.0    1.0    1.0           1.0    1.0           1.0   \n",
       "H & Z            4.0    2.0    6.0    4.0    8.0    8.0    7.0   11.0   14.0   \n",
       "S               69.0   77.0   92.0  100.0  104.0  113.0  140.0  151.0  153.0   \n",
       "S & Z                   1.0    2.0    5.0           6.0    2.0    8.0    6.0   \n",
       "Z              103.0  143.0  146.0  172.0  258.0  329.0  371.0  501.0  645.0   \n",
       "\n",
       "                 -1      1      2      3      4      5      6      7      8   \\\n",
       "G              847.0  476.0  558.0  562.0  505.0  446.0  353.0  292.0  253.0   \n",
       "G & H          112.0   35.0   27.0   26.0   27.0   19.0   27.0   17.0   14.0   \n",
       "G & H & S                      1.0                  2.0           3.0    1.0   \n",
       "G & H & S & Z    4.0           1.0                                             \n",
       "G & H & Z       32.0   15.0    3.0    3.0    4.0    2.0    3.0    1.0    3.0   \n",
       "G & S            8.0    3.0    4.0    3.0    4.0    8.0    6.0   10.0    1.0   \n",
       "G & S & Z               1.0           2.0           2.0           2.0          \n",
       "G & Z          375.0  224.0  226.0  202.0  165.0  115.0  110.0   68.0   61.0   \n",
       "H               96.0   77.0   70.0   67.0   67.0   64.0   67.0   61.0   58.0   \n",
       "H & S           19.0   15.0   15.0   11.0   11.0   10.0   14.0    7.0   11.0   \n",
       "H & S & Z        1.0                                1.0    2.0           1.0   \n",
       "H & Z           23.0   16.0    9.0    5.0    9.0    5.0    5.0    6.0    2.0   \n",
       "S               83.0   58.0   46.0   58.0   80.0   77.0   74.0   73.0   92.0   \n",
       "S & Z            4.0    8.0    3.0    3.0    5.0    5.0    4.0    1.0    2.0   \n",
       "Z              837.0  918.0  790.0  818.0  691.0  630.0  512.0  422.0  337.0   \n",
       "\n",
       "                  9      10  \n",
       "G              218.0  173.0  \n",
       "G & H            8.0   17.0  \n",
       "G & H & S               1.0  \n",
       "G & H & S & Z                \n",
       "G & H & Z        2.0         \n",
       "G & S            1.0    3.0  \n",
       "G & S & Z        1.0    1.0  \n",
       "G & Z           43.0   34.0  \n",
       "H               55.0   48.0  \n",
       "H & S            5.0    4.0  \n",
       "H & S & Z                    \n",
       "H & Z            8.0         \n",
       "S               89.0   78.0  \n",
       "S & Z            2.0    1.0  \n",
       "Z              290.0  251.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "flipon_type_num = {\"g4\": 1, \"h-dna\": 2, \"sidd\": 3, \"z-dna\": 4}\n",
    "flipon_num_type = {\"1\": \"G\", \"2\": \"H\", \"3\": \"S\", \"4\": \"Z\"}\n",
    "# flipon_num_type = {str(val):key for key,val in flipon_type_num.items()}\n",
    "counter = Counter()\n",
    "\n",
    "for idx, group_df in tqdm(overlaps_df_mirna.groupby(\"Gene Name\")):\n",
    "    group_df[\"Count\"] = group_df[\"Count\"].apply(lambda x: int(x != 0))\n",
    "    group_df[\"Count_type\"] = group_df[[\"Flipon\", \"Count\"]].apply(\n",
    "        lambda x: flipon_type_num[x[\"Flipon\"]] if x[\"Count\"] != 0 else 0, axis=1\n",
    "    )\n",
    "\n",
    "    pivot_df = (\n",
    "        group_df.astype({\"Window Num\": int})\n",
    "        .pivot_table(\n",
    "            values=[\"Count_type\"], index=[\"Flipon\"], columns=[\"Window Num\"], sort=True\n",
    "        )\n",
    "        .astype(str)\n",
    "    )\n",
    "\n",
    "    pivot_sum = pivot_df.sum().str.replace(\"0\", \"\")\n",
    "    for col, val in zip(pivot_sum.index, pivot_sum.values):\n",
    "        if val == \"\":\n",
    "            continue\n",
    "        counter[(col[1], val)] += 1\n",
    "\n",
    "result = pd.DataFrame()\n",
    "for idx, val in counter.items():\n",
    "    result.loc[idx[1], idx[0]] = val\n",
    "\n",
    "result = result.fillna('')\n",
    "result = result.loc[:,sorted(result.columns)]\n",
    "# result.columns = [f\"[{c*100}, {(c-1)*100}]\" for c in result.columns]\n",
    "result = result.sort_index()\n",
    "result.index = [' & '.join([flipon_num_type[y] for y in x]) for x in result.index]\n",
    "\n",
    "\n",
    "with pd.option_context('display.max_rows', None):\n",
    "    display(result)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Summary for all flipons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ae28e10d22494e949d1ec5bb11b78611",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/23869 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>-10</th>\n",
       "      <th>-9</th>\n",
       "      <th>-8</th>\n",
       "      <th>-7</th>\n",
       "      <th>-6</th>\n",
       "      <th>-5</th>\n",
       "      <th>-4</th>\n",
       "      <th>-3</th>\n",
       "      <th>-2</th>\n",
       "      <th>-1</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>23465.0</td>\n",
       "      <td>23406.0</td>\n",
       "      <td>23333.0</td>\n",
       "      <td>23221.0</td>\n",
       "      <td>23080.0</td>\n",
       "      <td>22900.0</td>\n",
       "      <td>22672.0</td>\n",
       "      <td>22315.0</td>\n",
       "      <td>21703.0</td>\n",
       "      <td>20946.0</td>\n",
       "      <td>21650.0</td>\n",
       "      <td>21757.0</td>\n",
       "      <td>21789.0</td>\n",
       "      <td>22004.0</td>\n",
       "      <td>22193.0</td>\n",
       "      <td>22477.0</td>\n",
       "      <td>22748.0</td>\n",
       "      <td>22915.0</td>\n",
       "      <td>23040.0</td>\n",
       "      <td>23171.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>342.0</td>\n",
       "      <td>398.0</td>\n",
       "      <td>449.0</td>\n",
       "      <td>563.0</td>\n",
       "      <td>671.0</td>\n",
       "      <td>849.0</td>\n",
       "      <td>1010.0</td>\n",
       "      <td>1305.0</td>\n",
       "      <td>1781.0</td>\n",
       "      <td>2264.0</td>\n",
       "      <td>1868.0</td>\n",
       "      <td>1781.0</td>\n",
       "      <td>1791.0</td>\n",
       "      <td>1605.0</td>\n",
       "      <td>1467.0</td>\n",
       "      <td>1197.0</td>\n",
       "      <td>985.0</td>\n",
       "      <td>843.0</td>\n",
       "      <td>751.0</td>\n",
       "      <td>629.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>58.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>114.0</td>\n",
       "      <td>115.0</td>\n",
       "      <td>182.0</td>\n",
       "      <td>238.0</td>\n",
       "      <td>367.0</td>\n",
       "      <td>620.0</td>\n",
       "      <td>334.0</td>\n",
       "      <td>324.0</td>\n",
       "      <td>282.0</td>\n",
       "      <td>256.0</td>\n",
       "      <td>202.0</td>\n",
       "      <td>189.0</td>\n",
       "      <td>128.0</td>\n",
       "      <td>105.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>67.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       -10      -9       -8       -7       -6       -5       -4       -3   \\\n",
       "0  23465.0  23406.0  23333.0  23221.0  23080.0  22900.0  22672.0  22315.0   \n",
       "1    342.0    398.0    449.0    563.0    671.0    849.0   1010.0   1305.0   \n",
       "2     58.0     61.0     84.0     80.0    114.0    115.0    182.0    238.0   \n",
       "3      4.0      4.0      3.0      5.0      4.0      5.0      5.0     11.0   \n",
       "4                                                                           \n",
       "\n",
       "       -2       -1        1        2        3        4        5        6   \\\n",
       "0  21703.0  20946.0  21650.0  21757.0  21789.0  22004.0  22193.0  22477.0   \n",
       "1   1781.0   2264.0   1868.0   1781.0   1791.0   1605.0   1467.0   1197.0   \n",
       "2    367.0    620.0    334.0    324.0    282.0    256.0    202.0    189.0   \n",
       "3     17.0     34.0     17.0      6.0      7.0      4.0      7.0      6.0   \n",
       "4      1.0      5.0               1.0                                       \n",
       "\n",
       "        7        8        9        10  \n",
       "0  22748.0  22915.0  23040.0  23171.0  \n",
       "1    985.0    843.0    751.0    629.0  \n",
       "2    128.0    105.0     75.0     67.0  \n",
       "3      8.0      6.0      3.0      2.0  \n",
       "4                                      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "counter = Counter()\n",
    "\n",
    "for idx, group_df in tqdm(overlaps_df_all.groupby(\"Gene Name\")):\n",
    "    group_df[\"Count\"] = group_df[\"Count\"].apply(lambda x: int(x != 0))\n",
    "\n",
    "    pivot_df = group_df.pivot_table(\n",
    "        values=[\"Count\"], index=[\"Flipon\"], columns=[\"Window Num\"]\n",
    "    )\n",
    "\n",
    "    pivot_sum = pivot_df.sum()\n",
    "    for col, val in zip(pivot_sum.index, pivot_sum.values):\n",
    "        counter[(int(col[1]), val)] += 1\n",
    "\n",
    "result = pd.DataFrame()\n",
    "for idx, val in counter.items():\n",
    "    result.loc[idx[1], idx[0]] = val\n",
    "result = result.loc[:,sorted(result.columns)].sort_index().fillna('')\n",
    "\n",
    "with pd.option_context('display.max_rows', None):\n",
    "    display(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4dcf37ce52eb41c4bb0b82a8465615d5",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/19731 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>-10</th>\n",
       "      <th>-9</th>\n",
       "      <th>-8</th>\n",
       "      <th>-7</th>\n",
       "      <th>-6</th>\n",
       "      <th>-5</th>\n",
       "      <th>-4</th>\n",
       "      <th>-3</th>\n",
       "      <th>-2</th>\n",
       "      <th>-1</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19385.0</td>\n",
       "      <td>19328.0</td>\n",
       "      <td>19270.0</td>\n",
       "      <td>19186.0</td>\n",
       "      <td>19051.0</td>\n",
       "      <td>18896.0</td>\n",
       "      <td>18687.0</td>\n",
       "      <td>18398.0</td>\n",
       "      <td>17873.0</td>\n",
       "      <td>17290.0</td>\n",
       "      <td>17885.0</td>\n",
       "      <td>17978.0</td>\n",
       "      <td>17971.0</td>\n",
       "      <td>18163.0</td>\n",
       "      <td>18345.0</td>\n",
       "      <td>18554.0</td>\n",
       "      <td>18768.0</td>\n",
       "      <td>18895.0</td>\n",
       "      <td>19009.0</td>\n",
       "      <td>19120.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>292.0</td>\n",
       "      <td>348.0</td>\n",
       "      <td>391.0</td>\n",
       "      <td>477.0</td>\n",
       "      <td>580.0</td>\n",
       "      <td>731.0</td>\n",
       "      <td>886.0</td>\n",
       "      <td>1116.0</td>\n",
       "      <td>1518.0</td>\n",
       "      <td>1863.0</td>\n",
       "      <td>1529.0</td>\n",
       "      <td>1464.0</td>\n",
       "      <td>1505.0</td>\n",
       "      <td>1343.0</td>\n",
       "      <td>1217.0</td>\n",
       "      <td>1006.0</td>\n",
       "      <td>848.0</td>\n",
       "      <td>740.0</td>\n",
       "      <td>652.0</td>\n",
       "      <td>550.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>51.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>101.0</td>\n",
       "      <td>153.0</td>\n",
       "      <td>207.0</td>\n",
       "      <td>324.0</td>\n",
       "      <td>541.0</td>\n",
       "      <td>301.0</td>\n",
       "      <td>284.0</td>\n",
       "      <td>250.0</td>\n",
       "      <td>221.0</td>\n",
       "      <td>162.0</td>\n",
       "      <td>166.0</td>\n",
       "      <td>109.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>59.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       -10      -9       -8       -7       -6       -5       -4       -3   \\\n",
       "0  19385.0  19328.0  19270.0  19186.0  19051.0  18896.0  18687.0  18398.0   \n",
       "1    292.0    348.0    391.0    477.0    580.0    731.0    886.0   1116.0   \n",
       "2     51.0     52.0     67.0     65.0     97.0    101.0    153.0    207.0   \n",
       "3      3.0      3.0      3.0      3.0      3.0      3.0      5.0     10.0   \n",
       "4                                                                           \n",
       "\n",
       "       -2       -1        1        2        3        4        5        6   \\\n",
       "0  17873.0  17290.0  17885.0  17978.0  17971.0  18163.0  18345.0  18554.0   \n",
       "1   1518.0   1863.0   1529.0   1464.0   1505.0   1343.0   1217.0   1006.0   \n",
       "2    324.0    541.0    301.0    284.0    250.0    221.0    162.0    166.0   \n",
       "3     15.0     33.0     16.0      4.0      5.0      4.0      7.0      5.0   \n",
       "4      1.0      4.0               1.0                                       \n",
       "\n",
       "        7        8        9        10  \n",
       "0  18768.0  18895.0  19009.0  19120.0  \n",
       "1    848.0    740.0    652.0    550.0  \n",
       "2    109.0     91.0     67.0     59.0  \n",
       "3      6.0      5.0      3.0      2.0  \n",
       "4                                      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "counter = Counter()\n",
    "\n",
    "for idx, group_df in tqdm(overlaps_df_mirna.groupby(\"Gene Name\")):\n",
    "    group_df[\"Count\"] = group_df[\"Count\"].apply(lambda x: int(x != 0))\n",
    "\n",
    "    pivot_df = group_df.pivot_table(\n",
    "        values=[\"Count\"], index=[\"Flipon\"], columns=[\"Window Num\"]\n",
    "    )\n",
    "\n",
    "    pivot_sum = pivot_df.sum()\n",
    "    for col, val in zip(pivot_sum.index, pivot_sum.values):\n",
    "        counter[(int(col[1]), val)] += 1\n",
    "\n",
    "result = pd.DataFrame()\n",
    "for idx, val in counter.items():\n",
    "    result.loc[idx[1], idx[0]] = val\n",
    "result = result.loc[:,sorted(result.columns)].sort_index().fillna('')\n",
    "\n",
    "with pd.option_context('display.max_rows', None):\n",
    "    display(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.11.0 ('mirna')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0 | packaged by conda-forge | (main, Oct 25 2022, 06:24:40) [GCC 10.4.0]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "aa4ca5e43b0b74394dc05f972f57eb4922712b4f4450a8ee4f1bf77784cf2ccc"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
