{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import plotly.express as px\n",
    "\n",
    "from gtfparse import read_gtf\n",
    "\n",
    "from datapaths import *\n",
    "\n",
    "pd.set_option('display.max_rows', 4)\n",
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Flipon</th>\n",
       "      <th>Coordinates</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene Feature</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Strand</th>\n",
       "      <th>Gene Type</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Transcript ID</th>\n",
       "      <th>c. M miR (+)</th>\n",
       "      <th>c. M miR (-)</th>\n",
       "      <th>c. MJ miR (+)</th>\n",
       "      <th>c. MJ miR (-)</th>\n",
       "      <th>c. J miR (+)</th>\n",
       "      <th>c. J miR (-)</th>\n",
       "      <th>M miR (+)</th>\n",
       "      <th>M miR (-)</th>\n",
       "      <th>MJ miR (+)</th>\n",
       "      <th>MJ miR (-)</th>\n",
       "      <th>J miR (+)</th>\n",
       "      <th>J miR (-)</th>\n",
       "      <th>cCRE (+-200bp)</th>\n",
       "      <th>LINE/LTR (+-200bp)</th>\n",
       "      <th>Feature Group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3014794-3014871</td>\n",
       "      <td>-</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>+</td>\n",
       "      <td>TEC</td>\n",
       "      <td>ENSMUSG00000102693.1</td>\n",
       "      <td>ENSMUST00000193812.1</td>\n",
       "      <td>miR-328 (1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-486 (1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-6994 (1), miR-129b (1)</td>\n",
       "      <td>miR-92a-2 (1), miR-698/miR-7078 (1), miR-3572 ...</td>\n",
       "      <td>miR-486a/miR-486b (1)</td>\n",
       "      <td>miR-3076 (1), miR-6418 (1), miR-6987 (1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-7649 (1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LINE@L1Md_F2(-), LINE@L1VL1(+)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3099888-3099963</td>\n",
       "      <td>+</td>\n",
       "      <td>Promoter (2-3kb)</td>\n",
       "      <td>Gm26206</td>\n",
       "      <td>+</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>ENSMUSG00000064842.1</td>\n",
       "      <td>ENSMUST00000082908.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-328 (1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-365-1/miR-365-2 (1), miR-698/miR-7078 (1),...</td>\n",
       "      <td>miR-221 (1), miR-129b (1)</td>\n",
       "      <td>miR-6418 (1), miR-7067 (1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LINE@L1Md_F2(+), LTR@MTC(+)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77706</th>\n",
       "      <td>z-dna</td>\n",
       "      <td>chrY:4203100-4203146</td>\n",
       "      <td>.</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td>Gm28191</td>\n",
       "      <td>+</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>ENSMUSG00000100300.1</td>\n",
       "      <td>ENSMUST00000189112.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-493 (1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-466d/miR-466i (4), miR-466l (5)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-493 (1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LINE@L1_Mur1(-)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77707</th>\n",
       "      <td>z-dna</td>\n",
       "      <td>chrY:4207093-4207113</td>\n",
       "      <td>.</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td>Gm28191</td>\n",
       "      <td>+</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>ENSMUSG00000100300.1</td>\n",
       "      <td>ENSMUST00000189112.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-466a/miR-466e/miR-466p (1), miR-466d/miR-4...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LINE@L1_Mus3(+), LINE@L1Md_A(+), LTR@ORR1B1(+)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77708 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Flipon           Coordinates Strand       Gene Feature      Gene Name  \\\n",
       "0         g4  chr1:3014794-3014871      -  Distal Intergenic  4933401J01Rik   \n",
       "1         g4  chr1:3099888-3099963      +   Promoter (2-3kb)        Gm26206   \n",
       "...      ...                   ...    ...                ...            ...   \n",
       "77706  z-dna  chrY:4203100-4203146      .  Distal Intergenic        Gm28191   \n",
       "77707  z-dna  chrY:4207093-4207113      .  Distal Intergenic        Gm28191   \n",
       "\n",
       "      Gene Strand               Gene Type               Gene ID  \\\n",
       "0               +                     TEC  ENSMUSG00000102693.1   \n",
       "1               +                   snRNA  ENSMUSG00000064842.1   \n",
       "...           ...                     ...                   ...   \n",
       "77706           +  unprocessed_pseudogene  ENSMUSG00000100300.1   \n",
       "77707           +  unprocessed_pseudogene  ENSMUSG00000100300.1   \n",
       "\n",
       "              Transcript ID c. M miR (+) c. M miR (-) c. MJ miR (+)  \\\n",
       "0      ENSMUST00000193812.1  miR-328 (1)          NaN   miR-486 (1)   \n",
       "1      ENSMUST00000082908.1          NaN  miR-328 (1)           NaN   \n",
       "...                     ...          ...          ...           ...   \n",
       "77706  ENSMUST00000189112.1          NaN          NaN   miR-493 (1)   \n",
       "77707  ENSMUST00000189112.1          NaN          NaN           NaN   \n",
       "\n",
       "      c. MJ miR (-) c. J miR (+) c. J miR (-)  \\\n",
       "0               NaN          NaN          NaN   \n",
       "1               NaN          NaN          NaN   \n",
       "...             ...          ...          ...   \n",
       "77706           NaN          NaN          NaN   \n",
       "77707           NaN          NaN          NaN   \n",
       "\n",
       "                                               M miR (+)  \\\n",
       "0                             miR-6994 (1), miR-129b (1)   \n",
       "1      miR-365-1/miR-365-2 (1), miR-698/miR-7078 (1),...   \n",
       "...                                                  ...   \n",
       "77706                miR-466d/miR-466i (4), miR-466l (5)   \n",
       "77707  miR-466a/miR-466e/miR-466p (1), miR-466d/miR-4...   \n",
       "\n",
       "                                               M miR (-)  \\\n",
       "0      miR-92a-2 (1), miR-698/miR-7078 (1), miR-3572 ...   \n",
       "1                              miR-221 (1), miR-129b (1)   \n",
       "...                                                  ...   \n",
       "77706                                                NaN   \n",
       "77707                                                NaN   \n",
       "\n",
       "                       MJ miR (+)                                MJ miR (-)  \\\n",
       "0           miR-486a/miR-486b (1)  miR-3076 (1), miR-6418 (1), miR-6987 (1)   \n",
       "1      miR-6418 (1), miR-7067 (1)                                       NaN   \n",
       "...                           ...                                       ...   \n",
       "77706                 miR-493 (1)                                       NaN   \n",
       "77707                         NaN                                       NaN   \n",
       "\n",
       "      J miR (+)     J miR (-) cCRE (+-200bp)  \\\n",
       "0           NaN  miR-7649 (1)            NaN   \n",
       "1           NaN           NaN            NaN   \n",
       "...         ...           ...            ...   \n",
       "77706       NaN           NaN            NaN   \n",
       "77707       NaN           NaN            NaN   \n",
       "\n",
       "                                   LINE/LTR (+-200bp) Feature Group  \n",
       "0                      LINE@L1Md_F2(-), LINE@L1VL1(+)      LINE/LTR  \n",
       "1                         LINE@L1Md_F2(+), LTR@MTC(+)      LINE/LTR  \n",
       "...                                               ...           ...  \n",
       "77706                                 LINE@L1_Mur1(-)      LINE/LTR  \n",
       "77707  LINE@L1_Mus3(+), LINE@L1Md_A(+), LTR@ORR1B1(+)      LINE/LTR  \n",
       "\n",
       "[77708 rows x 24 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flipon_to_data = pd.read_table(F_FLIPON_TO_DATA)\n",
    "flipon_to_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/fpavlov/miniconda3/envs/mirna/lib/python3.11/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "/home/fpavlov/miniconda3/envs/mirna/lib/python3.11/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The warn_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'level', 'mgi_id', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name', 'transcript_support_level', 'tag', 'havana_transcript', 'exon_number', 'exon_id', 'protein_id', 'ccdsid', 'ont']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>mgi_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>ont</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000102693.1</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1918292</td>\n",
       "      <td>OTTMUSG00000049935.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000102693.1</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1918292</td>\n",
       "      <td>OTTMUSG00000049935.1</td>\n",
       "      <td>ENSMUST00000193812.1</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTMUST00000127109.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1302161</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000064372.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>mt-Tp</td>\n",
       "      <td>3</td>\n",
       "      <td>MGI:102478</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000082423.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>mt-Tp-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1302162</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000064372.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>mt-Tp</td>\n",
       "      <td>3</td>\n",
       "      <td>MGI:102478</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000082423.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>mt-Tp-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000521550.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1302163 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname   source     feature    start      end  score strand frame  \\\n",
       "0          chr1   HAVANA        gene  3073253  3074322    NaN      +     0   \n",
       "1          chr1   HAVANA  transcript  3073253  3074322    NaN      +     0   \n",
       "...         ...      ...         ...      ...      ...    ...    ...   ...   \n",
       "1302161    chrM  ENSEMBL  transcript    15356    15422    NaN      -     0   \n",
       "1302162    chrM  ENSEMBL        exon    15356    15422    NaN      -     0   \n",
       "\n",
       "                      gene_id gene_type      gene_name level       mgi_id  \\\n",
       "0        ENSMUSG00000102693.1       TEC  4933401J01Rik     2  MGI:1918292   \n",
       "1        ENSMUSG00000102693.1       TEC  4933401J01Rik     2  MGI:1918292   \n",
       "...                       ...       ...            ...   ...          ...   \n",
       "1302161  ENSMUSG00000064372.1   Mt_tRNA          mt-Tp     3   MGI:102478   \n",
       "1302162  ENSMUSG00000064372.1   Mt_tRNA          mt-Tp     3   MGI:102478   \n",
       "\n",
       "                  havana_gene         transcript_id transcript_type  \\\n",
       "0        OTTMUSG00000049935.1                                         \n",
       "1        OTTMUSG00000049935.1  ENSMUST00000193812.1             TEC   \n",
       "...                       ...                   ...             ...   \n",
       "1302161                        ENSMUST00000082423.1         Mt_tRNA   \n",
       "1302162                        ENSMUST00000082423.1         Mt_tRNA   \n",
       "\n",
       "           transcript_name transcript_support_level    tag  \\\n",
       "0                                                            \n",
       "1        4933401J01Rik-201                       NA  basic   \n",
       "...                    ...                      ...    ...   \n",
       "1302161          mt-Tp-201                       NA  basic   \n",
       "1302162          mt-Tp-201                       NA  basic   \n",
       "\n",
       "            havana_transcript exon_number               exon_id protein_id  \\\n",
       "0                                                                            \n",
       "1        OTTMUST00000127109.1                                                \n",
       "...                       ...         ...                   ...        ...   \n",
       "1302161                                                                      \n",
       "1302162                                 1  ENSMUSE00000521550.1              \n",
       "\n",
       "        ccdsid ont  \n",
       "0                   \n",
       "1                   \n",
       "...        ...  ..  \n",
       "1302161             \n",
       "1302162             \n",
       "\n",
       "[1302163 rows x 25 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf = read_gtf(F_GENCODE)\n",
    "gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['c. M miR (+)',\n",
       " 'c. M miR (-)',\n",
       " 'c. MJ miR (+)',\n",
       " 'c. MJ miR (-)',\n",
       " 'c. J miR (+)',\n",
       " 'c. J miR (-)']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c_mir_list = ['c. M miR (+)', 'c. M miR (-)', 'c. MJ miR (+)', 'c. MJ miR (-)', 'c. J miR (+)', 'c. J miR (-)']\n",
    "c_mir_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>23</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>g4</th>\n",
       "      <td>1881.0</td>\n",
       "      <td>855.0</td>\n",
       "      <td>413.0</td>\n",
       "      <td>184.0</td>\n",
       "      <td>77.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>h-dna</th>\n",
       "      <td>1003.0</td>\n",
       "      <td>200.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sidd</th>\n",
       "      <td>796.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>z-dna</th>\n",
       "      <td>1641.0</td>\n",
       "      <td>863.0</td>\n",
       "      <td>549.0</td>\n",
       "      <td>350.0</td>\n",
       "      <td>206.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           1      2      3      4      5      6     7     8     9    10   11  \\\n",
       "g4     1881.0  855.0  413.0  184.0   77.0   38.0  28.0  13.0   7.0  1.0        \n",
       "h-dna  1003.0  200.0   60.0   18.0    8.0    4.0   1.0   2.0             1.0   \n",
       "sidd    796.0   29.0    2.0                                                    \n",
       "z-dna  1641.0  863.0  549.0  350.0  206.0  120.0  71.0  25.0  17.0  6.0  8.0   \n",
       "\n",
       "        12   13   14   15   23  \n",
       "g4          1.0                 \n",
       "h-dna            1.0            \n",
       "sidd                            \n",
       "z-dna  6.0  5.0  2.0  1.0  1.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# All\n",
    "with pd.option_context(\"display.max_rows\", None):\n",
    "    display(\n",
    "        flipon_to_data[flipon_to_data[\"Gene Feature\"] == \"Promoter (<=1kb)\"]\n",
    "        .groupby(\"Gene ID\")[\"Flipon\"]\n",
    "        .value_counts()\n",
    "        .rename(\"Concurrent\")\n",
    "        .reset_index()[[\"Flipon\", \"Concurrent\"]]\n",
    "        .value_counts()\n",
    "        .rename(\"Total\")\n",
    "        .reset_index()\n",
    "        .pivot_table(values='Total', index='Flipon', columns='Concurrent').fillna('')\n",
    "        .rename_axis(None, axis=0).rename_axis(None,axis=1)\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>12</th>\n",
       "      <th>22</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Flipon</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>g4</th>\n",
       "      <td>1722.0</td>\n",
       "      <td>553.0</td>\n",
       "      <td>195.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>h-dna</th>\n",
       "      <td>485.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sidd</th>\n",
       "      <td>756.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>z-dna</th>\n",
       "      <td>1570.0</td>\n",
       "      <td>515.0</td>\n",
       "      <td>191.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            1      2      3     4     5     6    7    8    9    12   22\n",
       "Flipon                                                                 \n",
       "g4      1722.0  553.0  195.0  61.0  29.0  14.0  2.0  1.0               \n",
       "h-dna    485.0   59.0   12.0   1.0                                     \n",
       "sidd     756.0   24.0    2.0                                           \n",
       "z-dna   1570.0  515.0  191.0  70.0  25.0   5.0  4.0  1.0  1.0  1.0  1.0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# c. miRNA\n",
    "\n",
    "with pd.option_context(\"display.max_rows\", None):\n",
    "    display(\n",
    "        flipon_to_data[(flipon_to_data[\"Gene Feature\"] == \"Promoter (<=1kb)\") & (flipon_to_data[c_mir_list].notna().sum(axis=1)>0)]\n",
    "        .groupby(\"Gene ID\")[\"Flipon\"]\n",
    "        .value_counts()\n",
    "        .rename(\"Concurrent\")\n",
    "        .reset_index()[[\"Flipon\", \"Concurrent\"]]\n",
    "        .value_counts()\n",
    "        .rename(\"Total\")\n",
    "        .reset_index()\n",
    "        .pivot_table(values='Total', index='Flipon', columns='Concurrent').fillna('')\n",
    "        .rename_axis(None, axis=1)\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Conc</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Type</th>\n",
       "      <th>Flipon</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000105054.1</th>\n",
       "      <th>Gm43664</th>\n",
       "      <th>processed_pseudogene</th>\n",
       "      <th>z-dna</th>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000014592.21</th>\n",
       "      <th>Camta1</th>\n",
       "      <th>protein_coding</th>\n",
       "      <th>z-dna</th>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000049556.5</th>\n",
       "      <th>Lingo1</th>\n",
       "      <th>protein_coding</th>\n",
       "      <th>z-dna</th>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000057897.14</th>\n",
       "      <th>Camk2b</th>\n",
       "      <th>protein_coding</th>\n",
       "      <th>z-dna</th>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000108904.1</th>\n",
       "      <th>Gm45147</th>\n",
       "      <th>processed_pseudogene</th>\n",
       "      <th>h-dna</th>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000055013.16</th>\n",
       "      <th>Agap1</th>\n",
       "      <th>protein_coding</th>\n",
       "      <th>z-dna</th>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000061411.12</th>\n",
       "      <th>Nol4l</th>\n",
       "      <th>protein_coding</th>\n",
       "      <th>z-dna</th>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000031511.16</th>\n",
       "      <th>Arhgef7</th>\n",
       "      <th>protein_coding</th>\n",
       "      <th>z-dna</th>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000061411.12</th>\n",
       "      <th>Nol4l</th>\n",
       "      <th>protein_coding</th>\n",
       "      <th>g4</th>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000010476.14</th>\n",
       "      <th>Ebf3</th>\n",
       "      <th>protein_coding</th>\n",
       "      <th>z-dna</th>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000022382.15</th>\n",
       "      <th>Wnt7b</th>\n",
       "      <th>protein_coding</th>\n",
       "      <th>z-dna</th>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000058317.12</th>\n",
       "      <th>Ube2e2</th>\n",
       "      <th>protein_coding</th>\n",
       "      <th>z-dna</th>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000063632.6</th>\n",
       "      <th>Sox11</th>\n",
       "      <th>protein_coding</th>\n",
       "      <th>z-dna</th>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000042589.18</th>\n",
       "      <th>Cux2</th>\n",
       "      <th>protein_coding</th>\n",
       "      <th>z-dna</th>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000033581.17</th>\n",
       "      <th>Igf2bp2</th>\n",
       "      <th>protein_coding</th>\n",
       "      <th>z-dna</th>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000002799.6</th>\n",
       "      <th>Jag2</th>\n",
       "      <th>protein_coding</th>\n",
       "      <th>z-dna</th>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000054036.6</th>\n",
       "      <th>Olfr279</th>\n",
       "      <th>protein_coding</th>\n",
       "      <th>z-dna</th>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000001911.16</th>\n",
       "      <th>Nfix</th>\n",
       "      <th>protein_coding</th>\n",
       "      <th>z-dna</th>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000043456.17</th>\n",
       "      <th>Zfp536</th>\n",
       "      <th>protein_coding</th>\n",
       "      <th>z-dna</th>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000008575.17</th>\n",
       "      <th>Nfib</th>\n",
       "      <th>protein_coding</th>\n",
       "      <th>z-dna</th>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                             Conc\n",
       "Gene ID               Gene Name Gene Type            Flipon      \n",
       "ENSMUSG00000105054.1  Gm43664   processed_pseudogene z-dna     23\n",
       "ENSMUSG00000014592.21 Camta1    protein_coding       z-dna     15\n",
       "ENSMUSG00000049556.5  Lingo1    protein_coding       z-dna     14\n",
       "ENSMUSG00000057897.14 Camk2b    protein_coding       z-dna     14\n",
       "ENSMUSG00000108904.1  Gm45147   processed_pseudogene h-dna     14\n",
       "ENSMUSG00000055013.16 Agap1     protein_coding       z-dna     13\n",
       "ENSMUSG00000061411.12 Nol4l     protein_coding       z-dna     13\n",
       "ENSMUSG00000031511.16 Arhgef7   protein_coding       z-dna     13\n",
       "ENSMUSG00000061411.12 Nol4l     protein_coding       g4        13\n",
       "ENSMUSG00000010476.14 Ebf3      protein_coding       z-dna     13\n",
       "ENSMUSG00000022382.15 Wnt7b     protein_coding       z-dna     13\n",
       "ENSMUSG00000058317.12 Ube2e2    protein_coding       z-dna     12\n",
       "ENSMUSG00000063632.6  Sox11     protein_coding       z-dna     12\n",
       "ENSMUSG00000042589.18 Cux2      protein_coding       z-dna     12\n",
       "ENSMUSG00000033581.17 Igf2bp2   protein_coding       z-dna     12\n",
       "ENSMUSG00000002799.6  Jag2      protein_coding       z-dna     12\n",
       "ENSMUSG00000054036.6  Olfr279   protein_coding       z-dna     12\n",
       "ENSMUSG00000001911.16 Nfix      protein_coding       z-dna     11\n",
       "ENSMUSG00000043456.17 Zfp536    protein_coding       z-dna     11\n",
       "ENSMUSG00000008575.17 Nfib      protein_coding       z-dna     11"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "with pd.option_context('display.max_rows', 20):\n",
    "    display(\n",
    "        flipon_to_data[(flipon_to_data[\"Gene Feature\"] == \"Promoter (<=1kb)\")]\n",
    "        .groupby(\"Gene ID\")[['Gene Name', 'Gene Type', \"Flipon\"]]\n",
    "        .value_counts().rename('Conc').to_frame().sort_values('Conc', ascending=False).head(20)\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Transcript ID</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Type</th>\n",
       "      <th>Flipon</th>\n",
       "      <th>Flipons per promoter</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000105054.1</td>\n",
       "      <td>ENSMUST00000196038.1</td>\n",
       "      <td>Gm43664</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>z-dna</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000054036.6</td>\n",
       "      <td>ENSMUST00000050451.5</td>\n",
       "      <td>Olfr279</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>z-dna</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6849</th>\n",
       "      <td>ENSMUSG00000034037.14</td>\n",
       "      <td>ENSMUST00000089334.8</td>\n",
       "      <td>Fgd5</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>z-dna</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6850</th>\n",
       "      <td>ENSMUSG00000041488.16</td>\n",
       "      <td>ENSMUST00000211047.1</td>\n",
       "      <td>Stx3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>g4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6851 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Gene ID         Transcript ID Gene Name  \\\n",
       "0      ENSMUSG00000105054.1  ENSMUST00000196038.1   Gm43664   \n",
       "1      ENSMUSG00000054036.6  ENSMUST00000050451.5   Olfr279   \n",
       "...                     ...                   ...       ...   \n",
       "6849  ENSMUSG00000034037.14  ENSMUST00000089334.8      Fgd5   \n",
       "6850  ENSMUSG00000041488.16  ENSMUST00000211047.1      Stx3   \n",
       "\n",
       "                 Gene Type Flipon  Flipons per promoter  \n",
       "0     processed_pseudogene  z-dna                    22  \n",
       "1           protein_coding  z-dna                    12  \n",
       "...                    ...    ...                   ...  \n",
       "6849        protein_coding  z-dna                     1  \n",
       "6850        protein_coding     g4                     1  \n",
       "\n",
       "[6851 rows x 6 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table = (\n",
    "    flipon_to_data[\n",
    "        (flipon_to_data[\"Gene Feature\"] == \"Promoter (<=1kb)\")\n",
    "        & (flipon_to_data[c_mir_list].notna().sum(axis=1)>0)\n",
    "    ]\n",
    "    .groupby([\"Gene ID\", 'Transcript ID'])[[\"Gene Name\", \"Gene Type\", \"Flipon\"]]\n",
    "    .value_counts()\n",
    "    .rename(\"Flipons per promoter\")\n",
    "    .to_frame()\n",
    "    .sort_values(\"Flipons per promoter\", ascending=False).reset_index()\n",
    ")\n",
    "\n",
    "table.to_csv(D_TABLES / \"flipons_per_promoter_mir.tsv\", sep='\\t', index=False)\n",
    "\n",
    "table\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mirna",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0 | packaged by conda-forge | (main, Oct 25 2022, 06:24:40) [GCC 10.4.0]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "aa4ca5e43b0b74394dc05f972f57eb4922712b4f4450a8ee4f1bf77784cf2ccc"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
