{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tempfile\n",
    "import pandas as pd\n",
    "\n",
    "from gtfparse import read_gtf\n",
    "from Bio import SeqIO\n",
    "from datapaths import *\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "pd.set_option('display.max_rows', 4)\n",
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Flipon</th>\n",
       "      <th>Coordinates</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene Feature</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Gene Strand</th>\n",
       "      <th>Gene Type</th>\n",
       "      <th>Gene ID</th>\n",
       "      <th>Transcript ID</th>\n",
       "      <th>c. M miR (+)</th>\n",
       "      <th>c. M miR (-)</th>\n",
       "      <th>c. MJ miR (+)</th>\n",
       "      <th>c. MJ miR (-)</th>\n",
       "      <th>c. J miR (+)</th>\n",
       "      <th>c. J miR (-)</th>\n",
       "      <th>all M miR (+)</th>\n",
       "      <th>all M miR (-)</th>\n",
       "      <th>all MJ miR (+)</th>\n",
       "      <th>all MJ miR (-)</th>\n",
       "      <th>all J miR (+)</th>\n",
       "      <th>all J miR (-)</th>\n",
       "      <th>cCRE (+-200bp)</th>\n",
       "      <th>LINE/LTR (+-200bp)</th>\n",
       "      <th>Feature Group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3014794-3014871</td>\n",
       "      <td>-</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>+</td>\n",
       "      <td>TEC</td>\n",
       "      <td>ENSMUSG00000102693.1</td>\n",
       "      <td>ENSMUST00000193812.1</td>\n",
       "      <td>miR-328 (1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-486 (1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-6994 (1), miR-129b (1)</td>\n",
       "      <td>miR-92a-2 (1), miR-698/miR-7078 (1), miR-3572 ...</td>\n",
       "      <td>miR-486a/miR-486b (1)</td>\n",
       "      <td>miR-3076 (1), miR-6418 (1), miR-6987 (1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-7649 (1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LINE@L1Md_F2(-), LINE@L1VL1(+)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>g4</td>\n",
       "      <td>chr1:3099888-3099963</td>\n",
       "      <td>+</td>\n",
       "      <td>Promoter (2-3kb)</td>\n",
       "      <td>Gm26206</td>\n",
       "      <td>+</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>ENSMUSG00000064842.1</td>\n",
       "      <td>ENSMUST00000082908.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-328 (1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-365-1/miR-365-2 (1), miR-698/miR-7078 (1),...</td>\n",
       "      <td>miR-221 (1), miR-129b (1)</td>\n",
       "      <td>miR-6418 (1), miR-7067 (1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LINE@L1Md_F2(+), LTR@MTC(+)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77706</th>\n",
       "      <td>z-dna</td>\n",
       "      <td>chrY:4203100-4203146</td>\n",
       "      <td>.</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td>Gm28191</td>\n",
       "      <td>+</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>ENSMUSG00000100300.1</td>\n",
       "      <td>ENSMUST00000189112.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-493 (1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-466d/miR-466i (4), miR-466l (5)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-493 (1)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LINE@L1_Mur1(-)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77707</th>\n",
       "      <td>z-dna</td>\n",
       "      <td>chrY:4207093-4207113</td>\n",
       "      <td>.</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td>Gm28191</td>\n",
       "      <td>+</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>ENSMUSG00000100300.1</td>\n",
       "      <td>ENSMUST00000189112.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miR-466a/miR-466e/miR-466p (1), miR-466d/miR-4...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LINE@L1_Mus3(+), LINE@L1Md_A(+), LTR@ORR1B1(+)</td>\n",
       "      <td>LINE/LTR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77708 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Flipon           Coordinates Strand       Gene Feature      Gene Name  \\\n",
       "0         g4  chr1:3014794-3014871      -  Distal Intergenic  4933401J01Rik   \n",
       "1         g4  chr1:3099888-3099963      +   Promoter (2-3kb)        Gm26206   \n",
       "...      ...                   ...    ...                ...            ...   \n",
       "77706  z-dna  chrY:4203100-4203146      .  Distal Intergenic        Gm28191   \n",
       "77707  z-dna  chrY:4207093-4207113      .  Distal Intergenic        Gm28191   \n",
       "\n",
       "      Gene Strand               Gene Type               Gene ID  \\\n",
       "0               +                     TEC  ENSMUSG00000102693.1   \n",
       "1               +                   snRNA  ENSMUSG00000064842.1   \n",
       "...           ...                     ...                   ...   \n",
       "77706           +  unprocessed_pseudogene  ENSMUSG00000100300.1   \n",
       "77707           +  unprocessed_pseudogene  ENSMUSG00000100300.1   \n",
       "\n",
       "              Transcript ID c. M miR (+) c. M miR (-) c. MJ miR (+)  \\\n",
       "0      ENSMUST00000193812.1  miR-328 (1)          NaN   miR-486 (1)   \n",
       "1      ENSMUST00000082908.1          NaN  miR-328 (1)           NaN   \n",
       "...                     ...          ...          ...           ...   \n",
       "77706  ENSMUST00000189112.1          NaN          NaN   miR-493 (1)   \n",
       "77707  ENSMUST00000189112.1          NaN          NaN           NaN   \n",
       "\n",
       "      c. MJ miR (-) c. J miR (+) c. J miR (-)  \\\n",
       "0               NaN          NaN          NaN   \n",
       "1               NaN          NaN          NaN   \n",
       "...             ...          ...          ...   \n",
       "77706           NaN          NaN          NaN   \n",
       "77707           NaN          NaN          NaN   \n",
       "\n",
       "                                           all M miR (+)  \\\n",
       "0                             miR-6994 (1), miR-129b (1)   \n",
       "1      miR-365-1/miR-365-2 (1), miR-698/miR-7078 (1),...   \n",
       "...                                                  ...   \n",
       "77706                miR-466d/miR-466i (4), miR-466l (5)   \n",
       "77707  miR-466a/miR-466e/miR-466p (1), miR-466d/miR-4...   \n",
       "\n",
       "                                           all M miR (-)  \\\n",
       "0      miR-92a-2 (1), miR-698/miR-7078 (1), miR-3572 ...   \n",
       "1                              miR-221 (1), miR-129b (1)   \n",
       "...                                                  ...   \n",
       "77706                                                NaN   \n",
       "77707                                                NaN   \n",
       "\n",
       "                   all MJ miR (+)                            all MJ miR (-)  \\\n",
       "0           miR-486a/miR-486b (1)  miR-3076 (1), miR-6418 (1), miR-6987 (1)   \n",
       "1      miR-6418 (1), miR-7067 (1)                                       NaN   \n",
       "...                           ...                                       ...   \n",
       "77706                 miR-493 (1)                                       NaN   \n",
       "77707                         NaN                                       NaN   \n",
       "\n",
       "      all J miR (+) all J miR (-) cCRE (+-200bp)  \\\n",
       "0               NaN  miR-7649 (1)            NaN   \n",
       "1               NaN           NaN            NaN   \n",
       "...             ...           ...            ...   \n",
       "77706           NaN           NaN            NaN   \n",
       "77707           NaN           NaN            NaN   \n",
       "\n",
       "                                   LINE/LTR (+-200bp) Feature Group  \n",
       "0                      LINE@L1Md_F2(-), LINE@L1VL1(+)      LINE/LTR  \n",
       "1                         LINE@L1Md_F2(+), LTR@MTC(+)      LINE/LTR  \n",
       "...                                               ...           ...  \n",
       "77706                                 LINE@L1_Mur1(-)      LINE/LTR  \n",
       "77707  LINE@L1_Mus3(+), LINE@L1Md_A(+), LTR@ORR1B1(+)      LINE/LTR  \n",
       "\n",
       "[77708 rows x 24 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flipon_to_data = pd.read_table(F_FLIPON_TO_DATA)\n",
    "flipon_to_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/fpavlov/miniconda3/envs/mirna/lib/python3.11/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "/home/fpavlov/miniconda3/envs/mirna/lib/python3.11/site-packages/gtfparse/read_gtf.py:82: FutureWarning: The warn_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
      "\n",
      "\n",
      "  chunk_iterator = pd.read_csv(\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'level', 'mgi_id', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_name', 'transcript_support_level', 'tag', 'havana_transcript', 'exon_number', 'exon_id', 'protein_id', 'ccdsid', 'ont']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>mgi_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>ont</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000102693.1</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1918292</td>\n",
       "      <td>OTTMUSG00000049935.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3073253</td>\n",
       "      <td>3074322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000102693.1</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>2</td>\n",
       "      <td>MGI:1918292</td>\n",
       "      <td>OTTMUSG00000049935.1</td>\n",
       "      <td>ENSMUST00000193812.1</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTMUST00000127109.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1302161</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000064372.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>mt-Tp</td>\n",
       "      <td>3</td>\n",
       "      <td>MGI:102478</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000082423.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>mt-Tp-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1302162</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15356</td>\n",
       "      <td>15422</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000064372.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>mt-Tp</td>\n",
       "      <td>3</td>\n",
       "      <td>MGI:102478</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000082423.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>mt-Tp-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00000521550.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1302163 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname   source     feature    start      end  score strand frame  \\\n",
       "0          chr1   HAVANA        gene  3073253  3074322    NaN      +     0   \n",
       "1          chr1   HAVANA  transcript  3073253  3074322    NaN      +     0   \n",
       "...         ...      ...         ...      ...      ...    ...    ...   ...   \n",
       "1302161    chrM  ENSEMBL  transcript    15356    15422    NaN      -     0   \n",
       "1302162    chrM  ENSEMBL        exon    15356    15422    NaN      -     0   \n",
       "\n",
       "                      gene_id gene_type      gene_name level       mgi_id  \\\n",
       "0        ENSMUSG00000102693.1       TEC  4933401J01Rik     2  MGI:1918292   \n",
       "1        ENSMUSG00000102693.1       TEC  4933401J01Rik     2  MGI:1918292   \n",
       "...                       ...       ...            ...   ...          ...   \n",
       "1302161  ENSMUSG00000064372.1   Mt_tRNA          mt-Tp     3   MGI:102478   \n",
       "1302162  ENSMUSG00000064372.1   Mt_tRNA          mt-Tp     3   MGI:102478   \n",
       "\n",
       "                  havana_gene         transcript_id transcript_type  \\\n",
       "0        OTTMUSG00000049935.1                                         \n",
       "1        OTTMUSG00000049935.1  ENSMUST00000193812.1             TEC   \n",
       "...                       ...                   ...             ...   \n",
       "1302161                        ENSMUST00000082423.1         Mt_tRNA   \n",
       "1302162                        ENSMUST00000082423.1         Mt_tRNA   \n",
       "\n",
       "           transcript_name transcript_support_level    tag  \\\n",
       "0                                                            \n",
       "1        4933401J01Rik-201                       NA  basic   \n",
       "...                    ...                      ...    ...   \n",
       "1302161          mt-Tp-201                       NA  basic   \n",
       "1302162          mt-Tp-201                       NA  basic   \n",
       "\n",
       "            havana_transcript exon_number               exon_id protein_id  \\\n",
       "0                                                                            \n",
       "1        OTTMUST00000127109.1                                                \n",
       "...                       ...         ...                   ...        ...   \n",
       "1302161                                                                      \n",
       "1302162                                 1  ENSMUSE00000521550.1              \n",
       "\n",
       "        ccdsid ont  \n",
       "0                   \n",
       "1                   \n",
       "...        ...  ..  \n",
       "1302161             \n",
       "1302162             \n",
       "\n",
       "[1302163 rows x 25 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf = read_gtf(F_GENCODE)\n",
    "gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_heptomers(path_fa: str, name: str) -> pd.DataFrame:\n",
    "    with tempfile.TemporaryDirectory() as tmpdir:\n",
    "        jellyfish_bc = tmpdir + f\"/jellyfish.bc\"\n",
    "        jellyfish_jf = tmpdir + f\"/jellyfish.jf\"\n",
    "        jellyfish_fa = tmpdir + f\"/jellyfish.fa\"\n",
    "        !jellyfish bc -m 7 -s 1G -t 60 -o {jellyfish_bc} {path_fa}\n",
    "        !jellyfish count -m 7 -s 100M -t 60 --bc {jellyfish_bc} -o {jellyfish_jf} {path_fa}\n",
    "        !jellyfish dump {jellyfish_jf} > {jellyfish_fa}\n",
    "\n",
    "        fasta = list(SeqIO.parse(jellyfish_fa, \"fasta\"))\n",
    "        heptomers, counts = [], []\n",
    "        for rec in fasta:\n",
    "            heptomers.append(str(rec.seq))\n",
    "            counts.append(rec.description)\n",
    "\n",
    "        df = (\n",
    "            pd.DataFrame(data={\"heptomer\": heptomers, \"heptomer_count\": counts})\n",
    "            .astype({\"heptomer_count\": int})\n",
    "            .sort_values(\"heptomer_count\", ascending=False)\n",
    "            .reset_index(drop=True)\n",
    "        )\n",
    "        df['col_lvl_0'] = name.split('_')[0]\n",
    "        df['col_lvl_1'] = name.split('_')[1]\n",
    "        df['col_lvl_2'] = name.split('_')[2]\n",
    "        df['total_regions'] = int(name.split('_')[3])\n",
    "        df['total_heptomers'] = len(heptomers)\n",
    "        return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_heptomers_for_flipons(flipon_type: str, mir: str):\n",
    "    with tempfile.TemporaryDirectory() as tmpdir:\n",
    "        # 1. Get flipon coordinates\n",
    "        path_flipons = tmpdir + '/flipons.bed'\n",
    "        path_flipons_fa = tmpdir + '/flipons.fa'\n",
    "\n",
    "        flipons = flipon_to_data.copy()\n",
    "        if mir == 'c. miR':\n",
    "            flipons = flipons[(flipons.filter(like='c. ').notna().sum(axis=1) > 0)]\n",
    "        if flipon_type!='Any flipon':\n",
    "            flipons = flipons[flipons['Flipon'] == (flipon_type)]\n",
    "        flipons = flipons['Coordinates'].str.split(r':|-', regex=True, expand=True)\n",
    "        \n",
    "        flipons.to_csv(path_flipons, sep='\\t', header=False, index=False)\n",
    "        !bedtools getfasta -bed {path_flipons} -fi {F_GENOME} > {path_flipons_fa}\n",
    "\n",
    "        # 2. Get heptomer count\n",
    "        heptomer_promoter_df = get_heptomers(path_flipons_fa, f\"{flipon_type}_{mir}_Flipon_{flipons.shape[0]}\")\n",
    "\n",
    "        return heptomer_promoter_df\n",
    "\n",
    "\n",
    "def get_heptomers_for_pp(transcript_id_list: list, name: str):\n",
    "    with tempfile.TemporaryDirectory() as tmpdir:\n",
    "        # 1. Get transcript coordinates        \n",
    "        annotated_transcripts = gtf[(gtf['feature']=='transcript') & (gtf['transcript_id'].isin(transcript_id_list))].drop_duplicates('transcript_id')[['seqname', 'start', 'end', 'transcript_id', 'score', 'strand']]\n",
    "\n",
    "        # 2. Get TSS coordinates\n",
    "        annotated_transcripts['end'] = annotated_transcripts.apply(lambda row: row['start']+1 if row['strand']=='+' else row['end'], axis=1)\n",
    "        annotated_transcripts['start'] = annotated_transcripts.apply(lambda row: row['end']-1 if row['strand']=='-' else row['start'], axis=1)\n",
    "\n",
    "        path_tss = tmpdir + '/tss.bed'\n",
    "        annotated_transcripts.to_csv(path_tss, sep='\\t', header=False, index=False)\n",
    "\n",
    "        # 3. Get promoter coordinates\n",
    "        path_proximal_promoters = tmpdir + '/proximal_promoters.bed'\n",
    "        path_proximal_promoters_fasta = path_proximal_promoters.replace('.bed', '.fa')\n",
    "        !bedtools slop -i {path_tss} -g {F_CHROM_SIZES} -r 999 -l 1000 -s > {path_proximal_promoters}\n",
    "        !bedtools getfasta -bed {path_proximal_promoters} -fi {F_GENOME} -s > {path_proximal_promoters_fasta}\n",
    "\n",
    "        # 4. Get heptomer count\n",
    "        heptomer_promoter_df = get_heptomers(path_proximal_promoters_fasta, f\"{name}_{annotated_transcripts.shape[0]}\")\n",
    "\n",
    "        return heptomer_promoter_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>heptomer</th>\n",
       "      <th>heptomer_count</th>\n",
       "      <th>col_lvl_0</th>\n",
       "      <th>col_lvl_1</th>\n",
       "      <th>col_lvl_2</th>\n",
       "      <th>total_regions</th>\n",
       "      <th>total_heptomers</th>\n",
       "      <th>heptomer_global_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TTTTTTT</td>\n",
       "      <td>233738</td>\n",
       "      <td>All</td>\n",
       "      <td>All regions</td>\n",
       "      <td>PP</td>\n",
       "      <td>81540</td>\n",
       "      <td>16384</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAAAA</td>\n",
       "      <td>216032</td>\n",
       "      <td>All</td>\n",
       "      <td>All regions</td>\n",
       "      <td>PP</td>\n",
       "      <td>81540</td>\n",
       "      <td>16384</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16382</th>\n",
       "      <td>CGATACG</td>\n",
       "      <td>317</td>\n",
       "      <td>All</td>\n",
       "      <td>All regions</td>\n",
       "      <td>PP</td>\n",
       "      <td>81540</td>\n",
       "      <td>16384</td>\n",
       "      <td>16383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16383</th>\n",
       "      <td>CGTATCG</td>\n",
       "      <td>251</td>\n",
       "      <td>All</td>\n",
       "      <td>All regions</td>\n",
       "      <td>PP</td>\n",
       "      <td>81540</td>\n",
       "      <td>16384</td>\n",
       "      <td>16384</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16384 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      heptomer  heptomer_count col_lvl_0    col_lvl_1 col_lvl_2  \\\n",
       "0      TTTTTTT          233738       All  All regions        PP   \n",
       "1      AAAAAAA          216032       All  All regions        PP   \n",
       "...        ...             ...       ...          ...       ...   \n",
       "16382  CGATACG             317       All  All regions        PP   \n",
       "16383  CGTATCG             251       All  All regions        PP   \n",
       "\n",
       "       total_regions  total_heptomers  heptomer_global_rank  \n",
       "0              81540            16384                     1  \n",
       "1              81540            16384                     2  \n",
       "...              ...              ...                   ...  \n",
       "16382          81540            16384                 16383  \n",
       "16383          81540            16384                 16384  \n",
       "\n",
       "[16384 rows x 8 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_id_list = gtf[gtf['feature']=='transcript']['transcript_id'].unique()\n",
    "name = \"All_All regions_PP\"\n",
    "df_all_all_pp = get_heptomers_for_pp(transcript_id_list, name)\n",
    "df_all_all_pp['heptomer_global_rank'] = df_all_all_pp.index + 1\n",
    "df_all_all_pp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c6325a4c0ba84e499e7f0d7ba3b9bb5c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/5 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>heptomer</th>\n",
       "      <th>heptomer_count</th>\n",
       "      <th>col_lvl_0</th>\n",
       "      <th>col_lvl_1</th>\n",
       "      <th>col_lvl_2</th>\n",
       "      <th>total_regions</th>\n",
       "      <th>total_heptomers</th>\n",
       "      <th>heptomer_global_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAAAA</td>\n",
       "      <td>59241</td>\n",
       "      <td>Any flipon</td>\n",
       "      <td>All regions</td>\n",
       "      <td>PP</td>\n",
       "      <td>23869</td>\n",
       "      <td>16384</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TGTGTGT</td>\n",
       "      <td>57821</td>\n",
       "      <td>Any flipon</td>\n",
       "      <td>All regions</td>\n",
       "      <td>PP</td>\n",
       "      <td>23869</td>\n",
       "      <td>16384</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258532</th>\n",
       "      <td>CGGACCT</td>\n",
       "      <td>2</td>\n",
       "      <td>z-dna</td>\n",
       "      <td>c. miR</td>\n",
       "      <td>Flipon</td>\n",
       "      <td>12348</td>\n",
       "      <td>3781</td>\n",
       "      <td>12664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258533</th>\n",
       "      <td>AAAAGCA</td>\n",
       "      <td>2</td>\n",
       "      <td>z-dna</td>\n",
       "      <td>c. miR</td>\n",
       "      <td>Flipon</td>\n",
       "      <td>12348</td>\n",
       "      <td>3781</td>\n",
       "      <td>629</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>258534 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       heptomer  heptomer_count   col_lvl_0    col_lvl_1 col_lvl_2  \\\n",
       "0       AAAAAAA           59241  Any flipon  All regions        PP   \n",
       "1       TGTGTGT           57821  Any flipon  All regions        PP   \n",
       "...         ...             ...         ...          ...       ...   \n",
       "258532  CGGACCT               2       z-dna       c. miR    Flipon   \n",
       "258533  AAAAGCA               2       z-dna       c. miR    Flipon   \n",
       "\n",
       "        total_regions  total_heptomers  heptomer_global_rank  \n",
       "0               23869            16384                     2  \n",
       "1               23869            16384                     3  \n",
       "...               ...              ...                   ...  \n",
       "258532          12348             3781                 12664  \n",
       "258533          12348             3781                   629  \n",
       "\n",
       "[258534 rows x 8 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_df = pd.DataFrame()\n",
    "\n",
    "for x in tqdm(['Any flipon', 'g4', 'h-dna', 'sidd', 'z-dna']):\n",
    "    for y in ['All regions', 'c. miR']:\n",
    "        transcript_id_list = flipon_to_data.copy()\n",
    "        if x != 'Any flipon':\n",
    "            transcript_id_list = transcript_id_list[transcript_id_list['Flipon']==x]\n",
    "        if y != 'All regions':\n",
    "            transcript_id_list = transcript_id_list[transcript_id_list.filter(like='c. ').notna().sum(axis=1)>0]\n",
    "        transcript_id_list = transcript_id_list['Transcript ID'].unique()\n",
    "        name = f\"{x}_{y}_PP\"\n",
    "        df_pp = get_heptomers_for_pp(transcript_id_list, name)\n",
    "        \n",
    "        df_flipon = get_heptomers_for_flipons(x, y)\n",
    "        \n",
    "        total_df = pd.concat([total_df, df_pp, df_flipon], ignore_index=True)\n",
    "\n",
    "total_df = total_df.merge(df_all_all_pp[['heptomer', 'heptomer_global_rank']], on='heptomer', how='left')\n",
    "total_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>col_lvl_0</th>\n",
       "      <th>All</th>\n",
       "      <th colspan=\"4\" halign=\"left\">Any flipon</th>\n",
       "      <th colspan=\"4\" halign=\"left\">g4</th>\n",
       "      <th colspan=\"4\" halign=\"left\">h-dna</th>\n",
       "      <th colspan=\"4\" halign=\"left\">sidd</th>\n",
       "      <th colspan=\"4\" halign=\"left\">z-dna</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>col_lvl_1</th>\n",
       "      <th>All regions</th>\n",
       "      <th colspan=\"2\" halign=\"left\">All regions</th>\n",
       "      <th colspan=\"2\" halign=\"left\">c. miR</th>\n",
       "      <th colspan=\"2\" halign=\"left\">All regions</th>\n",
       "      <th colspan=\"2\" halign=\"left\">c. miR</th>\n",
       "      <th colspan=\"2\" halign=\"left\">All regions</th>\n",
       "      <th colspan=\"2\" halign=\"left\">c. miR</th>\n",
       "      <th colspan=\"2\" halign=\"left\">All regions</th>\n",
       "      <th colspan=\"2\" halign=\"left\">c. miR</th>\n",
       "      <th colspan=\"2\" halign=\"left\">All regions</th>\n",
       "      <th colspan=\"2\" halign=\"left\">c. miR</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>col_lvl_2</th>\n",
       "      <th>PP</th>\n",
       "      <th>PP</th>\n",
       "      <th>Flipon</th>\n",
       "      <th>PP</th>\n",
       "      <th>Flipon</th>\n",
       "      <th>PP</th>\n",
       "      <th>Flipon</th>\n",
       "      <th>PP</th>\n",
       "      <th>Flipon</th>\n",
       "      <th>PP</th>\n",
       "      <th>Flipon</th>\n",
       "      <th>PP</th>\n",
       "      <th>Flipon</th>\n",
       "      <th>PP</th>\n",
       "      <th>Flipon</th>\n",
       "      <th>PP</th>\n",
       "      <th>Flipon</th>\n",
       "      <th>PP</th>\n",
       "      <th>Flipon</th>\n",
       "      <th>PP</th>\n",
       "      <th>Flipon</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>heptomer</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TTTTTTT</th>\n",
       "      <td>233738.0</td>\n",
       "      <td>57213.0</td>\n",
       "      <td>27382.0</td>\n",
       "      <td>46891.0</td>\n",
       "      <td>18623.0</td>\n",
       "      <td>27101.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19192.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>23828.0</td>\n",
       "      <td>10953.0</td>\n",
       "      <td>10703.0</td>\n",
       "      <td>2775.0</td>\n",
       "      <td>26063.0</td>\n",
       "      <td>16429.0</td>\n",
       "      <td>25464.0</td>\n",
       "      <td>15848.0</td>\n",
       "      <td>25294.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16799.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAAAAA</th>\n",
       "      <td>216032.0</td>\n",
       "      <td>59241.0</td>\n",
       "      <td>29575.0</td>\n",
       "      <td>49570.0</td>\n",
       "      <td>19955.0</td>\n",
       "      <td>27469.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>19608.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>25259.0</td>\n",
       "      <td>11810.0</td>\n",
       "      <td>11477.0</td>\n",
       "      <td>2759.0</td>\n",
       "      <td>28619.0</td>\n",
       "      <td>17761.0</td>\n",
       "      <td>28001.0</td>\n",
       "      <td>17193.0</td>\n",
       "      <td>25775.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17485.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGATACG</th>\n",
       "      <td>317.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>69.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>45.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>35.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>23.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>45.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGTATCG</th>\n",
       "      <td>251.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>35.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>38.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16384 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "col_lvl_0         All  Any flipon                                     g4  \\\n",
       "col_lvl_1 All regions All regions            c. miR          All regions   \n",
       "col_lvl_2          PP          PP   Flipon       PP   Flipon          PP   \n",
       "heptomer                                                                   \n",
       "TTTTTTT      233738.0     57213.0  27382.0  46891.0  18623.0     27101.0   \n",
       "AAAAAAA      216032.0     59241.0  29575.0  49570.0  19955.0     27469.0   \n",
       "...               ...         ...      ...      ...      ...         ...   \n",
       "CGATACG         317.0        75.0      NaN     69.0      NaN        45.0   \n",
       "CGTATCG         251.0        72.0      2.0     62.0      NaN        35.0   \n",
       "\n",
       "col_lvl_0                              h-dna                            \\\n",
       "col_lvl_1          c. miR        All regions            c. miR           \n",
       "col_lvl_2 Flipon       PP Flipon          PP   Flipon       PP  Flipon   \n",
       "heptomer                                                                 \n",
       "TTTTTTT      NaN  19192.0    NaN     23828.0  10953.0  10703.0  2775.0   \n",
       "AAAAAAA      4.0  19608.0    3.0     25259.0  11810.0  11477.0  2759.0   \n",
       "...          ...      ...    ...         ...      ...      ...     ...   \n",
       "CGATACG      NaN     35.0    NaN        19.0      NaN      9.0     NaN   \n",
       "CGTATCG      NaN     28.0    NaN        26.0      NaN     15.0     NaN   \n",
       "\n",
       "col_lvl_0        sidd                                  z-dna                  \\\n",
       "col_lvl_1 All regions            c. miR          All regions          c. miR   \n",
       "col_lvl_2          PP   Flipon       PP   Flipon          PP Flipon       PP   \n",
       "heptomer                                                                       \n",
       "TTTTTTT       26063.0  16429.0  25464.0  15848.0     25294.0    NaN  16799.0   \n",
       "AAAAAAA       28619.0  17761.0  28001.0  17193.0     25775.0    NaN  17485.0   \n",
       "...               ...      ...      ...      ...         ...    ...      ...   \n",
       "CGATACG          23.0      NaN     22.0      NaN        45.0    NaN     27.0   \n",
       "CGTATCG          30.0      NaN     30.0      NaN        38.0    NaN     26.0   \n",
       "\n",
       "col_lvl_0         \n",
       "col_lvl_1         \n",
       "col_lvl_2 Flipon  \n",
       "heptomer          \n",
       "TTTTTTT      NaN  \n",
       "AAAAAAA      NaN  \n",
       "...          ...  \n",
       "CGATACG      NaN  \n",
       "CGTATCG      NaN  \n",
       "\n",
       "[16384 rows x 21 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pivot = (\n",
    "    pd.concat([df_all_all_pp, total_df], ignore_index=True)\n",
    "    .pivot_table(index=\"heptomer\", columns=[\"col_lvl_0\", \"col_lvl_1\", \"col_lvl_2\"], values=[\"heptomer_count\"], sort=False)\n",
    "    .droplevel(0, axis=1)\n",
    ")\n",
    "\n",
    "pivot.to_csv(D_TABLES / 'heptomer_counts.tsv', sep='\\t')\n",
    "pivot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Data</th>\n",
       "      <th colspan=\"6\" halign=\"left\">0.promoters</th>\n",
       "      <th colspan=\"4\" halign=\"left\">g4</th>\n",
       "      <th colspan=\"4\" halign=\"left\">h-dna</th>\n",
       "      <th colspan=\"4\" halign=\"left\">sidd</th>\n",
       "      <th colspan=\"4\" halign=\"left\">z-dna</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Description</th>\n",
       "      <th colspan=\"2\" halign=\"left\">all mouse (55,401)</th>\n",
       "      <th colspan=\"2\" halign=\"left\">with flipons &amp; miR (15,515)</th>\n",
       "      <th colspan=\"2\" halign=\"left\">with flipons (21,356)</th>\n",
       "      <th colspan=\"2\" halign=\"left\">flipons &amp; miR (2,323)</th>\n",
       "      <th colspan=\"2\" halign=\"left\">flipons (5,721)</th>\n",
       "      <th colspan=\"2\" halign=\"left\">flipons &amp; miR (226)</th>\n",
       "      <th colspan=\"2\" halign=\"left\">flipons (1,537)</th>\n",
       "      <th colspan=\"2\" halign=\"left\">flipons &amp; miR (720)</th>\n",
       "      <th colspan=\"2\" halign=\"left\">flipons (779)</th>\n",
       "      <th colspan=\"2\" halign=\"left\">flipons &amp; miR (1,288)</th>\n",
       "      <th colspan=\"2\" halign=\"left\">flipons (8,245)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>number_type</th>\n",
       "      <th>heptomer_counts</th>\n",
       "      <th>heptomer_rank</th>\n",
       "      <th>heptomer_counts</th>\n",
       "      <th>heptomer_rank</th>\n",
       "      <th>heptomer_counts</th>\n",
       "      <th>heptomer_rank</th>\n",
       "      <th>heptomer_counts</th>\n",
       "      <th>heptomer_rank</th>\n",
       "      <th>heptomer_counts</th>\n",
       "      <th>heptomer_rank</th>\n",
       "      <th>heptomer_counts</th>\n",
       "      <th>heptomer_rank</th>\n",
       "      <th>heptomer_counts</th>\n",
       "      <th>heptomer_rank</th>\n",
       "      <th>heptomer_counts</th>\n",
       "      <th>heptomer_rank</th>\n",
       "      <th>heptomer_counts</th>\n",
       "      <th>heptomer_rank</th>\n",
       "      <th>heptomer_counts</th>\n",
       "      <th>heptomer_rank</th>\n",
       "      <th>heptomer_counts</th>\n",
       "      <th>heptomer_rank</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>heptomer</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAAAAAA</th>\n",
       "      <td>158560</td>\n",
       "      <td>1</td>\n",
       "      <td>41141</td>\n",
       "      <td>1</td>\n",
       "      <td>56065</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3915</td>\n",
       "      <td>1</td>\n",
       "      <td>5434</td>\n",
       "      <td>8</td>\n",
       "      <td>102</td>\n",
       "      <td>1326</td>\n",
       "      <td>5</td>\n",
       "      <td>1266</td>\n",
       "      <td>1</td>\n",
       "      <td>1362</td>\n",
       "      <td>1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTTTTT</th>\n",
       "      <td>153693</td>\n",
       "      <td>2</td>\n",
       "      <td>36769</td>\n",
       "      <td>3</td>\n",
       "      <td>50889</td>\n",
       "      <td>2</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>4</td>\n",
       "      <td>198</td>\n",
       "      <td>1276</td>\n",
       "      <td>6</td>\n",
       "      <td>1060</td>\n",
       "      <td>2</td>\n",
       "      <td>1184</td>\n",
       "      <td>2</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGATACG</th>\n",
       "      <td>180</td>\n",
       "      <td>16383</td>\n",
       "      <td>48</td>\n",
       "      <td>16382</td>\n",
       "      <td>66</td>\n",
       "      <td>16384</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGTATCG</th>\n",
       "      <td>169</td>\n",
       "      <td>16384</td>\n",
       "      <td>51</td>\n",
       "      <td>16381</td>\n",
       "      <td>68</td>\n",
       "      <td>16383</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16384 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Data               0.promoters                                            \\\n",
       "Description all mouse (55,401)               with flipons & miR (15,515)   \n",
       "number_type    heptomer_counts heptomer_rank             heptomer_counts   \n",
       "heptomer                                                                   \n",
       "AAAAAAA                 158560             1                       41141   \n",
       "TTTTTTT                 153693             2                       36769   \n",
       "...                        ...           ...                         ...   \n",
       "CGATACG                    180         16383                          48   \n",
       "CGTATCG                    169         16384                          51   \n",
       "\n",
       "Data                                                           \\\n",
       "Description               with flipons (21,356)                 \n",
       "number_type heptomer_rank       heptomer_counts heptomer_rank   \n",
       "heptomer                                                        \n",
       "AAAAAAA                 1                 56065             1   \n",
       "TTTTTTT                 3                 50889             2   \n",
       "...                   ...                   ...           ...   \n",
       "CGATACG             16382                    66         16384   \n",
       "CGTATCG             16381                    68         16383   \n",
       "\n",
       "Data                           g4                                              \\\n",
       "Description flipons & miR (2,323)               flipons (5,721)                 \n",
       "number_type       heptomer_counts heptomer_rank heptomer_counts heptomer_rank   \n",
       "heptomer                                                                        \n",
       "AAAAAAA                         1          3915               1          5434   \n",
       "TTTTTTT                      <NA>          <NA>            <NA>          <NA>   \n",
       "...                           ...           ...             ...           ...   \n",
       "CGATACG                      <NA>          <NA>            <NA>          <NA>   \n",
       "CGTATCG                      <NA>          <NA>            <NA>          <NA>   \n",
       "\n",
       "Data                      h-dna                                              \\\n",
       "Description flipons & miR (226)               flipons (1,537)                 \n",
       "number_type     heptomer_counts heptomer_rank heptomer_counts heptomer_rank   \n",
       "heptomer                                                                      \n",
       "AAAAAAA                       8           102            1326             5   \n",
       "TTTTTTT                       4           198            1276             6   \n",
       "...                         ...           ...             ...           ...   \n",
       "CGATACG                    <NA>          <NA>            <NA>          <NA>   \n",
       "CGTATCG                    <NA>          <NA>            <NA>          <NA>   \n",
       "\n",
       "Data                       sidd                                              \\\n",
       "Description flipons & miR (720)                 flipons (779)                 \n",
       "number_type     heptomer_counts heptomer_rank heptomer_counts heptomer_rank   \n",
       "heptomer                                                                      \n",
       "AAAAAAA                    1266             1            1362             1   \n",
       "TTTTTTT                    1060             2            1184             2   \n",
       "...                         ...           ...             ...           ...   \n",
       "CGATACG                    <NA>          <NA>            <NA>          <NA>   \n",
       "CGTATCG                    <NA>          <NA>            <NA>          <NA>   \n",
       "\n",
       "Data                        z-dna                                              \n",
       "Description flipons & miR (1,288)               flipons (8,245)                \n",
       "number_type       heptomer_counts heptomer_rank heptomer_counts heptomer_rank  \n",
       "heptomer                                                                       \n",
       "AAAAAAA                      <NA>          <NA>            <NA>          <NA>  \n",
       "TTTTTTT                      <NA>          <NA>            <NA>          <NA>  \n",
       "...                           ...           ...             ...           ...  \n",
       "CGATACG                      <NA>          <NA>            <NA>          <NA>  \n",
       "CGTATCG                      <NA>          <NA>            <NA>          <NA>  \n",
       "\n",
       "[16384 rows x 22 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_pivot = (\n",
    "    heptomer_total_df.pivot_table(\n",
    "        index=\"heptomer\", columns=[\"Data\", \"Description\", \"number_type\"], values=[\"value\"],\n",
    "    )\n",
    "    .astype(pd.Int32Dtype())\n",
    "    .droplevel(0, axis=1)\n",
    "    .sort_values(('0.promoters', 'all mouse (55,401)', 'heptomer_rank'))\n",
    ")\n",
    "\n",
    "total_pivot.to_csv(D_TABLES / 'promoter_ranks.tsv', sep='\\t')\n",
    "total_pivot\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Data</th>\n",
       "      <th colspan=\"6\" halign=\"left\">0.promoters</th>\n",
       "      <th colspan=\"4\" halign=\"left\">g4</th>\n",
       "      <th colspan=\"4\" halign=\"left\">h-dna</th>\n",
       "      <th colspan=\"4\" halign=\"left\">sidd</th>\n",
       "      <th colspan=\"4\" halign=\"left\">z-dna</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Description</th>\n",
       "      <th colspan=\"2\" halign=\"left\">all mouse (55,401)</th>\n",
       "      <th colspan=\"2\" halign=\"left\">with flipons &amp; miR (15,515)</th>\n",
       "      <th colspan=\"2\" halign=\"left\">with flipons (21,356)</th>\n",
       "      <th colspan=\"2\" halign=\"left\">flipons &amp; miR (2,323)</th>\n",
       "      <th colspan=\"2\" halign=\"left\">flipons (5,721)</th>\n",
       "      <th colspan=\"2\" halign=\"left\">flipons &amp; miR (226)</th>\n",
       "      <th colspan=\"2\" halign=\"left\">flipons (1,537)</th>\n",
       "      <th colspan=\"2\" halign=\"left\">flipons &amp; miR (720)</th>\n",
       "      <th colspan=\"2\" halign=\"left\">flipons (779)</th>\n",
       "      <th colspan=\"2\" halign=\"left\">flipons &amp; miR (1,288)</th>\n",
       "      <th colspan=\"2\" halign=\"left\">flipons (8,245)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>number_type</th>\n",
       "      <th>heptomer_counts</th>\n",
       "      <th>heptomer_rank</th>\n",
       "      <th>heptomer_counts</th>\n",
       "      <th>heptomer_rank</th>\n",
       "      <th>heptomer_counts</th>\n",
       "      <th>heptomer_rank</th>\n",
       "      <th>heptomer_counts</th>\n",
       "      <th>heptomer_rank</th>\n",
       "      <th>heptomer_counts</th>\n",
       "      <th>heptomer_rank</th>\n",
       "      <th>heptomer_counts</th>\n",
       "      <th>heptomer_rank</th>\n",
       "      <th>heptomer_counts</th>\n",
       "      <th>heptomer_rank</th>\n",
       "      <th>heptomer_counts</th>\n",
       "      <th>heptomer_rank</th>\n",
       "      <th>heptomer_counts</th>\n",
       "      <th>heptomer_rank</th>\n",
       "      <th>heptomer_counts</th>\n",
       "      <th>heptomer_rank</th>\n",
       "      <th>heptomer_counts</th>\n",
       "      <th>heptomer_rank</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>heptomer</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GGAGAGA</th>\n",
       "      <td>24594</td>\n",
       "      <td>110</td>\n",
       "      <td>7148</td>\n",
       "      <td>97</td>\n",
       "      <td>9824</td>\n",
       "      <td>99</td>\n",
       "      <td>24</td>\n",
       "      <td>568</td>\n",
       "      <td>24</td>\n",
       "      <td>1102</td>\n",
       "      <td>82</td>\n",
       "      <td>15</td>\n",
       "      <td>82</td>\n",
       "      <td>82</td>\n",
       "      <td>13</td>\n",
       "      <td>1451</td>\n",
       "      <td>13</td>\n",
       "      <td>1559</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTTTGTT</th>\n",
       "      <td>24299</td>\n",
       "      <td>117</td>\n",
       "      <td>5482</td>\n",
       "      <td>291</td>\n",
       "      <td>7515</td>\n",
       "      <td>301</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>2</td>\n",
       "      <td>361</td>\n",
       "      <td>2</td>\n",
       "      <td>838</td>\n",
       "      <td>46</td>\n",
       "      <td>188</td>\n",
       "      <td>46</td>\n",
       "      <td>205</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGTACCG</th>\n",
       "      <td>403</td>\n",
       "      <td>16347</td>\n",
       "      <td>96</td>\n",
       "      <td>16344</td>\n",
       "      <td>145</td>\n",
       "      <td>16342</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGGTACG</th>\n",
       "      <td>391</td>\n",
       "      <td>16352</td>\n",
       "      <td>96</td>\n",
       "      <td>16346</td>\n",
       "      <td>126</td>\n",
       "      <td>16359</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>422 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Data               0.promoters                                            \\\n",
       "Description all mouse (55,401)               with flipons & miR (15,515)   \n",
       "number_type    heptomer_counts heptomer_rank             heptomer_counts   \n",
       "heptomer                                                                   \n",
       "GGAGAGA                  24594           110                        7148   \n",
       "GTTTGTT                  24299           117                        5482   \n",
       "...                        ...           ...                         ...   \n",
       "CGTACCG                    403         16347                          96   \n",
       "CGGTACG                    391         16352                          96   \n",
       "\n",
       "Data                                                           \\\n",
       "Description               with flipons (21,356)                 \n",
       "number_type heptomer_rank       heptomer_counts heptomer_rank   \n",
       "heptomer                                                        \n",
       "GGAGAGA                97                  9824            99   \n",
       "GTTTGTT               291                  7515           301   \n",
       "...                   ...                   ...           ...   \n",
       "CGTACCG             16344                   145         16342   \n",
       "CGGTACG             16346                   126         16359   \n",
       "\n",
       "Data                           g4                                              \\\n",
       "Description flipons & miR (2,323)               flipons (5,721)                 \n",
       "number_type       heptomer_counts heptomer_rank heptomer_counts heptomer_rank   \n",
       "heptomer                                                                        \n",
       "GGAGAGA                        24           568              24          1102   \n",
       "GTTTGTT                      <NA>          <NA>            <NA>          <NA>   \n",
       "...                           ...           ...             ...           ...   \n",
       "CGTACCG                      <NA>          <NA>            <NA>          <NA>   \n",
       "CGGTACG                      <NA>          <NA>            <NA>          <NA>   \n",
       "\n",
       "Data                      h-dna                                              \\\n",
       "Description flipons & miR (226)               flipons (1,537)                 \n",
       "number_type     heptomer_counts heptomer_rank heptomer_counts heptomer_rank   \n",
       "heptomer                                                                      \n",
       "GGAGAGA                      82            15              82            82   \n",
       "GTTTGTT                       2           361               2           838   \n",
       "...                         ...           ...             ...           ...   \n",
       "CGTACCG                    <NA>          <NA>            <NA>          <NA>   \n",
       "CGGTACG                    <NA>          <NA>            <NA>          <NA>   \n",
       "\n",
       "Data                       sidd                                              \\\n",
       "Description flipons & miR (720)                 flipons (779)                 \n",
       "number_type     heptomer_counts heptomer_rank heptomer_counts heptomer_rank   \n",
       "heptomer                                                                      \n",
       "GGAGAGA                      13          1451              13          1559   \n",
       "GTTTGTT                      46           188              46           205   \n",
       "...                         ...           ...             ...           ...   \n",
       "CGTACCG                    <NA>          <NA>            <NA>          <NA>   \n",
       "CGGTACG                    <NA>          <NA>            <NA>          <NA>   \n",
       "\n",
       "Data                        z-dna                                              \n",
       "Description flipons & miR (1,288)               flipons (8,245)                \n",
       "number_type       heptomer_counts heptomer_rank heptomer_counts heptomer_rank  \n",
       "heptomer                                                                       \n",
       "GGAGAGA                      <NA>          <NA>            <NA>          <NA>  \n",
       "GTTTGTT                      <NA>          <NA>            <NA>          <NA>  \n",
       "...                           ...           ...             ...           ...  \n",
       "CGTACCG                      <NA>          <NA>            <NA>          <NA>  \n",
       "CGGTACG                      <NA>          <NA>            <NA>          <NA>  \n",
       "\n",
       "[422 rows x 22 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seed_total_pivot = total_pivot.loc[total_pivot.index.isin(mirna_list['Seed region']) | total_pivot.index.isin(mirna_list['Seed region (RC)']),:]\n",
    "\n",
    "seed_total_pivot.to_csv(D_TABLES / 'promoter_ranks_seed.tsv', sep='\\t')\n",
    "seed_total_pivot"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mirna",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0 | packaged by conda-forge | (main, Oct 25 2022, 06:24:40) [GCC 10.4.0]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "aa4ca5e43b0b74394dc05f972f57eb4922712b4f4450a8ee4f1bf77784cf2ccc"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
