{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Enable R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Loading required package: BiocGenerics\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: S4Vectors\n",
      "\n",
      "R[write to console]: Loading required package: stats4\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "R[write to console]: Loading required package: IRanges\n",
      "\n",
      "R[write to console]: Loading required package: GenomeInfoDb\n",
      "\n",
      "R[write to console]: Loading required package: GenomicRanges\n",
      "\n",
      "R[write to console]: Loading required package: AnnotationDbi\n",
      "\n",
      "R[write to console]: Loading required package: Biobase\n",
      "\n",
      "R[write to console]: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: ChIPseeker v1.34.0  For help: https://guangchuangyu.github.io/software/ChIPseeker\n",
      "\n",
      "If you use ChIPseeker in published research, please cite:\n",
      "Guangchuang Yu, Li-Gen Wang, Qing-Yu He. ChIPseeker: an R/Bioconductor package for ChIP peak annotation, comparison and visualization. Bioinformatics 2015, 31(14):2382-2383\n",
      "\n",
      "R[write to console]: clusterProfiler v4.6.0  For help: https://yulab-smu.top/biomedical-knowledge-mining-book/\n",
      "\n",
      "If you use clusterProfiler in published research, please cite:\n",
      "T Wu, E Hu, S Xu, M Chen, P Guo, Z Dai, T Feng, L Zhou, W Tang, L Zhan, X Fu, S Liu, X Bo, and G Yu. clusterProfiler 4.0: A universal enrichment tool for interpreting omics data. The Innovation. 2021, 2(3):100141\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: 'clusterProfiler'\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from 'package:AnnotationDbi':\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from 'package:IRanges':\n",
      "\n",
      "    slice\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from 'package:S4Vectors':\n",
      "\n",
      "    rename\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from 'package:stats':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "R[write to console]: \n",
      "Attaching package: 'dplyr'\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from 'package:AnnotationDbi':\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from 'package:Biobase':\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from 'package:GenomicRanges':\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "R[write to console]: The following object is masked from 'package:GenomeInfoDb':\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from 'package:IRanges':\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from 'package:S4Vectors':\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from 'package:BiocGenerics':\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from 'package:stats':\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "R[write to console]: The following objects are masked from 'package:base':\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "library(GenomicFeatures)\n",
    "library(ChIPpeakAnno)\n",
    "library(ChIPseeker)\n",
    "library(clusterProfiler)\n",
    "library(dplyr)\n",
    "library(stringr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Python imports and paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "\n",
    "d_root = Path(\"/home/fpavlov/projects/article_conserved_miRNA\")\n",
    "d_data = d_root / \"data\"\n",
    "\n",
    "# output\n",
    "f_mm_v25_annotation = d_data / \"gencode.vM25.annotation.gtf.gz\"\n",
    "f_mm_v25_annotation_str = str(d_data / \"gencode.vM25.annotation.gtf.gz\")\n",
    "f_flipon_to_gene_anno = d_data / \"flipon_to_gene_anno.tsv\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/fpavlov/proje 100%[===================>]  27,22M   882KB/s    in 26s     \n"
     ]
    }
   ],
   "source": [
    "# Download gencode gene annotation\n",
    "!wget -c -O {f_mm_v25_annotation} -q --show-progress http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.annotation.gtf.gz\n",
    "\n",
    "# Move the variable with path to R\n",
    "%R -i f_mm_v25_annotation_str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Import genomic features from the file as a GRanges object ... \n",
      "R[write to console]: OK\n",
      "\n",
      "R[write to console]: Prepare the 'metadata' data frame ... \n",
      "R[write to console]: OK\n",
      "\n",
      "R[write to console]: Make the TxDb object ... \n",
      "R[write to console]: OK\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "# Make a TxDb annotation\n",
    "txdb_all_annotation <- makeTxDbFromGFF(f_mm_v25_annotation_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "# Get gene names from the annotation\n",
    "gtf.file = f_mm_v25_annotation_str\n",
    "gtf.gr = rtracklayer::import(gtf.file) # creates a GRanges object\n",
    "gtf.df = as.data.frame(gtf.gr)\n",
    "genes = unique(gtf.df[ ,c(\"gene_id\",\"gene_name\")])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \n",
    "ssdna_list = [\n",
    "    str(d_data / \"mm10_kouzine_ssDNA_bed\" / \"mm10_kouzine_actb_ssdna_enriched_z-dna.bed\"),\n",
    "    str(d_data / \"mm10_kouzine_ssDNA_bed\" / \"mm10_kouzine_actb_ssdna_enriched_quadruplex.bed\"),\n",
    "    str(d_data / \"mm10_kouzine_ssDNA_bed\" / \"mm10_kouzine_actb_ssdna_enriched_sidd.bed\"),\n",
    "    str(d_data / \"mm10_kouzine_ssDNA_bed\" / \"mm10_kouzine_actb_ssdna_enriched_h-dna.bed\"),\n",
    "]\n",
    "\n",
    "%R -i ssdna_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">> loading peak file...\t\t\t\t 2022-11-14 20:54:54 \n",
      ">> preparing features information...\t\t 2022-11-14 20:54:54 \n",
      ">> identifying nearest features...\t\t 2022-11-14 20:54:54 \n",
      ">> calculating distance from peak to TSS...\t 2022-11-14 20:54:55 \n",
      ">> assigning genomic annotation...\t\t 2022-11-14 20:54:55 \n",
      ">> assigning chromosome lengths\t\t\t 2022-11-14 20:54:56 \n",
      ">> done...\t\t\t\t\t 2022-11-14 20:54:56 \n",
      ">> loading peak file...\t\t\t\t 2022-11-14 20:54:57 \n",
      ">> preparing features information...\t\t 2022-11-14 20:54:57 \n",
      ">> identifying nearest features...\t\t 2022-11-14 20:54:57 \n",
      ">> calculating distance from peak to TSS...\t 2022-11-14 20:54:57 \n",
      ">> assigning genomic annotation...\t\t 2022-11-14 20:54:57 \n",
      ">> assigning chromosome lengths\t\t\t 2022-11-14 20:54:59 \n",
      ">> done...\t\t\t\t\t 2022-11-14 20:54:59 \n",
      ">> loading peak file...\t\t\t\t 2022-11-14 20:54:59 \n",
      ">> preparing features information...\t\t 2022-11-14 20:54:59 \n",
      ">> identifying nearest features...\t\t 2022-11-14 20:54:59 \n",
      ">> calculating distance from peak to TSS...\t 2022-11-14 20:54:59 \n",
      ">> assigning genomic annotation...\t\t 2022-11-14 20:54:59 \n",
      ">> assigning chromosome lengths\t\t\t 2022-11-14 20:55:01 \n",
      ">> done...\t\t\t\t\t 2022-11-14 20:55:01 \n",
      ">> loading peak file...\t\t\t\t 2022-11-14 20:55:01 \n",
      ">> preparing features information...\t\t 2022-11-14 20:55:01 \n",
      ">> identifying nearest features...\t\t 2022-11-14 20:55:01 \n",
      ">> calculating distance from peak to TSS...\t 2022-11-14 20:55:01 \n",
      ">> assigning genomic annotation...\t\t 2022-11-14 20:55:01 \n",
      ">> assigning chromosome lengths\t\t\t 2022-11-14 20:55:03 \n",
      ">> done...\t\t\t\t\t 2022-11-14 20:55:03 \n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "df_final <- data.frame()\n",
    "\n",
    "for (ssdna in ssdna_list) {\n",
    "  peakAnno <- annotatePeak(ssdna, tssRegion=c(-3000, 3000), TxDb=txdb_all_annotation)\n",
    "\n",
    "  anno_df <- data.frame(peakAnno@anno)\n",
    "  res_df <- anno_df[c('V4', 'seqnames','start','end','geneStrand','annotation','geneId','distanceToTSS','transcriptId','geneStart','geneEnd')]\n",
    "  res_df$group <- str_replace(str_split(ssdna, 'enriched_')[[1]][2], '.bed', '')\n",
    "  res_df <- merge(res_df, genes, by.x='geneId', by.y='gene_id')\n",
    "\n",
    "  df_final <- bind_rows(df_final, res_df)\n",
    "  \n",
    "  print('--------------------------------------------')\n",
    "  print(ssdna)\n",
    "  print('--------------------------------------------')\n",
    "  print(peakAnno)\n",
    "  print(dim(res_df))\n",
    "}\n",
    "\n",
    "# write.table(df_final, 'flipon_to_gene_anno.tsv', append=FALSE, sep=\"\\t\", dec=\".\", row.names=FALSE, col.names = TRUE)\n",
    "# df_final\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>V4</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>geneId</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>geneStart</th>\n",
       "      <th>geneEnd</th>\n",
       "      <th>transcriptId</th>\n",
       "      <th>distanceToTSS</th>\n",
       "      <th>geneStrand</th>\n",
       "      <th>annotation</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr7:142572121-142572208</td>\n",
       "      <td>7</td>\n",
       "      <td>142572122</td>\n",
       "      <td>142572208</td>\n",
       "      <td>ENSMUSG00000000031.16</td>\n",
       "      <td>H19</td>\n",
       "      <td>142575634</td>\n",
       "      <td>142576538</td>\n",
       "      <td>ENSMUST00000140716.1</td>\n",
       "      <td>4330.0</td>\n",
       "      <td>2</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td>z-dna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr4:120414009-120414024</td>\n",
       "      <td>4</td>\n",
       "      <td>120414010</td>\n",
       "      <td>120414024</td>\n",
       "      <td>ENSMUSG00000000085.16</td>\n",
       "      <td>Scmh1</td>\n",
       "      <td>120405428</td>\n",
       "      <td>120463084</td>\n",
       "      <td>ENSMUST00000132116.7</td>\n",
       "      <td>8582.0</td>\n",
       "      <td>1</td>\n",
       "      <td>Intron (ENSMUST00000134375.7/ENSMUSG0000000008...</td>\n",
       "      <td>z-dna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr11:85832528-85832543</td>\n",
       "      <td>11</td>\n",
       "      <td>85832529</td>\n",
       "      <td>85832543</td>\n",
       "      <td>ENSMUSG00000000093.6</td>\n",
       "      <td>Tbx2</td>\n",
       "      <td>85832551</td>\n",
       "      <td>85841948</td>\n",
       "      <td>ENSMUST00000000095.6</td>\n",
       "      <td>-8.0</td>\n",
       "      <td>1</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>z-dna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11:85832843-85832860</td>\n",
       "      <td>11</td>\n",
       "      <td>85832844</td>\n",
       "      <td>85832860</td>\n",
       "      <td>ENSMUSG00000000093.6</td>\n",
       "      <td>Tbx2</td>\n",
       "      <td>85832551</td>\n",
       "      <td>85841948</td>\n",
       "      <td>ENSMUST00000000095.6</td>\n",
       "      <td>293.0</td>\n",
       "      <td>1</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>z-dna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr11:85833014-85833035</td>\n",
       "      <td>11</td>\n",
       "      <td>85833015</td>\n",
       "      <td>85833035</td>\n",
       "      <td>ENSMUSG00000000093.6</td>\n",
       "      <td>Tbx2</td>\n",
       "      <td>85832551</td>\n",
       "      <td>85841948</td>\n",
       "      <td>ENSMUST00000000095.6</td>\n",
       "      <td>464.0</td>\n",
       "      <td>1</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>z-dna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77696</th>\n",
       "      <td>chr7:27899690-27899773</td>\n",
       "      <td>7</td>\n",
       "      <td>27899691</td>\n",
       "      <td>27899773</td>\n",
       "      <td>ENSMUSG00000118641.1</td>\n",
       "      <td>AC139063.1</td>\n",
       "      <td>27894062</td>\n",
       "      <td>27894118</td>\n",
       "      <td>ENSMUST00000239484.1</td>\n",
       "      <td>-5573.0</td>\n",
       "      <td>2</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td>h-dna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77697</th>\n",
       "      <td>chr7:27899779-27899799</td>\n",
       "      <td>7</td>\n",
       "      <td>27899780</td>\n",
       "      <td>27899799</td>\n",
       "      <td>ENSMUSG00000118641.1</td>\n",
       "      <td>AC139063.1</td>\n",
       "      <td>27894062</td>\n",
       "      <td>27894118</td>\n",
       "      <td>ENSMUST00000239484.1</td>\n",
       "      <td>-5662.0</td>\n",
       "      <td>2</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td>h-dna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77698</th>\n",
       "      <td>chr7:27899863-27899888</td>\n",
       "      <td>7</td>\n",
       "      <td>27899864</td>\n",
       "      <td>27899888</td>\n",
       "      <td>ENSMUSG00000118641.1</td>\n",
       "      <td>AC139063.1</td>\n",
       "      <td>27894062</td>\n",
       "      <td>27894118</td>\n",
       "      <td>ENSMUST00000239484.1</td>\n",
       "      <td>-5746.0</td>\n",
       "      <td>2</td>\n",
       "      <td>Distal Intergenic</td>\n",
       "      <td>h-dna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77699</th>\n",
       "      <td>chr17:39848378-39848395</td>\n",
       "      <td>17</td>\n",
       "      <td>39848379</td>\n",
       "      <td>39848395</td>\n",
       "      <td>ENSMUSG00000118642.1</td>\n",
       "      <td>CT010467.2</td>\n",
       "      <td>39848103</td>\n",
       "      <td>39848827</td>\n",
       "      <td>ENSMUST00000205406.1</td>\n",
       "      <td>432.0</td>\n",
       "      <td>2</td>\n",
       "      <td>Promoter (&lt;=1kb)</td>\n",
       "      <td>h-dna</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77700</th>\n",
       "      <td>chr13:32633603-32633664</td>\n",
       "      <td>13</td>\n",
       "      <td>32633604</td>\n",
       "      <td>32633664</td>\n",
       "      <td>ENSMUSG00000118652.1</td>\n",
       "      <td>AL645799.1</td>\n",
       "      <td>32614016</td>\n",
       "      <td>32682515</td>\n",
       "      <td>ENSMUST00000239419.1</td>\n",
       "      <td>19588.0</td>\n",
       "      <td>1</td>\n",
       "      <td>Intron (ENSMUST00000239419.1/ENSMUSG0000011865...</td>\n",
       "      <td>h-dna</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77700 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             V4  seqnames      start        end  \\\n",
       "1      chr7:142572121-142572208         7  142572122  142572208   \n",
       "2      chr4:120414009-120414024         4  120414010  120414024   \n",
       "3       chr11:85832528-85832543        11   85832529   85832543   \n",
       "4       chr11:85832843-85832860        11   85832844   85832860   \n",
       "5       chr11:85833014-85833035        11   85833015   85833035   \n",
       "...                         ...       ...        ...        ...   \n",
       "77696    chr7:27899690-27899773         7   27899691   27899773   \n",
       "77697    chr7:27899779-27899799         7   27899780   27899799   \n",
       "77698    chr7:27899863-27899888         7   27899864   27899888   \n",
       "77699   chr17:39848378-39848395        17   39848379   39848395   \n",
       "77700   chr13:32633603-32633664        13   32633604   32633664   \n",
       "\n",
       "                      geneId   gene_name  geneStart    geneEnd  \\\n",
       "1      ENSMUSG00000000031.16         H19  142575634  142576538   \n",
       "2      ENSMUSG00000000085.16       Scmh1  120405428  120463084   \n",
       "3       ENSMUSG00000000093.6        Tbx2   85832551   85841948   \n",
       "4       ENSMUSG00000000093.6        Tbx2   85832551   85841948   \n",
       "5       ENSMUSG00000000093.6        Tbx2   85832551   85841948   \n",
       "...                      ...         ...        ...        ...   \n",
       "77696   ENSMUSG00000118641.1  AC139063.1   27894062   27894118   \n",
       "77697   ENSMUSG00000118641.1  AC139063.1   27894062   27894118   \n",
       "77698   ENSMUSG00000118641.1  AC139063.1   27894062   27894118   \n",
       "77699   ENSMUSG00000118642.1  CT010467.2   39848103   39848827   \n",
       "77700   ENSMUSG00000118652.1  AL645799.1   32614016   32682515   \n",
       "\n",
       "               transcriptId  distanceToTSS  geneStrand  \\\n",
       "1      ENSMUST00000140716.1         4330.0           2   \n",
       "2      ENSMUST00000132116.7         8582.0           1   \n",
       "3      ENSMUST00000000095.6           -8.0           1   \n",
       "4      ENSMUST00000000095.6          293.0           1   \n",
       "5      ENSMUST00000000095.6          464.0           1   \n",
       "...                     ...            ...         ...   \n",
       "77696  ENSMUST00000239484.1        -5573.0           2   \n",
       "77697  ENSMUST00000239484.1        -5662.0           2   \n",
       "77698  ENSMUST00000239484.1        -5746.0           2   \n",
       "77699  ENSMUST00000205406.1          432.0           2   \n",
       "77700  ENSMUST00000239419.1        19588.0           1   \n",
       "\n",
       "                                              annotation  group  \n",
       "1                                      Distal Intergenic  z-dna  \n",
       "2      Intron (ENSMUST00000134375.7/ENSMUSG0000000008...  z-dna  \n",
       "3                                       Promoter (<=1kb)  z-dna  \n",
       "4                                       Promoter (<=1kb)  z-dna  \n",
       "5                                       Promoter (<=1kb)  z-dna  \n",
       "...                                                  ...    ...  \n",
       "77696                                  Distal Intergenic  h-dna  \n",
       "77697                                  Distal Intergenic  h-dna  \n",
       "77698                                  Distal Intergenic  h-dna  \n",
       "77699                                   Promoter (<=1kb)  h-dna  \n",
       "77700  Intron (ENSMUST00000239419.1/ENSMUSG0000011865...  h-dna  \n",
       "\n",
       "[77700 rows x 13 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_final = %R df_final\n",
    "\n",
    "df_final = df_final[['V4', 'seqnames', 'start', 'end', 'geneId', 'gene_name', 'geneStart', 'geneEnd', 'transcriptId', 'distanceToTSS', 'geneStrand', 'annotation', 'group']]\n",
    "\n",
    "df_final.to_csv(f_flipon_to_gene_anno, sep='\\t', index=False)\n",
    "df_final"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.11.0 ('mirna')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "aa4ca5e43b0b74394dc05f972f57eb4922712b4f4450a8ee4f1bf77784cf2ccc"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
